diff --git a/crossdev-20101011/wrappers/site/linux-gnueabi b/crossdev-20101011.patch/wrappers/site/linux-gnueabi
index b022182..855c623 100644
--- a/crossdev-20101011/wrappers/site/linux-gnueabi
+++ b/crossdev-20101011.patch/wrappers/site/linux-gnueabi
@@ -11,7 +11,7 @@ ac_cv_func_realloc_0_nonnull=yes
 ac_cv_func_setgrent_void=yes
 ac_cv_func_setpgrp_void=yes
 ac_cv_func_strnlen_working=yes
-ac_cv_have_abstract_sockets=no
+ac_cv_have_abstract_sockets=yes
 ac_cv_header_pwd_h=yes
 ac_cv_printf_zd_format=yes
 enable_malloc0returnsnull=yes
