diff --git a/var/tmp/portage/sys-devel/crossdev-20101011/image/usr/bin/crossdev b/usr/bin/crossdev
index bd73021..7d45f88 100755
--- a/var/tmp/portage/sys-devel/crossdev-20101011/image/usr/bin/crossdev
+++ b/usr/bin/crossdev
@@ -619,6 +619,24 @@ set_portage() {
 	set_links ${cat} ${pkg}
 	set_env ${pkg} "${env}"
 }
+set_metadata() {
+	local s srcdir=${PORTDIR} d
+
+	d="${PORTDIR_OVERLAY}"/metadata
+	if [[ ! -L ${d} && -d ${d} ]] ; then #211386
+		einfo "Leaving existing dir alone: ${d}"
+	else
+		rm -f "${d}"
+		for s in ${PORTDIR_OVERLAYS} ; do
+			if [[ -e ${s}/metadata/layout.conf ]] ; then
+				einfo "Using metadata from ${s} instead of ${PORTDIR}"
+				srcdir=${s}
+				break
+			fi
+		done
+		ln -s "${srcdir}"/metadata "${d}"
+	fi
+}
 
 mkdir -p "${CONFIGROOT}"
 check_trailing_newline "${CONFIGROOT}"/categories
@@ -631,6 +649,7 @@ for f in package.{keywords,mask,use} ; do
 	mkdir -p ${f} || exit 1
 	rm -f ${f}/cross-${CTARGET}
 done
+set_metadata
 
 set_portage ${BCAT} ${BPKG} ${BVER} "${BENV}"
 set_portage ${GCAT} ${GPKG} ${GVER} "${GENV}"
