hack in nds32 support until they mainline their code

--- a/crossdev
+++ b/crossdev
@@ -214,6 +216,13 @@
 		      GUSE="-fortran -openmp";
 		      WITH_DEF_HEADERS="no";;
 
+		nds32*)
+		      BCAT="dev-embedded"; BPKG="nds32-binutils";
+		      GCAT="dev-embedded"; GPKG="nds32-gcc";
+		      DCAT="dev-embedded"; DPKG="nds32-gdb";
+		      GUSE="-fortran";
+		      WITH_DEF_HEADERS="no";;
+
 		nios2*)
 		      BPKG="binutils-nios2";
 		      GPKG="gcc-nios2";;
@@ -263,7 +274,7 @@
 			KPKG="[none]"
 			STAGE_DEFAULT=${STAGE_LIBC}
 			GUSE+=" -openmp" #489798
-			MULTILIB_USE="yes" #407275
+			MULTILIB_USE="no" #407275
 			WITH_DEF_HEADERS="no"
 			;;
 
