Index: chrome/browser/chromeos/arc/arc_auth_service.cc
diff --git a/chrome/browser/chromeos/arc/arc_auth_service.cc b/chrome/browser/chromeos/arc/arc_auth_service.cc
index 0c12cb97c2f2c5d033fcf5c624760fb23445bf31..a4df6dff67740257b68b7e267bde805c196c9629 100644
--- a/chrome/browser/chromeos/arc/arc_auth_service.cc
+++ b/chrome/browser/chromeos/arc/arc_auth_service.cc
@@ -116,9 +116,10 @@ void ArcAuthService::GetAuthCodeDeprecated(
 void ArcAuthService::GetAuthCode(const GetAuthCodeCallback& callback) {
   DCHECK(thread_checker_.CalledOnValidThread());
 
-  std::string auth_code = GetAndResetAuthCode();
-  if (!auth_code.empty()) {
-    callback.Run(mojo::String(auth_code), !IsOptInVerificationDisabled());
+  const std::string auth_code = GetAndResetAuthCode();
+  const bool verification_disabled = IsOptInVerificationDisabled();
+  if (!auth_code.empty() || verification_disabled) {
+    callback.Run(mojo::String(auth_code), !verification_disabled);
     return;
   }
 
