diff -Naur mosys-1.0.07.275/lib/smbios/smbios.c mosys-1.0.07.275-patch/lib/smbios/smbios.c
--- mosys-1.0.07.275/lib/smbios/smbios.c	2011-01-05 17:48:51.000000000 -0800
+++ mosys-1.0.07.275-patch/lib/smbios/smbios.c	2011-03-01 16:19:40.000000000 -0800
@@ -177,6 +177,32 @@
 	}
 
 	if (!found) {
+		FILE *fp;
+		char *line = NULL;
+		size_t bytes;
+
+		lprintf(LOG_DEBUG, "%s: Attempting to get SMBIOS base address "
+			"from /sys/kernel/debug/efi_smbios_base\n", __func__);
+
+		/* ChromeOS currently exports smbios base in debugfs. */
+		fp = fopen("/sys/kernel/debug/efi_smbios_base", "r");
+		if ((fp > 0) && (getline(&line, &bytes, fp)) > 0) {
+			baseaddr = strtoull(line, NULL, 0);
+			lprintf(LOG_DEBUG, "%s: SMBIOS=0x%08x\n",
+				__func__, baseaddr);
+			data = mmio_map(intf, O_RDONLY, baseaddr, len);
+			offset = 0;
+			found = 1;
+
+			if (line) {
+				free(line);
+				line = NULL;
+			}
+			fclose(fp);
+		}
+	}
+
+	if (!found) {
 		char *buf;
 		int klog_size;
 		size_t klog_offset;
@@ -215,43 +241,6 @@
 		free(buf);
 	}
 
-	/*
-	 * Try obtaining SMBIOS address from /var/log/messages (first entry
-	 * found will be used). This should be considered unreliable and used
-	 * as a last resort.
-	 */
-	if (!found) {
-		FILE *fp;
-		char *line = NULL;
-		size_t line_offset, bytes;
-
-		lprintf(LOG_DEBUG, "%s: Attempting to get SMBIOS base address "
-		                   "from /var/log/messages\n", __func__);
-		fp = fopen("/var/log/messages", "r");
-		/* read /var/log/messages one line at a time since it can get
-		   rather large */
-		while (fp && ((getline(&line, &bytes, fp)) > 0)) {
-			if (find_pattern(line, bytes,
-			             "SMBIOS=", 7, 1, &line_offset) == 0) {
-				baseaddr = strtoull(line + line_offset + 7,
-				                    NULL, 0);
-				lprintf(LOG_DEBUG, "%s: SMBIOS=0x%08x\n",
-					           __func__, baseaddr);
-				data = mmio_map(intf, O_RDONLY, baseaddr, len);
-				offset = 0;
-				found = 1;
-			}
-			if (line) {
-				free(line);
-				line = NULL;
-			}
-
-			if (found)
-				break;
-		}
-		fclose(fp);
-	}
-
 	if (!found) {
 		lprintf(LOG_DEBUG, "Unable to find SMBIOS entry.\n");
 		return -1;
