diff --git a/src/gallium/drivers/llvmpipe/lp_texture.c b/src/gallium/drivers/llvmpipe/lp_texture.c
index f601181..6fd7790 100644
--- a/src/gallium/drivers/llvmpipe/lp_texture.c
+++ b/src/gallium/drivers/llvmpipe/lp_texture.c
@@ 117,12 +114,6 @@ llvmpipe_texture_layout(struct llvmpipe_screen *screen,
          else
             lpr->row_stride[level] = align(nblocksx * block_size, util_cpu_caps.cacheline);
 
-         /* if row_stride * height > LP_MAX_TEXTURE_SIZE */
-         if (lpr->row_stride[level] > LP_MAX_TEXTURE_SIZE / nblocksy) {
-            /* image too large */
-            goto fail;
-         }
-
          lpr->img_stride[level] = lpr->row_stride[level] * nblocksy;
       }
 
@@ -142,13 +134,6 @@ llvmpipe_texture_layout(struct llvmpipe_screen *screen,
          lpr->num_slices_faces[level] = num_slices;
       }
 
-      /* if img_stride * num_slices_faces > LP_MAX_TEXTURE_SIZE */
-      if (lpr->img_stride[level] >
-          LP_MAX_TEXTURE_SIZE / lpr->num_slices_faces[level]) {
-         /* volume too large */
-         goto fail;
-      }
-
       total_size += (uint64_t) lpr->num_slices_faces[level]
                   * (uint64_t) lpr->img_stride[level];
       if (total_size > LP_MAX_TEXTURE_SIZE) {
