diff -upr Mesa-7.11-p1/src/egl/wayland/wayland-drm/Makefile Mesa-7.11-p1.pkg/src/egl/wayland/wayland-drm/Makefile
--- Mesa-7.11-p1/src/egl/wayland/wayland-drm/Makefile	2011-09-27 16:10:38.000000000 -0700
+++ Mesa-7.11-p1.pkg/src/egl/wayland/wayland-drm/Makefile	2011-09-27 17:06:12.060093000 -0700
@@ -16,8 +16,8 @@ wayland_drm_INCLUDES = \
 		-I$(TOP)/include
 
 # Generate protocol sources
-prefix=$(shell pkg-config --variable=prefix wayland-server)
-exec_prefx=$(shell pkg-config --variable=exec_prefix wayland-server)
+prefix=$(shell $(PKG_CONFIG) --variable=prefix wayland-server)
+exec_prefx=$(shell $(PKG_CONFIG) --variable=exec_prefix wayland-server)
 wayland_protocoldir = $(PWD)/protocol
 wayland_scanner=$(exec_prefix)/bin/wayland-scanner
 
diff -upr Mesa-7.11-p1/src/gallium/state_trackers/dri/drm/Makefile Mesa-7.11-p1.pkg/src/gallium/state_trackers/dri/drm/Makefile
--- Mesa-7.11-p1/src/gallium/state_trackers/dri/drm/Makefile	2011-09-27 16:10:43.000000000 -0700
+++ Mesa-7.11-p1.pkg/src/gallium/state_trackers/dri/drm/Makefile	2011-09-27 17:06:12.071091000 -0700
@@ -10,7 +10,7 @@ LIBRARY_INCLUDES = \
 	-I$(TOP)/src/gallium/state_trackers/dri/common \
 	-I$(TOP)/src/mesa/drivers/dri/common \
 	-I$(TOP)/src/mesa/main \
-         $(shell pkg-config --cflags-only-I libdrm)
+         $(shell $(PKG_CONFIG) --cflags-only-I libdrm)
 
 
 C_SOURCES = \
diff -upr Mesa-7.11-p1/src/gallium/state_trackers/egl/Makefile Mesa-7.11-p1.pkg/src/gallium/state_trackers/egl/Makefile
--- Mesa-7.11-p1/src/gallium/state_trackers/egl/Makefile	2011-09-27 16:10:43.000000000 -0700
+++ Mesa-7.11-p1.pkg/src/gallium/state_trackers/egl/Makefile	2011-09-27 17:06:12.083093000 -0700
@@ -18,7 +18,7 @@ x11_INCLUDES = \
 	-I$(TOP)/src/mapi \
 	-I$(TOP)/src/mesa \
 	$(X11_CFLAGS) \
-	$(shell pkg-config --cflags-only-I libdrm dri2proto)
+	$(shell $(PKG_CONFIG) --cflags-only-I libdrm dri2proto)
 
 x11_SOURCES = $(wildcard x11/*.c) \
 	      $(TOP)/src/glx/dri2.c
@@ -28,12 +28,12 @@ wayland_INCLUDES = \
 	-I$(TOP)/src/gallium/winsys \
 	-I$(TOP)/src/egl/wayland/wayland-egl \
 	-I$(TOP)/src/egl/wayland/wayland-drm \
-	$(shell pkg-config --cflags-only-I libdrm wayland-client)
+	$(shell $(PKG_CONFIG) --cflags-only-I libdrm wayland-client)
 
 wayland_SOURCES = $(wildcard wayland/*.c)
 wayland_OBJECTS = $(wayland_SOURCES:.c=.o)
 
-drm_INCLUDES = -I$(TOP)/src/gallium/winsys $(shell pkg-config --cflags-only-I libdrm) \
+drm_INCLUDES = -I$(TOP)/src/gallium/winsys $(shell $(PKG_CONFIG) --cflags-only-I libdrm)
 	       -I$(TOP)/src/gbm/main -I$(TOP)/src/gallium/state_trackers/gbm
 drm_SOURCES = $(wildcard drm/*.c)
 drm_OBJECTS = $(drm_SOURCES:.c=.o)
diff -upr Mesa-7.11-p1/src/gallium/state_trackers/va/Makefile Mesa-7.11-p1.pkg/src/gallium/state_trackers/va/Makefile
--- Mesa-7.11-p1/src/gallium/state_trackers/va/Makefile	2011-09-27 16:10:43.000000000 -0700
+++ Mesa-7.11-p1.pkg/src/gallium/state_trackers/va/Makefile	2011-09-27 17:10:19.104085000 -0700
@@ -8,7 +8,7 @@ VA_MINOR = 3
 LIBRARY_DEFINES = -DVER_MAJOR=$(VA_MAJOR) -DVER_MINOR=$(VA_MINOR) $(STATE_TRACKER_DEFINES)
 
 LIBRARY_INCLUDES = \
-	$(shell pkg-config --cflags-only-I libva) \
+	$(shell $(PKG_CONFIG) --cflags-only-I libva) \
 	-I$(TOP)/src/gallium/winsys/g3dvl
 
 C_SOURCES = htab.c \
diff -upr Mesa-7.11-p1/src/gallium/state_trackers/vdpau/Makefile Mesa-7.11-p1.pkg/src/gallium/state_trackers/vdpau/Makefile
--- Mesa-7.11-p1/src/gallium/state_trackers/vdpau/Makefile	2011-09-27 16:10:43.000000000 -0700
+++ Mesa-7.11-p1.pkg/src/gallium/state_trackers/vdpau/Makefile	2011-09-27 17:10:31.494120000 -0700
@@ -8,7 +8,7 @@ VDPAU_MINOR = 0
 LIBRARY_DEFINES = -DVER_MAJOR=$(VDPAU_MAJOR) -DVER_MINOR=$(VDPAU_MINOR) $(STATE_TRACKER_DEFINES)
 
 LIBRARY_INCLUDES = \
-	$(shell pkg-config --cflags-only-I vdpau) \
+	$(shell $(PKG_CONFIG) --cflags-only-I vdpau) \
 	-I$(TOP)/src/gallium/winsys/g3dvl
 
 C_SOURCES = htab.c \
diff -upr Mesa-7.11-p1/src/gallium/state_trackers/xorg/Makefile Mesa-7.11-p1.pkg/src/gallium/state_trackers/xorg/Makefile
--- Mesa-7.11-p1/src/gallium/state_trackers/xorg/Makefile	2011-09-27 16:10:43.000000000 -0700
+++ Mesa-7.11-p1.pkg/src/gallium/state_trackers/xorg/Makefile	2011-09-27 17:06:12.099090000 -0700
@@ -5,12 +5,12 @@ LIBNAME = xorgtracker
 
 LIBRARY_INCLUDES = \
 	-DHAVE_CONFIG_H \
-	$(shell pkg-config xextproto --atleast-version=7.0.99.1 \
+	$(shell $(PKG_CONFIG) xextproto --atleast-version=7.0.99.1 \
 				&& echo "-DHAVE_XEXTPROTO_71") \
-	$(shell pkg-config libkms --atleast-version=1.0 \
+	$(shell $(PKG_CONFIG) libkms --atleast-version=1.0 \
 				&& echo "-DHAVE_LIBKMS") \
-	$(shell pkg-config libkms --silence-errors --cflags-only-I) \
-	$(shell pkg-config --cflags-only-I pixman-1 xorg-server libdrm xproto dri2proto) \
+	$(shell $(PKG_CONFIG) libkms --silence-errors --cflags-only-I) \
+	$(shell $(PKG_CONFIG) --cflags-only-I pixman-1 xorg-server libdrm xproto dri2proto) \
 	-I$(TOP)/src/gallium/include \
 	-I$(TOP)/src/gallium/auxiliary \
 	-I$(TOP)/include \
diff -upr Mesa-7.11-p1/src/gallium/state_trackers/xorg/xvmc/Makefile Mesa-7.11-p1.pkg/src/gallium/state_trackers/xorg/xvmc/Makefile
--- Mesa-7.11-p1/src/gallium/state_trackers/xorg/xvmc/Makefile	2011-09-27 16:10:44.000000000 -0700
+++ Mesa-7.11-p1.pkg/src/gallium/state_trackers/xorg/xvmc/Makefile	2011-09-27 17:10:43.473145000 -0700
@@ -4,7 +4,7 @@ include $(TOP)/configs/current
 LIBNAME = xvmctracker
 
 LIBRARY_INCLUDES = \
-	$(shell pkg-config --cflags-only-I xvmc) \
+	$(shell $(PKG_CONFIG) --cflags-only-I xvmc) \
 	-I$(TOP)/src/gallium/winsys/g3dvl
 
 C_SOURCES = block.c \
diff -upr Mesa-7.11-p1/src/gallium/targets/dri-nouveau/Makefile Mesa-7.11-p1.pkg/src/gallium/targets/dri-nouveau/Makefile
--- Mesa-7.11-p1/src/gallium/targets/dri-nouveau/Makefile	2011-09-27 16:10:44.000000000 -0700
+++ Mesa-7.11-p1.pkg/src/gallium/targets/dri-nouveau/Makefile	2011-09-27 17:06:12.102092000 -0700
@@ -23,6 +23,6 @@ DRIVER_DEFINES = \
 
 include ../Makefile.dri
 
-DRI_LIB_DEPS += $(shell pkg-config libdrm_nouveau --libs)
+DRI_LIB_DEPS += $(shell $(PKG_CONFIG) libdrm_nouveau --libs)
 
 symlinks:
diff -upr Mesa-7.11-p1/src/gallium/targets/Makefile.xorg Mesa-7.11-p1.pkg/src/gallium/targets/Makefile.xorg
--- Mesa-7.11-p1/src/gallium/targets/Makefile.xorg	2011-09-27 16:10:44.000000000 -0700
+++ Mesa-7.11-p1.pkg/src/gallium/targets/Makefile.xorg	2011-09-27 17:09:36.168569000 -0700
@@ -19,7 +19,7 @@ OBJECTS = $(C_SOURCES:.c=.o) \
 	$(ASM_SOURCES:.S=.o)
 
 INCLUDES = \
-	$(shell pkg-config --cflags-only-I pixman-1 xorg-server libdrm xproto) \
+	$(shell $(PKG_CONFIG) --cflags-only-I pixman-1 xorg-server libdrm xproto) \
 	-I$(TOP)/src/gallium/include \
 	-I$(TOP)/src/gallium/drivers \
 	-I$(TOP)/src/gallium/auxiliary \
diff -upr Mesa-7.11-p1/src/gallium/targets/va-r300/Makefile Mesa-7.11-p1.pkg/src/gallium/targets/va-r300/Makefile
--- Mesa-7.11-p1/src/gallium/targets/va-r300/Makefile	2011-09-27 16:10:44.000000000 -0700
+++ Mesa-7.11-p1.pkg/src/gallium/targets/va-r300/Makefile	2011-09-27 17:11:23.702899000 -0700
@@ -19,7 +19,7 @@ C_SOURCES = \
 	$(COMMON_GALLIUM_SOURCES) \
 	$(DRIVER_SOURCES)
 
-DRIVER_LIBS = $(shell pkg-config libdrm --libs) -lXfixes
+DRIVER_LIBS = $(shell $(PKG_CONFIG) libdrm --libs) -lXfixes
 
 include ../Makefile.va
 
diff -upr Mesa-7.11-p1/src/gallium/targets/va-r600/Makefile Mesa-7.11-p1.pkg/src/gallium/targets/va-r600/Makefile
--- Mesa-7.11-p1/src/gallium/targets/va-r600/Makefile	2011-09-27 16:10:44.000000000 -0700
+++ Mesa-7.11-p1.pkg/src/gallium/targets/va-r600/Makefile	2011-09-27 17:11:34.714007000 -0700
@@ -20,7 +20,7 @@ C_SOURCES = \
 	$(COMMON_GALLIUM_SOURCES) \
 	$(DRIVER_SOURCES)
 
-DRIVER_LIBS = $(shell pkg-config libdrm --libs) -lXfixes
+DRIVER_LIBS = $(shell $(PKG_CONFIG) libdrm --libs) -lXfixes
 
 include ../Makefile.va
 
diff -upr Mesa-7.11-p1/src/gallium/targets/vdpau-r300/Makefile Mesa-7.11-p1.pkg/src/gallium/targets/vdpau-r300/Makefile
--- Mesa-7.11-p1/src/gallium/targets/vdpau-r300/Makefile	2011-09-27 16:10:44.000000000 -0700
+++ Mesa-7.11-p1.pkg/src/gallium/targets/vdpau-r300/Makefile	2011-09-27 17:11:42.563364000 -0700
@@ -20,7 +20,7 @@ C_SOURCES = \
 	$(COMMON_GALLIUM_SOURCES) \
 	$(DRIVER_SOURCES)
 
-DRIVER_LIBS = $(shell pkg-config libdrm --libs) -lXfixes
+DRIVER_LIBS = $(shell $(PKG_CONFIG) libdrm --libs) -lXfixes
 
 include ../Makefile.vdpau
 
diff -upr Mesa-7.11-p1/src/gallium/targets/vdpau-r600/Makefile Mesa-7.11-p1.pkg/src/gallium/targets/vdpau-r600/Makefile
--- Mesa-7.11-p1/src/gallium/targets/vdpau-r600/Makefile	2011-09-27 16:10:44.000000000 -0700
+++ Mesa-7.11-p1.pkg/src/gallium/targets/vdpau-r600/Makefile	2011-09-27 17:11:49.573795000 -0700
@@ -16,7 +16,7 @@ C_SOURCES = \
 	$(COMMON_GALLIUM_SOURCES) \
 	$(DRIVER_SOURCES)
 
-DRIVER_LIBS = $(shell pkg-config libdrm --libs) -lXfixes
+DRIVER_LIBS = $(shell $(PKG_CONFIG) libdrm --libs) -lXfixes
 
 include ../Makefile.vdpau
 
diff -upr Mesa-7.11-p1/src/gallium/targets/xorg-i915/Makefile Mesa-7.11-p1.pkg/src/gallium/targets/xorg-i915/Makefile
--- Mesa-7.11-p1/src/gallium/targets/xorg-i915/Makefile	2011-09-27 16:10:44.000000000 -0700
+++ Mesa-7.11-p1.pkg/src/gallium/targets/xorg-i915/Makefile	2011-09-27 17:06:12.116097000 -0700
@@ -19,6 +19,6 @@ DRIVER_PIPES = \
 	$(TOP)/src/gallium/drivers/rbug/librbug.a
 
 DRIVER_LINKS = \
-	$(shell pkg-config --libs libdrm libdrm_intel)
+	$(shell $(PKG_CONFIG) --libs libdrm libdrm_intel)
 
 include ../Makefile.xorg
diff -upr Mesa-7.11-p1/src/gallium/targets/xorg-i965/Makefile Mesa-7.11-p1.pkg/src/gallium/targets/xorg-i965/Makefile
--- Mesa-7.11-p1/src/gallium/targets/xorg-i965/Makefile	2011-09-27 16:10:44.000000000 -0700
+++ Mesa-7.11-p1.pkg/src/gallium/targets/xorg-i965/Makefile	2011-09-27 17:06:12.120090000 -0700
@@ -21,6 +21,6 @@ DRIVER_PIPES = \
 	$(TOP)/src/gallium/drivers/softpipe/libsoftpipe.a
 
 DRIVER_LINKS = \
-	$(shell pkg-config --libs libdrm libdrm_intel)
+	$(shell $(PKG_CONFIG) --libs libdrm libdrm_intel)
 
 include ../Makefile.xorg
diff -upr Mesa-7.11-p1/src/gallium/targets/xorg-nouveau/Makefile Mesa-7.11-p1.pkg/src/gallium/targets/xorg-nouveau/Makefile
--- Mesa-7.11-p1/src/gallium/targets/xorg-nouveau/Makefile	2011-09-27 16:10:44.000000000 -0700
+++ Mesa-7.11-p1.pkg/src/gallium/targets/xorg-nouveau/Makefile	2011-09-27 17:06:12.134090000 -0700
@@ -21,9 +21,9 @@ DRIVER_PIPES = \
 	$(TOP)/src/gallium/drivers/rbug/librbug.a
 
 DRIVER_LINKS = \
-	$(shell pkg-config --libs libdrm libdrm_nouveau)
+	$(shell $(PKG_CONFIG) --libs libdrm libdrm_nouveau)
 
 DRIVER_INCLUDES = \
-	$(shell pkg-config --cflags-only-I libdrm libdrm_nouveau xf86driproto)
+	$(shell $(PKG_CONFIG) --cflags-only-I libdrm libdrm_nouveau xf86driproto)
 
 include ../Makefile.xorg
diff -upr Mesa-7.11-p1/src/gallium/targets/xorg-r300/Makefile Mesa-7.11-p1.pkg/src/gallium/targets/xorg-r300/Makefile
--- Mesa-7.11-p1/src/gallium/targets/xorg-r300/Makefile	2011-09-27 16:10:44.000000000 -0700
+++ Mesa-7.11-p1.pkg/src/gallium/targets/xorg-r300/Makefile	2011-09-27 17:06:12.139104000 -0700
@@ -19,6 +19,6 @@ DRIVER_PIPES = \
 	$(TOP)/src/gallium/drivers/rbug/librbug.a
 
 DRIVER_LINKS = \
-	$(shell pkg-config --libs libdrm)
+	$(shell $(PKG_CONFIG) --libs libdrm)
 
 include ../Makefile.xorg
diff -upr Mesa-7.11-p1/src/gallium/targets/xorg-r600/Makefile Mesa-7.11-p1.pkg/src/gallium/targets/xorg-r600/Makefile
--- Mesa-7.11-p1/src/gallium/targets/xorg-r600/Makefile	2011-09-27 16:10:44.000000000 -0700
+++ Mesa-7.11-p1.pkg/src/gallium/targets/xorg-r600/Makefile	2011-09-27 17:11:57.512153000 -0700
@@ -20,6 +20,6 @@ DRIVER_PIPES = \
 	$(TOP)/src/gallium/drivers/rbug/librbug.a
 
 DRIVER_LINKS = \
-	$(shell pkg-config --libs libdrm)
+	$(shell $(PKG_CONFIG) --libs libdrm)
 
 include ../Makefile.xorg
diff -upr Mesa-7.11-p1/src/gallium/targets/xorg-vmwgfx/Makefile Mesa-7.11-p1.pkg/src/gallium/targets/xorg-vmwgfx/Makefile
--- Mesa-7.11-p1/src/gallium/targets/xorg-vmwgfx/Makefile	2011-09-27 16:10:44.000000000 -0700
+++ Mesa-7.11-p1.pkg/src/gallium/targets/xorg-vmwgfx/Makefile	2011-09-27 17:06:12.143090000 -0700
@@ -28,6 +28,6 @@ DRIVER_PIPES = \
 	$(TOP)/src/gallium/drivers/rbug/librbug.a
 
 DRIVER_LINKS = \
-	$(shell pkg-config --libs libdrm libkms)
+	$(shell $(PKG_CONFIG) --libs libdrm libkms)
 
 include ../Makefile.xorg
diff -upr Mesa-7.11-p1/src/gallium/targets/xvmc-nouveau/Makefile Mesa-7.11-p1.pkg/src/gallium/targets/xvmc-nouveau/Makefile
--- Mesa-7.11-p1/src/gallium/targets/xvmc-nouveau/Makefile	2011-09-27 16:10:44.000000000 -0700
+++ Mesa-7.11-p1.pkg/src/gallium/targets/xvmc-nouveau/Makefile	2011-09-27 17:12:04.044660000 -0700
@@ -19,7 +19,7 @@ C_SOURCES = \
 	$(COMMON_GALLIUM_SOURCES) \
 	$(DRIVER_SOURCES)
 
-DRIVER_LIBS = $(shell pkg-config libdrm_nouveau --libs) -lXfixes
+DRIVER_LIBS = $(shell $(PKG_CONFIG) libdrm_nouveau --libs) -lXfixes
 
 include ../Makefile.xvmc
 
diff -upr Mesa-7.11-p1/src/gallium/targets/xvmc-r300/Makefile Mesa-7.11-p1.pkg/src/gallium/targets/xvmc-r300/Makefile
--- Mesa-7.11-p1/src/gallium/targets/xvmc-r300/Makefile	2011-09-27 16:10:44.000000000 -0700
+++ Mesa-7.11-p1.pkg/src/gallium/targets/xvmc-r300/Makefile	2011-09-27 17:12:10.563101000 -0700
@@ -15,7 +15,7 @@ C_SOURCES = \
 	$(COMMON_GALLIUM_SOURCES) \
 	$(DRIVER_SOURCES)
 
-DRIVER_LIBS = $(shell pkg-config libdrm --libs) -lXfixes
+DRIVER_LIBS = $(shell $(PKG_CONFIG) libdrm --libs) -lXfixes
 
 include ../Makefile.xvmc
 
diff -upr Mesa-7.11-p1/src/gallium/targets/xvmc-r600/Makefile Mesa-7.11-p1.pkg/src/gallium/targets/xvmc-r600/Makefile
--- Mesa-7.11-p1/src/gallium/targets/xvmc-r600/Makefile	2011-09-27 16:10:44.000000000 -0700
+++ Mesa-7.11-p1.pkg/src/gallium/targets/xvmc-r600/Makefile	2011-09-27 17:12:15.572690000 -0700
@@ -16,7 +16,7 @@ C_SOURCES = \
 	$(COMMON_GALLIUM_SOURCES) \
 	$(DRIVER_SOURCES)
 
-DRIVER_LIBS = $(shell pkg-config libdrm --libs) -lXfixes
+DRIVER_LIBS = $(shell $(PKG_CONFIG) libdrm --libs) -lXfixes
 
 include ../Makefile.xvmc
 
diff -upr Mesa-7.11-p1/src/gallium/winsys/g3dvl/dri/Makefile Mesa-7.11-p1.pkg/src/gallium/winsys/g3dvl/dri/Makefile
--- Mesa-7.11-p1/src/gallium/winsys/g3dvl/dri/Makefile	2011-09-27 16:10:44.000000000 -0700
+++ Mesa-7.11-p1.pkg/src/gallium/winsys/g3dvl/dri/Makefile	2011-09-27 17:12:22.002210000 -0700
@@ -4,7 +4,7 @@ include $(TOP)/configs/current
 LIBNAME = vldri
 
 LIBRARY_INCLUDES = -I$(TOP)/src/gallium/winsys/g3dvl \
-                   $(shell pkg-config libdrm --cflags-only-I)
+                   $(shell $(PKG_CONFIG) libdrm --cflags-only-I)
 
 C_SOURCES = \
 	driclient.c \
diff -upr Mesa-7.11-p1/src/gallium/winsys/i915/drm/Makefile Mesa-7.11-p1.pkg/src/gallium/winsys/i915/drm/Makefile
--- Mesa-7.11-p1/src/gallium/winsys/i915/drm/Makefile	2011-09-27 16:10:44.000000000 -0700
+++ Mesa-7.11-p1.pkg/src/gallium/winsys/i915/drm/Makefile	2011-09-27 17:06:12.157097000 -0700
@@ -6,8 +6,8 @@ LIBNAME = i915drm
 # get C_SOURCES
 include Makefile.sources
 
-LIBRARY_INCLUDES = $(shell pkg-config libdrm --cflags-only-I)
+LIBRARY_INCLUDES = $(shell $(PKG_CONFIG) libdrm --cflags-only-I)
 
-LIBRARY_DEFINES = $(shell pkg-config libdrm --cflags-only-other)
+LIBRARY_DEFINES = $(shell $(PKG_CONFIG) libdrm --cflags-only-other)
 
 include ../../../Makefile.template
diff -upr Mesa-7.11-p1/src/gallium/winsys/i965/drm/Makefile Mesa-7.11-p1.pkg/src/gallium/winsys/i965/drm/Makefile
--- Mesa-7.11-p1/src/gallium/winsys/i965/drm/Makefile	2011-09-27 16:10:44.000000000 -0700
+++ Mesa-7.11-p1.pkg/src/gallium/winsys/i965/drm/Makefile	2011-09-27 17:06:12.161097000 -0700
@@ -7,8 +7,8 @@ C_SOURCES = \
 	i965_drm_buffer.c \
 	i965_drm_winsys.c
 
-LIBRARY_INCLUDES = $(shell pkg-config libdrm --cflags-only-I)
+LIBRARY_INCLUDES = $(shell $(PKG_CONFIG) libdrm --cflags-only-I)
 
-LIBRARY_DEFINES = $(shell pkg-config libdrm --cflags-only-other)
+LIBRARY_DEFINES = $(shell $(PKG_CONFIG) --cflags-only-other)
 
 include ../../../Makefile.template
diff -upr Mesa-7.11-p1/src/gallium/winsys/nouveau/drm/Makefile Mesa-7.11-p1.pkg/src/gallium/winsys/nouveau/drm/Makefile
--- Mesa-7.11-p1/src/gallium/winsys/nouveau/drm/Makefile	2011-09-27 16:10:44.000000000 -0700
+++ Mesa-7.11-p1.pkg/src/gallium/winsys/nouveau/drm/Makefile	2011-09-27 17:06:12.166091000 -0700
@@ -5,7 +5,7 @@ LIBNAME = nouveaudrm
 
 C_SOURCES = nouveau_drm_winsys.c
 
-LIBRARY_INCLUDES = $(shell pkg-config libdrm libdrm_nouveau --cflags-only-I)
-LIBRARY_DEFINES = $(shell pkg-config libdrm libdrm_nouveau --cflags-only-other)
+LIBRARY_INCLUDES = $(shell $(PKG_CONFIG) libdrm libdrm_nouveau --cflags-only-I)
+LIBRARY_DEFINES = $(shell $(PKG_CONFIG) libdrm libdrm_nouveau --cflags-only-other)
 
 include ../../../Makefile.template
diff -upr Mesa-7.11-p1/src/gallium/winsys/radeon/drm/Makefile Mesa-7.11-p1.pkg/src/gallium/winsys/radeon/drm/Makefile
--- Mesa-7.11-p1/src/gallium/winsys/radeon/drm/Makefile	2011-09-27 16:10:45.000000000 -0700
+++ Mesa-7.11-p1.pkg/src/gallium/winsys/radeon/drm/Makefile	2011-09-27 17:06:12.185090000 -0700
@@ -8,7 +8,7 @@ LIBNAME = radeonwinsys
 include Makefile.sources
 
 LIBRARY_INCLUDES = -I$(TOP)/include \
-		   $(shell pkg-config libdrm --cflags-only-I)
+		   $(shell $(PKG_CONFIG) libdrm --cflags-only-I)
 
 include ../../../Makefile.template
 
diff -upr Mesa-7.11-p1/src/gallium/winsys/svga/drm/Makefile Mesa-7.11-p1.pkg/src/gallium/winsys/svga/drm/Makefile
--- Mesa-7.11-p1/src/gallium/winsys/svga/drm/Makefile	2011-09-27 16:10:45.000000000 -0700
+++ Mesa-7.11-p1.pkg/src/gallium/winsys/svga/drm/Makefile	2011-09-27 17:06:12.190092000 -0700
@@ -9,11 +9,11 @@ include Makefile.sources
 LIBRARY_INCLUDES = \
        -I$(TOP)/src/gallium/drivers/svga \
        -I$(TOP)/src/gallium/drivers/svga/include \
-       $(shell pkg-config libdrm --cflags-only-I)
+       $(shell $(PKG_CONFIG) libdrm --cflags-only-I)
 
 LIBRARY_DEFINES = \
        -std=gnu99 -fvisibility=hidden \
        -DHAVE_STDINT_H -D_FILE_OFFSET_BITS=64 \
-       $(shell pkg-config libdrm --cflags-only-other)
+       $(shell $(PKG_CONFIG) libdrm --cflags-only-other)
 
 include ../../../Makefile.template
diff -upr Mesa-7.11-p1/src/mesa/drivers/dri/i915/Makefile Mesa-7.11-p1.pkg/src/mesa/drivers/dri/i915/Makefile
--- Mesa-7.11-p1/src/mesa/drivers/dri/i915/Makefile	2011-09-27 16:10:48.000000000 -0700
+++ Mesa-7.11-p1.pkg/src/mesa/drivers/dri/i915/Makefile	2011-09-27 17:09:05.412078000 -0700
@@ -16,7 +16,7 @@ C_SOURCES = \
 ASM_SOURCES = 
 
 DRIVER_DEFINES = $(addprefix -I$(TOP)/, $(i915_INCLUDES)) -DI915 \
-	$(shell pkg-config libdrm --atleast-version=2.3.1 \
+	$(shell $(PKG_CONFIG) libdrm --atleast-version=2.3.1 \
 				&& echo "-DDRM_VBLANK_FLIP=DRM_VBLANK_FLIP")
 
 INCLUDES += $(INTEL_CFLAGS)
