From 1eb7a81f2e43842acd59929ce65db2142b69134d Mon Sep 17 00:00:00 2001
From: Kenneth Graunke <kenneth@whitecape.org>
Date: Thu, 25 Nov 2010 05:33:07 +0000
Subject: glsl: Add a virtual as_discard() method.

NOTE: This is candidate for the 7.9 branch.
---
diff --git a/src/glsl/ir.h b/src/glsl/ir.h
index 850033b..62bb0fd 100644
--- a/src/glsl/ir.h
+++ b/src/glsl/ir.h
@@ -122,6 +122,7 @@ public:
    virtual class ir_if *                as_if()               { return NULL; }
    virtual class ir_swizzle *           as_swizzle()          { return NULL; }
    virtual class ir_constant *          as_constant()         { return NULL; }
+   virtual class ir_discard *           as_discard()          { return NULL; }
    /*@}*/
 
 protected:
@@ -1122,6 +1123,11 @@ public:
 
    virtual ir_visitor_status accept(ir_hierarchical_visitor *);
 
+   virtual ir_discard *as_discard()
+   {
+      return this;
+   }
+
    ir_rvalue *condition;
 };
 /*@}*/
--
cgit v0.8.3-6-g21f6
