diff --git a/Makefile.tools b/Makefile.tools
index 4d45682..9c13d45 100644
--- a/Makefile.tools
+++ b/Makefile.tools
@@ -149,10 +149,17 @@ else
 EXTRA_DIST += tools/hid2hci.1
 endif
 
+noinst_PROGRAMS += tools/btmgmt
+
+tools_btmgmt_SOURCES = tools/btmgmt.c src/glib-helper.c src/eir.c \
+				src/shared/util.h src/shared/util.c \
+				src/shared/mgmt.h src/shared/mgmt.c
+tools_btmgmt_LDADD = lib/libbluetooth-private.la @GLIB_LIBS@
+
 if EXPERIMENTAL
 noinst_PROGRAMS += tools/bdaddr tools/avinfo tools/avtest tools/scotest \
 			tools/hcieventmask tools/hcisecfilter tools/hwdb \
-			tools/btmgmt tools/btattach tools/btsnoop \
+			tools/btattach tools/btsnoop \
 			tools/btiotest tools/mpris-player
 
 tools_bdaddr_SOURCES = tools/bdaddr.c src/oui.h src/oui.c
@@ -168,11 +175,6 @@ tools_hcieventmask_LDADD = lib/libbluetooth-private.la
 
 tools_hwdb_LDADD = lib/libbluetooth-private.la
 
-tools_btmgmt_SOURCES = tools/btmgmt.c src/glib-helper.c src/eir.c \
-				src/shared/util.h src/shared/util.c \
-				src/shared/mgmt.h src/shared/mgmt.c
-tools_btmgmt_LDADD = lib/libbluetooth-private.la @GLIB_LIBS@
-
 tools_btiotest_SOURCES = tools/btiotest.c btio/btio.h btio/btio.c
 tools_btiotest_LDADD = lib/libbluetooth-private.la @GLIB_LIBS@
 
