diff --git a/Makefile.tools b/Makefile.tools
index 7ffdbc6..3ec1ad9 100644
--- a/Makefile.tools
+++ b/Makefile.tools
@@ -257,11 +257,21 @@ else
 EXTRA_DIST += tools/hid2hci.1
 endif
 
+noinst_PROGRAMS += tools/btmgmt
+
+tools_btmgmt_SOURCES = tools/btmgmt.c src/uuid-helper.c \
+				monitor/mainloop.h monitor/mainloop.c \
+				src/shared/io.h src/shared/io-mainloop.c \
+				src/shared/queue.h src/shared/queue.c \
+				src/shared/util.h src/shared/util.c \
+				src/shared/mgmt.h src/shared/mgmt.c
+tools_btmgmt_LDADD = lib/libbluetooth-internal.la
+
 if EXPERIMENTAL
 noinst_PROGRAMS += tools/bdaddr tools/avinfo tools/avtest \
 			tools/scotest tools/amptest tools/hwdb \
 			tools/hcieventmask tools/hcisecfilter \
-			tools/btmgmt tools/btinfo tools/btattach \
+			tools/btinfo tools/btattach \
 			tools/btsnoop tools/btproxy tools/btiotest \
 			tools/mpris-player tools/cltest tools/seq2bseq \
 			tools/ibeacon
@@ -281,14 +291,6 @@ tools_hwdb_LDADD = lib/libbluetooth-internal.la
 
 tools_hcieventmask_LDADD = lib/libbluetooth-internal.la
 
-tools_btmgmt_SOURCES = tools/btmgmt.c src/uuid-helper.c \
-				monitor/mainloop.h monitor/mainloop.c \
-				src/shared/io.h src/shared/io-mainloop.c \
-				src/shared/queue.h src/shared/queue.c \
-				src/shared/util.h src/shared/util.c \
-				src/shared/mgmt.h src/shared/mgmt.c
-tools_btmgmt_LDADD = lib/libbluetooth-internal.la
-
 tools_btinfo_SOURCES = tools/btinfo.c monitor/bt.h \
 				monitor/mainloop.h monitor/mainloop.c \
 				src/shared/io.h src/shared/io-mainloop.c \
