diff --git a/scripts/bluetooth.rules.in b/scripts/bluetooth.rules.in
index 64df69d..e0945c8 100644
--- a/scripts/bluetooth.rules.in
+++ b/scripts/bluetooth.rules.in
@@ -1,4 +1,4 @@
 # Run helper every time a Bluetooth device appears
 # On remove actions, bluetoothd should go away by itself
-ACTION=="add", SUBSYSTEM=="bluetooth", RUN+="@prefix@/sbin/bluetoothd --udev"
-ACTION=="change", SUBSYSTEM=="bluetooth", RUN+="@prefix@/sbin/bluetoothd --udev"
+ACTION=="add", SUBSYSTEM=="bluetooth", RUN+="/sbin/minijail0 -u bluetooth -g bluetooth -c 3500 -p -- @prefix@/sbin/bluetoothd --udev"
+ACTION=="change", SUBSYSTEM=="bluetooth", RUN+="/sbin/minijail0 -u bluetooth -g bluetooth -c 3500 -p -- @prefix@/sbin/bluetoothd --udev"
