diff --git a/lib/sdp.c b/lib/sdp.c
index cc685b4..0a10c6b 100644
--- a/lib/sdp.c
+++ b/lib/sdp.c
@@ -4247,7 +4247,7 @@ int sdp_process(sdp_session_t *session)
		SDPDBG("Attrlist byte count : %d\n", rsp_count);

		/* Valid range for rsp_count is 0x0002-0xFFFF */
-		if (rsp_count < 0x0002) {
+		if (t->rsp_concat_buf.data_size == 0 && rsp_count < 0x0002) {
			t->err = EPROTO;
			SDPERR("Protocol error: invalid AttrList size");
			status = SDP_INVALID_PDU_SIZE;
