From 2a16b506420710b0b745fbf02ccf40be1b58b33e Mon Sep 17 00:00:00 2001
From: Arman Uguray <armansito@chromium.org>
Date: Tue, 28 Oct 2014 14:07:34 -0700
Subject: [PATCH] Make btmgmt and btgatt-client non-experimental.

---
 Makefile.tools | 18 ++++++++++--------
 1 file changed, 10 insertions(+), 8 deletions(-)

diff --git a/Makefile.tools b/Makefile.tools
index 42cccc6..b9ed9db 100644
--- a/Makefile.tools
+++ b/Makefile.tools
@@ -218,11 +218,20 @@ else
 EXTRA_DIST += tools/hid2hci.1
 endif
 
+noinst_PROGRAMS += tools/btmgmt tools/btgatt-client
+
+tools_btmgmt_SOURCES = tools/btmgmt.c src/uuid-helper.c
+tools_btmgmt_LDADD = lib/libbluetooth-internal.la src/libshared-mainloop.la
+
+tools_btgatt_client_SOURCES = tools/btgatt-client.c src/uuid-helper.c
+tools_btgatt_client_LDADD = lib/libbluetooth-internal.la \
+				src/libshared-mainloop.la
+
 if EXPERIMENTAL
 noinst_PROGRAMS += tools/bdaddr tools/avinfo tools/avtest \
 			tools/scotest tools/amptest tools/hwdb \
 			tools/hcieventmask tools/hcisecfilter \
-			tools/btmgmt tools/btinfo tools/btattach \
+			tools/btinfo tools/btattach \
 			tools/btsnoop tools/btproxy tools/btiotest \
 			tools/cltest tools/seq2bseq tools/hex2hcd \
 			tools/ibeacon tools/btgatt-client
@@ -242,9 +251,6 @@ tools_hwdb_LDADD = lib/libbluetooth-internal.la
 
 tools_hcieventmask_LDADD = lib/libbluetooth-internal.la
 
-tools_btmgmt_SOURCES = tools/btmgmt.c src/uuid-helper.c
-tools_btmgmt_LDADD = lib/libbluetooth-internal.la src/libshared-mainloop.la
-
 tools_btinfo_SOURCES = tools/btinfo.c monitor/bt.h
 tools_btinfo_LDADD = src/libshared-mainloop.la
 
@@ -270,10 +276,6 @@ tools_hex2hcd_SOURCES = tools/hex2hcd.c
 tools_ibeacon_SOURCES = tools/ibeacon.c monitor/bt.h
 tools_ibeacon_LDADD = src/libshared-mainloop.la
 
-tools_btgatt_client_SOURCES = tools/btgatt-client.c src/uuid-helper.c
-tools_btgatt_client_LDADD = lib/libbluetooth-internal.la \
-				src/libshared-mainloop.la
-
 EXTRA_DIST += tools/bdaddr.1
 endif
 
-- 
2.1.0.rc2.206.gedb03e5

