--- opengl.eselect.orig	2011-02-10 15:23:53.000000000 -0800
+++ opengl.eselect	2011-02-10 15:52:12.000000000 -0800
@@ -85,7 +85,7 @@
 		# Note that we don't do .so*, just .so on purpose.  The
 		# loader knows to look in the profile dir, and the
 		# linked just needs the .so
-		for file in ${profile_libdir}/libGL{,core}.{so,dylib,a}; do
+		for file in ${profile_libdir}/libGL{,core}.{so*,dylib,a}; do
 			upgrade_file ${file} ${target}
 		done
 	popd &> /dev/null
