diff --git a/pym/portage/dbapi/vartree.py.orig b/pym/portage/dbapi/vartree.py
index 89ac917..7097a05 100644
--- a/pym/portage/dbapi/vartree.py.orig
+++ b/pym/portage/dbapi/vartree.py
@@ -1215,7 +1215,8 @@ class dblink(object):
 			raise AssertionError("Lock already held.")
 		# At least the parent needs to exist for the lock file.
 		ensure_dirs(self.dbroot)
-		self._lock_vdb = lockdir(self.dbroot)
+		if os.environ.get("PORTAGE_LOCKS") != "false":
+			self._lock_vdb = lockdir(self.dbroot)
 
 	def unlockdb(self):
 		if self._lock_vdb:
