From 0bb454d0aecba0be1dd197038eefc4be6be2580e Mon Sep 17 00:00:00 2001
From: Zdenek Behan <rain@matfyz.cz>
Date: Fri, 13 Apr 2012 14:51:46 -0700
Subject: [PATCH] portageq mass_best_visible: optional "type" arg

---
 bin/portageq |   22 ++++++++++++++++++----
 1 files changed, 18 insertions(+), 4 deletions(-)

diff --git a/bin/portageq b/bin/portageq
index fcdb9d9..dc465d2 100755
--- a/bin/portageq
+++ b/bin/portageq
@@ -473,6 +473,9 @@ def best_visible(argv):
 			if pkg.visible:
 				writemsg_stdout("%s\n" % (pkg.cpv,), noiselevel=-1)
 				return os.EX_OK
+
+		# No package found, write out an empty line.
+		writemsg_stdout("\n", noiselevel=-1)
 	except KeyError:
 		pass
 	return 1
@@ -480,16 +483,27 @@
 
 
 def mass_best_visible(argv):
-	"""<root> [<category/package>]+
+	"""<root> [<type>] [<category/package>]+
 	Returns category/package-version (without .ebuild).
+	The pkgtype argument defaults to "ebuild" if unspecified,
+	otherwise it must be one of ebuild, binary, or installed.
 	"""
+	type_map = {
+		"ebuild":"porttree",
+		"binary":"bintree",
+		"installed":"vartree"}
+
 	if (len(argv) < 2):
 		print("ERROR: insufficient parameters!")
 		sys.exit(2)
 	try:
-		for pack in argv[1:]:
-			mylist=portage.db[argv[0]]["porttree"].dbapi.match(pack)
-			print(pack+":"+portage.best(mylist))
+		root = argv.pop(0)
+		pkgtype = "ebuild"
+		if argv[0] in type_map:
+			pkgtype = argv.pop(0)
+		for pack in argv:
+			writemsg_stdout("%s:" % pack, noiselevel=-1)
+			best_visible([root, pkgtype, pack])
 	except KeyError:
 		sys.exit(1)
 mass_best_visible.uses_root = True
-- 
1.7.8.5

