diff -ur xkeyboard-config-2.4.1.orig/symbols/pc xkeyboard-config-2.4.1/symbols/pc
--- xkeyboard-config-2.4.1.orig/symbols/pc	2011-10-05 07:19:54.000000000 +0900
+++ xkeyboard-config-2.4.1/symbols/pc	2012-02-14 19:32:29.361799530 +0900
@@ -14,7 +14,16 @@
     include "pc(editing)"
     include "keypad(x11)"
 
-    key <BKSP> {	[ BackSpace	]	};
+    key <BKSP> {
+        type="CONTROL_ALT",
+        repeat=yes,
+        symbols[Group1] = [ BackSpace, Delete, Delete ],
+        actions[Group1] = [
+            NoAction(),
+            RedirectKey(key=<DELE>, clearmods=Alt),
+            RedirectKey(key=<DELE>, clearmods=Alt)
+        ]
+    };
 
     key  <TAB> {	[ Tab,	ISO_Left_Tab	]	};
     key <RTRN> {	[ Return	]	};
@@ -83,8 +92,27 @@
     key  <END> {	[  End			]	};
     key <PGDN> {	[  Next			]	};
 
-    key   <UP> {	[  Up			]	};
+    key <UP> {
+        type="CONTROL_ALT",
+        repeat=yes,
+        symbols[Group1] = [ Up, Prior, Home ],
+        actions[Group1] = [
+            NoAction(),
+            RedirectKey(key=<PGUP>, clearmods=Alt),
+            RedirectKey(key=<HOME>, clearmods=Control+Alt)
+        ]
+    };
+    key <DOWN> {
+        type="CONTROL_ALT",
+        repeat=yes,
+        symbols[Group1] = [ Down, Next, End ],
+        actions[Group1] = [
+            NoAction(),
+            RedirectKey(key=<PGDN>, clearmods=Alt),
+            RedirectKey(key=<END>, clearmods=Control+Alt)
+        ]
+    };
+
     key <LEFT> {	[  Left			]	};
-    key <DOWN> {	[  Down			]	};
     key <RGHT> {	[  Right		]	};
 };
diff -ur xkeyboard-config-2.4.1.orig/types/basic xkeyboard-config-2.4.1/types/basic
--- xkeyboard-config-2.4.1.orig/types/basic	2011-10-05 07:19:54.000000000 +0900
+++ xkeyboard-config-2.4.1/types/basic	2012-02-14 19:33:38.411499653 +0900
@@ -3,6 +3,7 @@
     // Fairly standard definitions for
     // the four required key types
 
+    virtual_modifiers Alt;
     virtual_modifiers NumLock;
 
     type "ONE_LEVEL" {
@@ -26,4 +27,16 @@
         level_name[Level2] = "Caps";
     };
 
+    // Extra type for keys that we want to simulate other keys when Alt or
+    // Control+Alt are down.  Level 1 is the key by itself, 2 is Alt, and 3
+    // is Ctrl+Alt.
+    type "CONTROL_ALT" {
+        modifiers              = Alt+Control;
+        map[None]              = Level1;
+        map[Alt]               = Level2;
+        map[Control+Alt]       = Level3;
+        level_name[Level1]     = "Base";
+        level_name[Level2]     = "Alt";
+        level_name[Level3]     = "ControlAlt";
+    };
 };
