diff -uNr xkb.orig/symbols/pc xkb/symbols/pc
--- xkb.orig/symbols/pc	2012-09-11 22:02:32.195462135 -0700
+++ xkb/symbols/pc	2012-09-20 00:37:59.349400128 -0700
@@ -40,6 +40,7 @@
     modifier_map Lock   { Caps_Lock, ISO_Lock };
     modifier_map Control{ Control_L, Control_R };
     modifier_map Mod2   { Num_Lock };
+    modifier_map Mod3   { F15 };
     modifier_map Mod4   { Super_L, Super_R };
 
     //  Fake keys for virtual<->real modifiers mapping 
