diff -ur xkeyboard-config-2.4.1.patched/symbols/br xkeyboard-config-2.4.1.patched.new/symbols/br
--- xkeyboard-config-2.4.1.patched/symbols/br	2011-10-05 07:19:54.000000000 +0900
+++ xkeyboard-config-2.4.1.patched.new/symbols/br	2013-10-04 18:14:11.805824883 +0900
@@ -21,7 +21,7 @@
 
     key <AD01>  { [         q,          Q,        slash,        slash ] };   
     key <AD02>  { [         w,          W,     question,     question ] };   
-    key <AD03>	{ [         e,          E,     EuroSign,     EuroSign ]	};
+    key <AD03>	{ [         e,          E,       degree,       degree ]	};
     key <AD04>	{ [         r,          R,   registered,   registered ]	};
     key <AD11>	{ [dead_acute, dead_grave,        acute,        grave ]	};
     key <AD12>	{ [bracketleft,  braceleft, ordfeminine,  dead_macron ]	};
