diff -ur xkeyboard-config-2.4.1.orig/symbols/tr xkeyboard-config-2.4.1/symbols/tr
--- xkeyboard-config-2.4.1.orig/symbols/tr	2015-10-14 16:06:01.295599228 +0800
+++ xkeyboard-config-2.4.1/symbols/tr	2015-10-15 15:52:17.715101262 +0800
@@ -25,7 +25,7 @@

     key <AD02>	{ [         w,          W,   VoidSymbol,   VoidSymbol ]	};
     key <AD03>	{ [         e,          E,     EuroSign,   VoidSymbol ]	};
-    key <AD05>	{ [         t,          T,    trademark,   VoidSymbol ]	};
+    key <AD05>	{ [         t,          T,        U20BA,   VoidSymbol ]	};
     key <AD07>	{ [         u,          U,  ucircumflex,  Ucircumflex ]	};
     key <AD08>	{ type[group1] = "FOUR_LEVEL_ALPHABETIC",
                   [  idotless,          I,  icircumflex,  Icircumflex ] };
@@ -98,7 +98,7 @@
     key <AC03>	{ [         e,          E,     EuroSign,   VoidSymbol ]	};
     key <AC04>	{ [         a,          A,  acircumflex,  Acircumflex ]	};
     key <AC05>	{ [udiaeresis, Udiaeresis,  ucircumflex,  Ucircumflex ]	};
-    key <AC06>	{ [         t,          T,    trademark,   VoidSymbol ]	};
+    key <AC06>	{ [         t,          T,        U20BA,   VoidSymbol ]	};
     key <AC07>	{ [         k,          K,   VoidSymbol,   VoidSymbol ]	};
     key <AC08>	{ [         m,          M,           mu,   VoidSymbol ]	};
     key <AC09>	{ [         l,          L,   VoidSymbol,   VoidSymbol ]	};
