diff -ur xkeyboard-config-2.4.1.orig/symbols/altwin xkeyboard-config-2.4.1/symbols/altwin
--- xkeyboard-config-2.4.1.orig/symbols/altwin	2011-10-05 07:19:54.000000000 +0900
+++ xkeyboard-config-2.4.1/symbols/altwin	2012-02-14 19:12:21.267032284 +0900
@@ -3,7 +3,7 @@
     key <LALT> {	[ 	Alt_L,	Meta_L		]	};
     key <RALT> { type[Group1] = "TWO_LEVEL",
                  symbols[Group1] = [ Alt_R, Meta_R ] };
-    modifier_map Mod1	{ Alt_L, Alt_R, Meta_L, Meta_R };
+    modifier_map Mod1	{ Alt_R, Meta_R };
 //  modifier_map Mod4	{};
 };
 
diff -ur xkeyboard-config-2.4.1.orig/symbols/pc xkeyboard-config-2.4.1/symbols/pc
--- xkeyboard-config-2.4.1.orig/symbols/pc	2011-10-05 07:19:54.000000000 +0900
+++ xkeyboard-config-2.4.1/symbols/pc	2012-02-14 19:16:41.025896801 +0900
@@ -40,7 +40,7 @@
     modifier_map Lock   { Caps_Lock, ISO_Lock };
     modifier_map Control{ Control_L, Control_R };
     modifier_map Mod2   { Num_Lock };
-    modifier_map Mod4   { Super_L, Super_R };
+    // modifier_map Mod4   { Super_L, Super_R };
 
     //  Fake keys for virtual<->real modifiers mapping 
     key <LVL3> {	[ ISO_Level3_Shift	]	};
