--- xkeyboard-config-2.15.orig/symbols/ca	2015-10-26 14:56:30.780984562 -0700
+++ xkeyboard-config-2.15/symbols/ca	2015-10-26 14:57:53.124666779 -0700
@@ -383,8 +383,8 @@
     			NoSymbol, dead_abovering ]	};
 
     key.type[group1]="EIGHT_LEVEL_SEMIALPHABETIC";
-    key <AD12>	{ [  ccedilla,   Ccedilla,     asciitilde, NoSymbol,
-		   dead_tilde,    dead_macron ]	};
+    key <AD12>	{ [  ccedilla,   Ccedilla,     dead_tilde, NoSymbol,
+		   asciitilde,    dead_macron ]	};
 
     key <AC01>	{ [         a,          A,     NoSymbol, NoSymbol,
     			   ae,         AE ]	};
@@ -467,6 +467,10 @@
 xkb_symbols "eng" {
     include "us(basic)"
     name[Group1] = "English (Canada)";
+
+    // Remap international-backslash key as backslash/bar instead of
+    // less/greater to meet key top indicator.
+    Key <LSGT> { [ backslash, bar ] };
 };
 
 // EXTRAS:
