diff -ur xkeyboard-config-2.4.1.orig/symbols/Makefile.am xkeyboard-config-2.4.1/symbols/Makefile.am
--- xkeyboard-config-2.4.1.orig/symbols/Makefile.am	2011-10-05 07:19:54.000000000 +0900
+++ xkeyboard-config-2.4.1/symbols/Makefile.am	2012-02-14 18:29:06.398469788 +0900
@@ -34,7 +34,7 @@
 ua us uz vn \
 za \
 altwin capslock compose ctrl empty eurosign rupeesign group inet \
-keypad kpdl level3 level5 nbsp olpc shift srvr_ctrl typo
+keypad kpdl level3 level5 nbsp olpc shift srvr_ctrl typo chromeos
 
 dir_data = $(dist_symbols_DATA)
 
