--- policy/org.freedesktop.devicekit.disks.policy.in.orig	2011-02-02 22:32:01.000000000 -0800
+++ policy/org.freedesktop.devicekit.disks.policy.in	2011-02-02 22:29:05.000000000 -0800
@@ -13,7 +13,7 @@
     <_description>Mount a device</_description>
     <_message>Authentication is required to mount the device</_message>
     <defaults>
-      <allow_any>no</allow_any>
+      <allow_any>yes</allow_any>
       <allow_inactive>no</allow_inactive>
       <allow_active>yes</allow_active>
     </defaults>
