diff -upr opencryptoki-2.2.8.orig/usr/lib/pkcs11/tpm_stdll/Makefile.am opencryptoki-2.2.8/usr/lib/pkcs11/tpm_stdll/Makefile.am
--- opencryptoki-2.2.8.orig/usr/lib/pkcs11/tpm_stdll/Makefile.am	2011-08-17 11:28:07.207947978 -0700
+++ opencryptoki-2.2.8/usr/lib/pkcs11/tpm_stdll/Makefile.am	2011-08-18 08:54:24.627863624 -0700
@@ -14,12 +14,13 @@ opencryptoki_stdll_libpkcs11_tpm_la_CFLA
 					     -I. -I../../../include	\
 					     -I../../../include/pkcs11	\
 					     -I../common -DMMAP		\
-					     -DSTDLL_NAME=\"tpmtok\"
+					     -DSTDLL_NAME=\"tpmtok\"    \
+					     -DNO_BACKUP
 
 # Not all versions of automake observe libname_CFLAGS
 AM_CFLAGS = -DLINUX -DMMAP -DSPINXPL -DNOCDMF -DNODSA -DNODH -I. \
 	    -I../../../include -I../../../include/pkcs11 \
-	    -I../common
+	    -I../common -DNO_BACKUP
 
 opencryptoki_stdll_libpkcs11_tpm_la_LDFLAGS = -shared -Wl,-Bsymbolic	\
 					      -lcrypto -ltspi -lpthread
diff -upr opencryptoki-2.2.8.orig/usr/lib/pkcs11/tpm_stdll/tpm_specific.c opencryptoki-2.2.8/usr/lib/pkcs11/tpm_stdll/tpm_specific.c
--- opencryptoki-2.2.8.orig/usr/lib/pkcs11/tpm_stdll/tpm_specific.c	2011-08-17 11:28:07.207947978 -0700
+++ opencryptoki-2.2.8/usr/lib/pkcs11/tpm_stdll/tpm_specific.c	2011-08-18 08:54:32.867474927 -0700
@@ -1155,11 +1155,13 @@ token_create_private_tree(CK_BYTE *pinHa
 		return rc;
 	}
 
+#ifndef NO_BACKUP
 	if (openssl_write_key(rsa, TPMTOK_PRIV_ROOT_KEY_FILE, pPin)) {
 		LogError1("openssl_write_key");
 		RSA_free(rsa);
 		return CKR_FUNCTION_FAILED;
 	}
+#endif
 
 	RSA_free(rsa);
 
@@ -1220,11 +1222,13 @@ token_create_public_tree(CK_BYTE *pinHas
 		return rc;
 	}
 
+#ifndef NO_BACKUP
 	if (openssl_write_key(rsa, TPMTOK_PUB_ROOT_KEY_FILE, pPin)) {
 		LogError1("openssl_write_key");
 		RSA_free(rsa);
 		return CKR_FUNCTION_FAILED;
 	}
+#endif
 
 	RSA_free(rsa);
 
