--- usr/lib/pkcs11/tpm_stdll/h_extern.h.orig	2011-05-18 14:54:16.000000000 -0700
+++ usr/lib/pkcs11/tpm_stdll/h_extern.h	2011-05-18 14:54:39.000000000 -0700
@@ -2016,14 +2016,23 @@
 #define LogDebug(...)		do { } while (0)
 #define LogDebug1(...)		do { } while (0)
 #define LogBlob(...)		do { } while (0)
-#define LogError(...)		do { } while (0)
-#define LogError1(...)		do { } while (0)
 #define LogWarn(...)		do { } while (0)
 #define LogWarn1(...)		do { } while (0)
 #define LogInfo(...)		do { } while (0)
 #define LogInfo1(...)		do { } while (0)
 
+#ifdef HIDE_BAD_ERRORS
+#define LogError(...)		do { } while (0)
+#define LogError1(...)		do { } while (0)
 #define st_err_log(...)		do { } while (0)
+#else
+#include <syslog.h>
+#define LogError(fmt, ...) syslog(LOG_ERR, "libpkcs11_tpm: " fmt, ##__VA_ARGS__)
+#define LogError1(data) \
+  syslog(LOG_ERR, "libpkcs11_tpm: %s:%d: %s", __FILE__, __LINE__, data)
+#define st_err_log(...) \
+  syslog(LOG_ERR, "libpkcs11_tpm: st_err_log at %s:%d", __FILE__, __LINE__)
+#endif
 #endif
 
 /* custom attributes for the TPM token */
