diff -upr opencryptoki-2.2.8.orig/usr/lib/pkcs11/tpm_stdll/tpm_specific.c opencryptoki-2.2.8/usr/lib/pkcs11/tpm_stdll/tpm_specific.c
--- opencryptoki-2.2.8.orig/usr/lib/pkcs11/tpm_stdll/tpm_specific.c	2011-03-04 14:20:08.147000363 -0800
+++ opencryptoki-2.2.8/usr/lib/pkcs11/tpm_stdll/tpm_specific.c	2011-03-04 14:19:35.246934169 -0800
@@ -2375,6 +2375,7 @@ token_specific_rsa_generate_keypair( TEM
 	CK_ULONG	mod_bits = 0;
 	CK_BBOOL	flag;
 	CK_RV		rc;
+	CK_BYTE		tpm_pubexp[] = { 0, 1, 0, 1 };
 
 	TSS_FLAG	initFlags = 0;
 	BYTE		authHash[SHA1_HASH_SIZE];
@@ -2491,6 +2492,14 @@ token_specific_rsa_generate_keypair( TEM
 	template_update_attribute( priv_tmpl, attr );
 	Tspi_Context_FreeMemory(tspContext, rgbBlob);
 
+	/* put the public exponent into the private key object */
+	if ((rc = build_attribute(CKA_PUBLIC_EXPONENT, tpm_pubexp,
+				  sizeof(tpm_pubexp), &attr))) {
+		st_err_log(84, __FILE__, __LINE__);
+		return rc;
+	}
+	template_update_attribute( priv_tmpl, attr );
+
 	/* wrap the authdata and put it into an object */
 	if (authData != NULL) {
 		if ((rc = token_wrap_auth_data(authData, publ_tmpl, priv_tmpl))) {
