--- llvm-2.8-r6/work/llvm-2.8/lib/ExecutionEngine/JIT/Intercept.cpp.orig	2012-04-19 15:05:56.833693203 -0700
+++ llvm-2.8-r6/work/llvm-2.8/lib/ExecutionEngine/JIT/Intercept.cpp	2012-04-19 15:06:19.834006058 -0700
@@ -19,6 +19,10 @@
 #include "llvm/Support/ErrorHandling.h"
 #include "llvm/System/DynamicLibrary.h"
 #include "llvm/Config/config.h"
+
+#include <sys/types.h>
+#include <unistd.h>
+
 using namespace llvm;
 
 // AtExitHandlers - List of functions to call when the program exits,
