https://breakpad.appspot.com/526002

--- a/src/third_party/curl/curlbuild.h
+++ b/src/third_party/curl/curlbuild.h
@@ -154,7 +154,7 @@
 #endif
 
 /* The size of `long', as computed by sizeof. */
-#if defined(_M_X64) || defined(__x86_64__)
+#if defined(_M_X64) || (defined(__x86_64__) && !defined(__ILP32__))
 #define CURL_SIZEOF_LONG 8
 #else
 #define CURL_SIZEOF_LONG 4
@@ -170,7 +170,7 @@
 typedef CURL_TYPEOF_CURL_SOCKLEN_T curl_socklen_t;
 
 /* Signed integral data type used for curl_off_t. */
-#if defined(_M_X64) || defined(__x86_64__)
+#if defined(_M_X64) || (defined(__x86_64__) && !defined(__ILP32__))
 #define CURL_TYPEOF_CURL_OFF_T long
 #else
 #define CURL_TYPEOF_CURL_OFF_T int64_t
