--- src/common/linux/http_upload.cc.orig	2010-10-07 17:23:10.000000000 -0700
+++ src/common/linux/http_upload.cc	2010-10-07 17:21:47.000000000 -0700
@@ -141,6 +141,9 @@
                      reinterpret_cast<void *>(response_body));
   }
 
+  // Fail if 400+ is returned from the web server.
+  (*curl_easy_setopt)(curl, CURLOPT_FAILONERROR, 1);
+
   CURLcode (*curl_easy_perform)(CURL *);
   *(void**) (&curl_easy_perform) = dlsym(curl_lib, "curl_easy_perform");
   err_code = (*curl_easy_perform)(curl);
