diff -urN ../mozc-1.1.785.102.old/hangul/session.cc ./hangul/session.cc
--- ../mozc-1.1.785.102.old/hangul/session.cc	2011-08-04 20:52:24.864402973 +0900
+++ ./hangul/session.cc	2011-08-04 21:00:15.114343579 +0900
@@ -350,6 +350,7 @@
 
   if (current_mode_ == kHangulMode) {
     if (IsHanjaSelectionMode()) {
+      CancelHanjaSelection(output);
       return false;
     }
     if (::hangul_ic_is_empty(context_)) {
@@ -463,7 +464,6 @@
 
 void Session::ProcessHanjaSelection(commands::Command *command) {
   const KeyEvent &key_event = command->input().key();
-  commands::Output *output = command->mutable_output();
 
   // When hanja selection window is shown, user can only do selecting hanja or
   // canceling conversion. After commit hanja character does not anything.
@@ -508,12 +508,8 @@
         return;
         break;
       case KeyEvent::HANJA:
-      case KeyEvent::ESCAPE: {
-          string preedit;
-          ResetHanjaList();
-          GetPreeditString(&preedit);
-          SetPreeditToOutput(preedit, output);
-        }
+      case KeyEvent::ESCAPE:
+        CancelHanjaSelection(command->mutable_output());
         return;
         break;
       default:
@@ -588,12 +584,24 @@
 }
 
 void Session::CancelContext(commands::Output *output) {
+  DCHECK(output);
   string preedit;
   FlushPreedit(&preedit);
   SetResultToOutput(preedit, output);
   ResetHanjaList();
 }
 
+void Session::CancelHanjaSelection(commands::Output *output) {
+  DCHECK(output);
+  if (!IsHanjaSelectionMode()) {
+    return;
+  }
+  string preedit;
+  GetPreeditString(&preedit);
+  SetPreeditToOutput(preedit, output);
+  ResetHanjaList();
+}
+
 void Session::UpdatePreeditWithHanjaString(const string &hanja,
                                            commands::Command *command) {
   if (current_mode_ == kHangulMode) {
@@ -639,7 +647,6 @@
 
   if (!HasReproduciblePreedit()) {
     CancelContext(output);
-    ResetHanjaList();
     return true;
   }
 
diff -urN ../mozc-1.1.785.102.old/hangul/session.h ./hangul/session.h
--- ../mozc-1.1.785.102.old/hangul/session.h	2011-08-04 12:34:17.824810844 +0900
+++ ./hangul/session.h	2011-08-04 20:57:25.034186084 +0900
@@ -145,6 +145,9 @@
   // as a result string. Then clear current hanja candidates if exists.
   void CancelContext(commands::Output *output);
 
+  // Cancels hanja-selection.
+  void CancelHanjaSelection(commands::Output *output);
+
   // Looks up hanja list from hanja list with preedit string.
   bool HanjaLookup(commands::Command *command);
 
