*** mozc-1.1.785.102/hangul/session.cc	Tue Jul 19 16:33:43 2011
--- mozc-1.1.785.102/hangul/session.cc	Tue Jul 26 10:42:58 2011
***************
*** 655,660 ****
--- 655,673 ----
  
  
  bool Session::SendKey(commands::Command *command) {
+   // This is quick fix for ChromeOS/R14
+   // TODO(nona): Read hanja key map from ChromeOS.
+   commands::KeyEvent *key_event = command->mutable_input()->mutable_key();
+   if (key_event->has_key_code() && key_event->modifier_keys_size() == 1) {
+     if (key_event->modifier_keys(0) == commands::KeyEvent::CTRL &&
+         key_event->key_code() == '9') {
+       key_event->clear_key_code();
+       key_event->clear_modifiers();
+       key_event->clear_modifier_keys();
+       key_event->set_special_key(commands::KeyEvent::HANJA);
+     }
+   }
+ 
    if (current_mode_ == kHanjaLockMode) {
      command->mutable_output()->set_consumed(ProcessHanjaLockMode(command));
    } else {
