*** mozc-1.1.785.102/hangul/session.cc	Tue Jul 26 10:44:15 2011
--- mozc-1.1.785.102/hangul/session.cc	Tue Jul 26 10:44:59 2011
***************
*** 312,323 ****
      new_candidate->set_id(i);
      new_candidate->set_index(i);
      new_candidate->set_value(::hanja_get_value(hanja));
-     // TODO(nona): Description is too long to pack one line.
-     // Find better way to show like ibus-hangul.
-     // new_candidate->mutable_annotation()->set_description(
-     //    hanja_get_comment(hanja));
    }
  
    const string kDigits = "1234567890";
    session::SessionOutput::FillShortcuts(kDigits, candidates);
  
--- 312,325 ----
      new_candidate->set_id(i);
      new_candidate->set_index(i);
      new_candidate->set_value(::hanja_get_value(hanja));
    }
  
+   const char *comment = hanja_list_get_nth_comment(hanja_list_, hanja_index_);
+   if (comment) {
+     commands::Footer *footer = candidates->mutable_footer();
+     footer->set_label(comment);
+   }
+ 
    const string kDigits = "1234567890";
    session::SessionOutput::FillShortcuts(kDigits, candidates);
  
