diff -urN ../mozc-1.1.785.102.old/hangul/session.cc ./hangul/session.cc
--- ../mozc-1.1.785.102.old/hangul/session.cc	2011-08-04 22:16:03.214388328 +0900
+++ ./hangul/session.cc	2011-08-05 00:10:27.234234963 +0900
@@ -731,17 +731,14 @@
       consumed = true;
       break;
     case commands::SessionCommand::SUBMIT:
-      FlushPreedit(&hangul_string);
-      SetResultToOutput(hangul_string, command->mutable_output());
+      CancelContext(command->mutable_output());
       break;
     case commands::SessionCommand::SELECT_CANDIDATE:
       SelectCandidateByShortcut(session_command.id());
       CommitSelectedCandidate(command);
       break;
     case commands::SessionCommand::SWITCH_INPUT_MODE:
-      FlushPreedit(&hangul_string);
-      SetResultToOutput(hangul_string, command->mutable_output());
-      ResetHanjaList();
+      CancelContext(command->mutable_output());
       switch (session_command.composition_mode()) {
         case commands::HIRAGANA:
           current_mode_ = kHangulMode;
@@ -770,6 +767,7 @@
 
 void Session::RenewContext() {
   ResetHanjaList();
+  hanja_lock_preedit_.clear();
   const HangulConfigMap *config_map = Singleton<HangulConfigMap>::get();
   const HangulConfig &hangul_config = GET_CONFIG(hangul_config);
   string keyboard_id =
diff -urN ../mozc-1.1.785.102.old/unix/ibus/mozc_engine.cc ./unix/ibus/mozc_engine.cc
--- ../mozc-1.1.785.102.old/unix/ibus/mozc_engine.cc	2011-08-04 12:34:17.824810844 +0900
+++ ./unix/ibus/mozc_engine.cc	2011-08-04 22:20:47.974233710 +0900
@@ -613,13 +613,14 @@
     // Commit a preedit string.
     command.set_type(commands::SessionCommand::SUBMIT);
     session_->SendCommand(command, &output);
-    UpdateAll(engine, output);
   } else {
     command.set_type(commands::SessionCommand::SWITCH_INPUT_MODE);
     command.set_composition_mode(composition_mode);
     session_->SendCommand(command, &output);
+    output.clear_mode();
   }
   current_composition_mode_ = composition_mode;
+  UpdateAll(engine, output);
 }
 
 void MozcEngine::PropertyActivate(IBusEngine *engine,
