diff -crN mozc-1.1.785.102.old/hangul/session.cc mozc-1.1.785.102/hangul/session.cc
*** mozc-1.1.785.102.old/hangul/session.cc	Fri Jul 29 16:59:11 2011
--- mozc-1.1.785.102/hangul/session.cc	Fri Jul 29 16:59:57 2011
***************
*** 476,481 ****
--- 476,482 ----
      } else {
        UpdateCandidate(DO_NOT_CHANGE_SELECTED_CANDIDATE, command);
      }
+     return;
    }
  
    if (key_event.has_special_key()) {
