diff -urN ../mozc-1.1.785.102.old/unix/ibus/mozc_engine.cc ./unix/ibus/mozc_engine.cc
--- ../mozc-1.1.785.102.old/unix/ibus/mozc_engine.cc	2011-08-12 11:25:58.638497767 +0900
+++ ./unix/ibus/mozc_engine.cc	2011-08-16 12:03:39.508596083 +0900
@@ -478,7 +478,8 @@
   // Stop ignoring "reset" signal.  See ProcessKeyevent().
   ignore_reset_for_deletion_range_workaround_ = false;
 
-  RevertSession(engine);
+  // Following change is for crosbug.com/18555
+  SubmitSession(engine);
   SyncData(false);
 }
 
@@ -886,7 +887,9 @@
     return true;
   }
   IBusText *text = ComposePreeditText(output.preedit());
-  ibus_engine_update_preedit_text(engine, text, CursorPos(output), TRUE);
+  // Following change is related to crosbug.com/18555
+  ibus_engine_update_preedit_text_with_mode(engine, text, CursorPos(output),
+                                            TRUE, IBUS_ENGINE_PREEDIT_COMMIT);
   // |text| is released by ibus_engine_update_preedit_text.
   return true;
 }
@@ -1105,5 +1108,16 @@
   }
 }
 
+void MozcEngine::SubmitSession(IBusEngine *engine) {
+  commands::SessionCommand command;
+  command.set_type(commands::SessionCommand::SUBMIT);
+  commands::Output output;
+  if (!session_->SendCommand(command, &output)) {
+    LOG(ERROR) << "SubmitSession() failed";
+    return;
+  }
+  UpdateAll(engine, output);
+}
+
 }  // namespace ibus
 }  // namespace mozc
diff -urN ../mozc-1.1.785.102.old/unix/ibus/mozc_engine.h ./unix/ibus/mozc_engine.h
--- ../mozc-1.1.785.102.old/unix/ibus/mozc_engine.h	2011-08-12 11:25:58.638497767 +0900
+++ ./unix/ibus/mozc_engine.h	2011-08-16 11:57:14.248440339 +0900
@@ -172,6 +172,8 @@
   // message, then hides a preedit string and the candidate window.
   void RevertSession(IBusEngine *engine);
 
+  void SubmitSession(IBusEngine *engine);
+
   uint64 last_sync_time_;
   scoped_ptr<KeyTranslator> key_translator_;
   scoped_ptr<client::SessionInterface> session_;
