diff -urN ../mozc-1.1.785.102.old/hangul/session.cc ./hangul/session.cc
--- ../mozc-1.1.785.102.old/hangul/session.cc	2011-08-12 11:25:58.638497767 +0900
+++ ./hangul/session.cc	2011-08-13 02:32:37.178641948 +0900
@@ -350,8 +350,9 @@
 
   if (current_mode_ == kHangulMode) {
     if (IsHanjaSelectionMode()) {
-      CancelHanjaSelection(output);
-      return false;
+      // Following changes are issued as crosbug.com/18419.
+      UpdateCandidate(DO_NOT_CHANGE_SELECTED_CANDIDATE, command);
+      return true;
     }
     if (::hangul_ic_is_empty(context_)) {
       return false;
@@ -517,7 +518,8 @@
     }
   }
 
-  if (current_mode_ == kHanjaLockMode) {
+  // Fixing for crosbug.com/19074
+  if (current_mode_ == kHanjaLockMode && key_event.modifier_keys_size() == 0) {
     return ProcessKeyEvent(command);
   } else {
     // Other keys are not handled when hanja selection window is shown in hangul
@@ -678,6 +680,11 @@
       key_event.special_key() == KeyEvent::BACKSPACE) {
     return ProcessBSKey(command);
   }
+
+  if (key_event.has_special_key() &&
+      key_event.special_key() == KeyEvent::HANJA) {
+    return HanjaLookup(command);
+  }
   return ProcessKeyEvent(command);
 }
 
