diff -ru ../mozc-1.1.785.102.orig/hangul/hangul.gyp ./hangul/hangul.gyp
--- ../mozc-1.1.785.102.orig/hangul/hangul.gyp	2011-07-19 16:33:43.000000000 +0900
+++ ./hangul/hangul.gyp	2011-07-22 17:20:15.717590810 +0900
@@ -68,14 +68,12 @@
       'target_name': 'hangul_session_server',
       'type': 'static_library',
       'sources': [
-        '../converter/segments.cc',
         'hangul_session_factory.cc',
       ],
       'dependencies': [
         '../base/base.gyp:base',
         '../client/client.gyp:client',
         '../session/session.gyp:session_handler',
-        '../session/session.gyp:session_internal',
         '../session/session_base.gyp:genproto_session',
       ],
     },
diff -ru ../mozc-1.1.785.102.orig/hangul/session.cc ./hangul/session.cc
--- ../mozc-1.1.785.102.orig/hangul/session.cc	2011-07-19 16:33:43.000000000 +0900
+++ ./hangul/session.cc	2011-07-22 17:19:37.747752642 +0900
@@ -319,7 +319,21 @@
   }
 
   const string kDigits = "1234567890";
-  session::SessionOutput::FillShortcuts(kDigits, candidates);
+  {
+    // Logic here is copied from SessionOutput::FillShortcuts.  We
+    // can't reuse this at this time because SessionOutput depends on
+    // converter/segments.cc, which depends on the Japanese language
+    // model.
+    // TODO(mukai): extract FillShortcuts() method to another library.
+    const size_t num_loop = min(
+        static_cast<size_t>(candidates->candidate_size()),
+        kDigits.size());
+    for (size_t i = 0; i < num_loop; ++i) {
+      const string shortcut = kDigits.substr(i, 1);
+      candidates->mutable_candidate(i)->mutable_annotation()->
+          set_shortcut(shortcut);
+    }
+  }
 
   candidates->set_focused_index(hanja_index_);
   candidates->set_direction(commands::Candidates::HORIZONTAL);
diff -ru ../mozc-1.1.785.102.orig/session/session.gyp ./session/session.gyp
--- ../mozc-1.1.785.102.orig/session/session.gyp	2011-07-19 16:33:43.000000000 +0900
+++ ./session/session.gyp	2011-07-22 17:26:08.227385317 +0900
@@ -83,7 +83,7 @@
         'session_watch_dog.cc',
       ],
       'dependencies': [
-        '../converter/converter.gyp:converter',
+        '../client/client.gyp:client',
         'session_base.gyp:genproto_session',
         'session_base.gyp:session_protocol',
       ]
