From 623160fa59c7fa2c01467336e5fa62f53203d47a Mon Sep 17 00:00:00 2001
From: Dave Mielke <dave@mielke.cc>
Date: Wed, 18 Jan 2017 20:24:40 -0500
Subject: [PATCH 4/6] Create d Baum device entry for the Orbit. (dm)

---
 Drivers/Braille/Baum/braille.c | 13 +++++++++++++
 1 file changed, 13 insertions(+)

diff --git a/Drivers/Braille/Baum/braille.c b/Drivers/Braille/Baum/braille.c
index 75ce2bccb..2881952df 100644
--- a/Drivers/Braille/Baum/braille.c
+++ b/Drivers/Braille/Baum/braille.c
@@ -179,6 +179,12 @@ BEGIN_KEY_NAME_TABLES(conny)
   KEY_NAME_TABLE(routing),
 END_KEY_NAME_TABLES
 
+BEGIN_KEY_NAME_TABLES(orbit)
+  KEY_NAME_SUBTABLE(display,6),
+  KEY_NAME_TABLE(entry),
+  KEY_NAME_TABLE(joystick),
+END_KEY_NAME_TABLES
+
 BEGIN_KEY_NAME_TABLES(pronto)
   KEY_NAME_SUBTABLE(display,6),
   KEY_NAME_TABLE(entry),
@@ -251,6 +257,7 @@ END_KEY_NAME_TABLES
 DEFINE_KEY_TABLE(default)
 DEFINE_KEY_TABLE(connect)
 DEFINE_KEY_TABLE(conny)
+DEFINE_KEY_TABLE(orbit)
 DEFINE_KEY_TABLE(pronto)
 DEFINE_KEY_TABLE(pv)
 DEFINE_KEY_TABLE(rb)
@@ -267,6 +274,7 @@ BEGIN_KEY_TABLE_LIST
   &KEY_TABLE_DEFINITION(default),
   &KEY_TABLE_DEFINITION(connect),
   &KEY_TABLE_DEFINITION(conny),
+  &KEY_TABLE_DEFINITION(orbit),
   &KEY_TABLE_DEFINITION(pronto),
   &KEY_TABLE_DEFINITION(pv),
   &KEY_TABLE_DEFINITION(rb),
@@ -803,6 +811,7 @@ typedef union {
 typedef enum {
   BAUM_DEVICE_Default,
   BAUM_DEVICE_Conny,
+  BAUM_DEVICE_Orbit,
   BAUM_DEVICE_PocketVario,
   BAUM_DEVICE_Pronto,
   BAUM_DEVICE_Refreshabraille,
@@ -835,6 +844,10 @@ static const BaumDeviceIdentityEntry baumDeviceIdentityTable[] = {
     .type = BAUM_DEVICE_Conny
   },
 
+  { .string = "Orbit",
+    .type = BAUM_DEVICE_Orbit
+  },
+
   { .string = "PocketVario",
     .type = BAUM_DEVICE_PocketVario
   },
-- 
2.11.0.483.g087da7b7c-goog

