From 8648487e687bbac36ffb32e6df91ab1ec7b94603 Mon Sep 17 00:00:00 2001
From: Dave Mielke <dave@mielke.cc>
Date: Wed, 18 Jan 2017 20:30:37 -0500
Subject: [PATCH 5/6] Add an entry for the Orbit's Bluetooth name. (dm)

---
 Programs/bluetooth_names.c | 5 +++++
 1 file changed, 5 insertions(+)

diff --git a/Programs/bluetooth_names.c b/Programs/bluetooth_names.c
index e81ee6fd4..00f6da147 100644
--- a/Programs/bluetooth_names.c
+++ b/Programs/bluetooth_names.c
@@ -47,6 +47,7 @@ BLUETOOTH_NAME_DRIVERS(EL12, "al", "vo");
 BLUETOOTH_NAME_DRIVERS(Focus, "fs");
 BLUETOOTH_NAME_DRIVERS(HWGBrailliant, "bm");
 BLUETOOTH_NAME_DRIVERS(MB248, "md");
+BLUETOOTH_NAME_DRIVERS(OrbitReader, "bm");
 BLUETOOTH_NAME_DRIVERS(Pronto, "bm");
 BLUETOOTH_NAME_DRIVERS(Refreshabraille, "bm");
 BLUETOOTH_NAME_DRIVERS(SuperVario, "bm");
@@ -159,6 +160,10 @@ const BluetoothNameEntry bluetoothNameTable[] = {
     .driverCodes = bluetoothNameDrivers_MB248
   },
 
+  { .namePrefix = "Orbit Reader",
+    .driverCodes = bluetoothNameDrivers_OrbitReader
+  },
+
   { .namePrefix = "Pronto!",
     .driverCodes = bluetoothNameDrivers_Pronto
   },
-- 
2.11.0.483.g087da7b7c-goog

