--- o3d/build/gyp_o3d       (revision 182761)
+++ o3d/build/gyp_o3d       (working copy)
@@ -13,11 +13,10 @@
 chrome_src = os.path.join(os.path.dirname(sys.argv[0]), os.pardir, os.pardir)
 o3d_src = os.path.join(os.path.dirname(sys.argv[0]), os.pardir)
 
-try:
-  import gyp
-except ImportError, e:
-  sys.path.append(os.path.join(chrome_src, 'tools', 'gyp', 'pylib'))
-  import gyp
+# There's a copy of GYP in the chroot that's not compatible
+# so we need to force use of the copy in Chrome now.
+sys.path.insert(0, os.path.join(chrome_src, 'tools', 'gyp', 'pylib'))
+import gyp
 
 if __name__ == '__main__':
   args = sys.argv[1:]
