--- build/linux/system.gyp.orig	2011-08-23 16:21:40.435736620 -0700
+++ build/linux/system.gyp	2011-08-23 16:22:01.907612431 -0700
@@ -184,26 +184,6 @@
       }]]
     },
     {
-      'target_name': 'gconf',
-      'type': 'settings',
-      'conditions': [
-        ['_toolset=="target"', {
-          'direct_dependent_settings': {
-            'cflags': [
-              '<!@(<(pkg-config) --cflags gconf-2.0)',
-            ],
-          },
-          'link_settings': {
-            'ldflags': [
-              '<!@(<(pkg-config) --libs-only-L --libs-only-other gconf-2.0)',
-            ],
-            'libraries': [
-              '<!@(<(pkg-config) --libs-only-l gconf-2.0)',
-            ],
-          },
-      }]]
-    },
-    {
       'target_name': 'x11',
       'type': 'settings',
       'conditions': [
