--- tools/gyp/pylib/gyp/__init__.py	2011-06-02 20:50:26.000000000 +0200
+++ tools/gyp/pylib/gyp/__init__.py.new	2011-08-26 04:22:39.000000000 +0200
@@ -316,6 +316,7 @@
                            'freebsd7': 'make',
                            'freebsd8': 'make',
                            'linux2':   'make',
+                           'linux3':   'make',
                            'openbsd4': 'make',
                            'sunos5':   'make',}[sys.platform] ]
 
