--- o3d/plugin/plugin.gyp.orig	2011-06-06 20:49:48.705076650 -0700
+++ o3d/plugin/plugin.gyp	2011-06-06 20:49:48.725110333 -0700
@@ -311,16 +311,25 @@
             'ldflags': [
               '-Wl,-znodelete',
               '-Wl,--gc-sections',
-              '<!@(pkg-config --libs-only-L xt)',
+              '<!@(<(pkg-config) --libs-only-L xt)',
               '<(as_needed_ldflags)',
               # Directs the linker to only generate dependencies on libraries
               # that we actually use. Must come last.
               '-Wl,--as-needed',
             ],
             'libraries': [
-              '<!@(pkg-config --libs-only-l xt)',
+              '<!@(<(pkg-config) --libs-only-l xt)',
             ],
             'conditions' : [
+              ['sysroot!=""', {
+                'variables': {
+                  'pkg-config': './../../build/linux/pkg-config-wrapper "<(sysroot)"',
+                },
+              }, {
+                'variables': {
+                  'pkg-config': 'pkg-config'
+                },
+              }],
               ['plugin_rpath != ""',
                 {
                   'ldflags': [
--- third_party/sqlite/sqlite.gyp.orig	2011-06-06 20:56:14.825048138 -0700
+++ third_party/sqlite/sqlite.gyp	2011-06-06 20:58:05.285127341 -0700
@@ -38,9 +38,9 @@
               # This next command produces no output but it it will fail (and
               # cause GYP to fail) if we don't have a recent enough version of
               # sqlite.
-              '<!@(pkg-config --atleast-version=<(required_sqlite_version) sqlite3)',
+              '<!@(<(pkg-config) --atleast-version=<(required_sqlite_version) sqlite3)',
 
-              '<!@(pkg-config --cflags sqlite3)',
+              '<!@(<(pkg-config) --cflags sqlite3)',
             ],
             'defines': [
               'USE_SYSTEM_SQLITE',
@@ -48,12 +48,23 @@
           },
           'link_settings': {
             'ldflags': [
-              '<!@(pkg-config --libs-only-L --libs-only-other sqlite3)',
+              '<!@(<(pkg-config) --libs-only-L --libs-only-other sqlite3)',
             ],
             'libraries': [
-              '<!@(pkg-config --libs-only-l sqlite3)',
+              '<!@(<(pkg-config) --libs-only-l sqlite3)',
             ],
           },
+          'conditions': [
+            ['sysroot!=""', {
+              'variables': {
+                'pkg-config': './../../build/linux/pkg-config-wrapper "<(sysroot)"',
+              },
+            }, {
+              'variables': {
+                'pkg-config': 'pkg-config'
+              },
+            }],
+          ]
         }, { # else: OS != "linux" or ! use_system_sqlite
           'product_name': 'sqlite3',
           'type': '<(library)',
--- o3d/build/libs.gyp.orig	2011-06-06 21:05:48.265047484 -0700
+++ o3d/build/libs.gyp	2011-06-06 21:05:54.944729757 -0700
@@ -25,15 +25,26 @@
           {
             'all_dependent_settings': {
               'cflags': [
-                '<!@(pkg-config --cflags cairo)',
+                '<!@(<(pkg-config) --cflags cairo)',
               ],
               'ldflags': [
-                '<!@(pkg-config --libs-only-L --libs-only-other cairo)',
+                '<!@(<(pkg-config) --libs-only-L --libs-only-other cairo)',
               ],
               'libraries': [
-                '<!@(pkg-config --libs-only-l cairo)',
+                '<!@(<(pkg-config) --libs-only-l cairo)',
               ],
             },
+            'conditions': [
+              ['sysroot!=""', {
+                'variables': {
+                  'pkg-config': './../../build/linux/pkg-config-wrapper "<(sysroot)"',
+                },
+              }, {
+                'variables': {
+                  'pkg-config': 'pkg-config'
+                },
+              }],
+            ],
           },
         ],
         ['OS=="mac"',
