diff -urN nacl_mozc_org/nacl_mozc.js nacl_mozc/nacl_mozc.js
--- nacl_mozc_org/nacl_mozc.js	2013-06-12 09:52:17.777154000 +0900
+++ nacl_mozc/nacl_mozc.js	2013-06-12 11:28:53.205072000 +0900
@@ -273,6 +273,10 @@
     chrome.input.ime.onSurroundingTextChanged.addListener(
         this.wrapAsyncHandler_(this.onSurroundingTextChanged_));
   }
+  // chrome.input.ime.onReset is available from ChromeOS 29.
+  if (chrome.input.ime.onReset) {
+    chrome.input.ime.onReset.addListener(this.wrapAsyncHandler_(this.onReset_));
+  }
 };
 
 /**
@@ -1006,6 +1010,29 @@
 };
 
 /**
+ * Callback method called when Chrome terminates ongoing text input session.
+ * @param {string} engineID ID of the engine.
+ * @private
+ */
+mozc.NaclMozc.prototype.onReset_ = function(engineID) {
+  if (!this.sessionID_) {
+    console.error('Session has not been created.');
+    return;
+  }
+  this.postMozcCommand_(
+      {
+        'input': {
+          'type': 'SEND_COMMAND',
+          'id': this.sessionID_,
+          'command': {
+            'type': 'RESET_CONTEXT'
+          }
+        }
+      },
+      this.outputResponse_.bind(this));
+};
+
+/**
  * Switches the composition mode.
  * @param {mozc.CompositionMode} newMode The new composition mode to be set.
  * @private
