diff -u -r nacl-mozc-1.10.1401.4_org/key_translator.js nacl-mozc-1.10.1401.4/key_translator.js
--- nacl-mozc-1.10.1401.4_org/key_translator.js	2013-05-08 12:25:54.824052000 +0900
+++ nacl-mozc-1.10.1401.4/key_translator.js	2013-05-08 12:26:38.000703000 +0900
@@ -59,6 +59,9 @@
   'KeyO': ['o', 'O', '\u3089', '\u3089'],  // 'ら', 'ら'
   'KeyP': ['p', 'P', '\u305B', '\u305B'],  // 'せ', 'せ'
   'BracketLeft': ['@', '`', '\u309B', '\u309B'],  // '゛', '゛'
+  'BracketRight': ['[', '{', '\u309C', '\u300C'],  // '゜', '「'
+  // 'BlacketRight' is typographical error which is used in Chrome 28.
+  // TODO(horo): In Chrome 29 this error will be fixed and we should remove it.
   'BlacketRight': ['[', '{', '\u309C', '\u300C'],  // '゜', '「'
 
   'KeyA': ['a', 'A', '\u3061', '\u3061'],  // 'ち', 'ち'
@@ -119,6 +122,9 @@
   'KeyO': ['o', 'O', '\u3089', '\u3089'],  // 'ら', 'ら'
   'KeyP': ['p', 'P', '\u305B', '\u305B'],  // 'せ', 'せ'
   'BracketLeft': ['[', '{', '\u309B', '\u309B'],  // '゛', '゛'
+  'BracketRight': [']', '}', '\u309C', '\u300C'],  // '゜', '「'
+  // 'BlacketRight' is typographical error which is used in Chrome 28.
+  // TODO(horo): In Chrome 29 this error will be fixed and we should remove it.
   'BlacketRight': [']', '}', '\u309C', '\u300C'],  // '゜', '「'
   'Backslash': ['\\', '|', '\u3080', '\u300D'],  // 'む', '」'
 
