--- nacl-mozc-1.12.1599.104_orig/key_translator.js	2014-02-04 17:23:10.510220227 +0900
+++ nacl-mozc-1.12.1599.104/key_translator.js	2014-02-04 17:23:42.750218679 +0900
@@ -195,9 +195,7 @@
   'NumpadMultiply': 'MULTIPLY',
   'Space': 'SPACE',
   'CapsLock': 'CAPS_LOCK',
-  'Convert': 'HENKAN',
   'KanaMode': 'KANA',
-  'NoConvert': 'MUHENKAN',
   'PageUp': 'PAGE_UP',
   'End': 'END',
   'Delete': 'DEL',
@@ -238,9 +236,7 @@
   'NumpadMultiply': 'MULTIPLY',
   'Space': 'SPACE',
   'CapsLock': 'CAPS_LOCK',
-  'Convert': 'HENKAN',
   'KanaMode': 'KANA',
-  'NoConvert': 'MUHENKAN',
   'PageUp': 'PAGE_UP',
   'End': 'END',
   'Delete': 'DEL',
