diff -urN nacl-mozc-1.11.1503.4_3/key_translator.js nacl-mozc-1.11.1503.4_4/key_translator.js
--- nacl-mozc-1.11.1503.4_3/key_translator.js	2013-07-26 13:26:31.922436000 +0900
+++ nacl-mozc-1.11.1503.4_4/key_translator.js	2013-08-13 15:51:23.891554000 +0900
@@ -162,7 +162,6 @@
  * @private
  */
 mozc.SPECIAL_KEY_CODE_MAP_JP_ = {
-  'BackQuote': 'HANKAKU',
   'Esc': 'ESCAPE',
   'Backspace': 'BACKSPACE',
   'Tab': 'TAB',
