--- nacl-mozc-1.12.1600.4_orig/key_translator.js	2013-12-26 16:02:59.965259407 +0900
+++ nacl-mozc-1.12.1600.4/key_translator.js	2013-12-26 16:04:26.175245701 +0900
@@ -97,6 +97,9 @@
  * @private
  */
 mozc.KEY_CODE_MAP_US_ = {
+  'Backquote': ['`', '~', '\u308D', '\u308D'],  // 'ろ', 'ろ'
+  // 'BackQuote' is used before Chrome 33.
+  // TODO(hsumita): Remove this.
   'BackQuote': ['`', '~', '\u308D', '\u308D'],  // 'ろ', 'ろ'
   'Digit1': ['1', '!', '\u306C', '\u306C'],  // 'ぬ', 'ぬ'
   'Digit2': ['2', '@', '\u3075', '\u3075'],  // 'ふ', 'ふ'
@@ -162,6 +165,9 @@
  * @private
  */
 mozc.SPECIAL_KEY_CODE_MAP_JP_ = {
+  'Escape': 'ESCAPE',
+  // 'Esc' is used before Chrome 33
+  // TODO(hsumita): Remove this.
   'Esc': 'ESCAPE',
   'Backspace': 'BACKSPACE',
   'Tab': 'TAB',
@@ -205,6 +211,9 @@
  * @private
  */
 mozc.SPECIAL_KEY_CODE_MAP_US_ = {
+  'Escape': 'ESCAPE',
+  // 'Esc' is used before Chrome 33.
+  // TODO(hsumita): Remove this.
   'Esc': 'ESCAPE',
   'Backspace': 'BACKSPACE',
   'Tab': 'TAB',
