diff -urN nacl-mozc_org/manifest.json nacl-mozc/manifest.json
--- nacl-mozc_org/manifest.json	2013-07-03 17:44:52.521630000 +0900
+++ nacl-mozc/manifest.json	2013-07-03 17:44:55.257688000 +0900
@@ -7,7 +7,8 @@
     "https://clients4.google.com/",
     "https://dl.google.com/",
     "input",
-    "unlimitedStorage"
+    "unlimitedStorage",
+    "inputMethodPrivate"
   ],
   "manifest_version": 2,
   "minimum_chrome_version": "22",
diff -urN nacl-mozc_org/nacl_mozc_init.js nacl-mozc/nacl_mozc_init.js
--- nacl-mozc_org/nacl_mozc_init.js	2013-07-03 16:52:38.000000000 +0900
+++ nacl-mozc/nacl_mozc_init.js	2013-07-03 17:38:23.640606000 +0900
@@ -31,5 +31,9 @@
     window['newOptionPage'] = naclMozc.newOptionPage.bind(naclMozc);
     var body = document.getElementsByTagName('body').item(0);
     body.appendChild(naclModule);
+    // To notify to the browser the end of initialization, otherwise some
+    // important event(onActivate or onFocus) will be missed.
+    // TODO(nona): Provide same/similar feature for public.
+    chrome.inputMethodPrivate.startIme();
   }
 }, true);
