diff -urN nacl-mozc-1.11.1503.4/manifest.json nacl-mozc-1.11.1503.4_2/manifest.json
--- nacl-mozc-1.11.1503.4/manifest.json	2013-07-26 13:26:13.706153000 +0900
+++ nacl-mozc-1.11.1503.4_2/manifest.json	2013-07-26 13:27:18.774177000 +0900
@@ -38,7 +38,8 @@
     "https://clients4.google.com/", 
     "https://dl.google.com/", 
     "input", 
-    "unlimitedStorage"
+    "unlimitedStorage",
+    "inputMethodPrivate"
   ], 
   "version": "1.11.1503.4"
 }
\ No newline at end of file
diff -urN nacl-mozc-1.11.1503.4/nacl_mozc_init.js nacl-mozc-1.11.1503.4_2/nacl_mozc_init.js
--- nacl-mozc-1.11.1503.4/nacl_mozc_init.js	2013-07-26 11:53:22.000000000 +0900
+++ nacl-mozc-1.11.1503.4_2/nacl_mozc_init.js	2013-07-26 13:27:57.406789000 +0900
@@ -31,5 +31,9 @@
     window['newOptionPage'] = naclMozc.newOptionPage.bind(naclMozc);
     var body = document.getElementsByTagName('body').item(0);
     body.appendChild(naclModule);
+    // To notify to the browser the end of initialization, otherwise some
+    // important event(onActivate or onFocus) will be missed.
+    // TODO(nona): Provide same/similar feature for public.
+    chrome.inputMethodPrivate.startIme();
   }
 }, true);
