--- nacl-mozc-1.12.1599.104_orig/key_translator.js	2013-12-26 15:56:03.825319243 +0900
+++ nacl-mozc-1.12.1599.104/key_translator.js	2013-12-26 15:58:12.825300949 +0900
@@ -123,6 +123,9 @@
  * @private
  */
 mozc.KEY_CODE_MAP_US_ = {
+  'Backquote': ['`', '~', '\u308D', '\u308D'],  // 'ろ', 'ろ'
+  // 'BackQuote' is used before Chrome 33.
+  // TODO(hsumita): Remove this.
   'BackQuote': ['`', '~', '\u308D', '\u308D'],  // 'ろ', 'ろ'
   'Digit1': ['1', '!', '\u306C', '\u306C'],  // 'ぬ', 'ぬ'
   'Digit2': ['2', '@', '\u3075', '\u3075'],  // 'ふ', 'ふ'
@@ -188,6 +191,9 @@
  * @private
  */
 mozc.SPECIAL_KEY_CODE_MAP_JP_ = {
+  'Escape': 'ESCAPE',
+  // 'Esc' is used before Chrome 33
+  // TODO(hsumita): Remove this.
   'Esc': 'ESCAPE',
   'Backspace': 'BACKSPACE',
   'Tab': 'TAB',
@@ -231,6 +237,9 @@
  * @private
  */
 mozc.SPECIAL_KEY_CODE_MAP_US_ = {
+  'Escape': 'ESCAPE',
+  // 'Esc' is used before Chrome 33.
+  // TODO(hsumita): Remove this.
   'Esc': 'ESCAPE',
   'Backspace': 'BACKSPACE',
   'Tab': 'TAB',
