diff -urN nacl-mozc-1.11.1503.4_2/option_page.js nacl-mozc-1.11.1503.4_3/option_page.js
--- nacl-mozc-1.11.1503.4_2/option_page.js	2013-07-26 13:26:32.244446000 +0900
+++ nacl-mozc-1.11.1503.4_3/option_page.js	2013-07-30 15:46:43.755996000 +0900
@@ -752,6 +752,17 @@
         optionCheckbox.name;
   }
 
+  // Disables clear_history_ok button according to check boxes in Clear History
+  // dialog.
+  this.document_.getElementById('clear_history_conversion_history')
+      .addEventListener('change',
+                        this.updateClearHistoryOkButton_.bind(this),
+                        true);
+  this.document_.getElementById('clear_history_suggestion_history')
+      .addEventListener('change',
+                        this.updateClearHistoryOkButton_.bind(this),
+                        true);
+
   this.document_.getElementById('user_dictionary_select').addEventListener(
       'change', this.onDictionarySelectChanged_.bind(this), true);
   this.document_.getElementById('dictionary_tool_reading_new_input')
@@ -1240,6 +1251,19 @@
 };
 
 /**
+ * Disables clear_history_ok button according to check boxes in Clear History
+ * dialog.
+ * @private
+ */
+mozc.OptionPage.prototype.updateClearHistoryOkButton_ = function() {
+  this.document_.getElementById('clear_history_ok').disabled =
+      !this.document_.getElementById('clear_history_conversion_history')
+          .checked &&
+      !this.document_.getElementById('clear_history_suggestion_history')
+          .checked;
+};
+
+/**
  * Called when clear_history_ok button is clicked.
  * This method calls CLEAR_USER_HISTORY and CLEAR_USER_PREDICTION command
  * according to clear_history_conversion_history and
