diff -urN nacl-mozc-1.12.1577.4_org/nacl_mozc.js nacl-mozc-1.12.1577.4/nacl_mozc.js
--- nacl-mozc-1.12.1577.4_org/nacl_mozc.js	2013-10-30 13:27:22.621478000 +0900
+++ nacl-mozc-1.12.1577.4/nacl_mozc.js	2013-10-30 13:29:42.857612000 +0900
@@ -873,9 +873,9 @@
 mozc.NaclMozc.prototype.onActivate_ = function(engineID) {
   this.engine_id_ = engineID;
   // Sets keyboardLayout_.
-  var appDetails = chrome.app.getDetails();
-  if (appDetails) {
-    var input_components = chrome.app.getDetails()['input_components'];
+  var getManifest = chrome.runtime.getManifest;
+  if (getManifest) {
+    var input_components = getManifest()['input_components'];
     for (var i = 0; i < input_components.length; ++i) {
       if (input_components[i].id == engineID) {
         this.keyboardLayout_ = input_components[i]['layouts'][0];
