From c2c896679b1c120265fb6c769e472c2f4115c584 Mon Sep 17 00:00:00 2001
From: Peter Hutterer <peter.hutterer@who-t.net>
Date: Wed, 5 Jan 2011 10:11:31 +1000
Subject: [PATCH 11/18] Print MT sync events differently to SYN_REPORT.

Signed-off-by: Peter Hutterer <peter.hutterer@who-t.net>
---
 evtest.c |    8 +++++++-
 1 files changed, 7 insertions(+), 1 deletions(-)

diff --git a/evtest.c b/evtest.c
index a5525bf..06f7948 100644
--- a/evtest.c
+++ b/evtest.c
@@ -41,6 +41,9 @@
 #ifndef EV_SYN
 #define EV_SYN 0
 #endif
+#ifndef SYN_MT_REPORT
+#define SYN_MT_REPORT 2
+#endif
 
 char *events[EV_MAX + 1] = {
 	[0 ... EV_MAX] = NULL,
@@ -450,7 +453,10 @@ int main (int argc, char **argv)
 			printf("Event: time %ld.%06ld, ", ev[i].time.tv_sec, ev[i].time.tv_usec);
 
 			if (ev[i].type == EV_SYN) {
-				printf("-------------- %s ------------\n", syns[ev[i].code]);
+				if (ev[i].code == SYN_MT_REPORT)
+					printf("++++++++++++++ %s ++++++++++++\n", syns[ev[i].code]);
+				else
+					printf("-------------- %s ------------\n", syns[ev[i].code]);
 			} else {
 				printf("type %d (%s), code %d (%s), ",
 					ev[i].type,
-- 
1.7.3.1

