From 403808b4d257cb0062cec626a1a5a5c81751bf48 Mon Sep 17 00:00:00 2001
From: Peter Hutterer <peter.hutterer@who-t.net>
Date: Wed, 5 Jan 2011 10:17:36 +1000
Subject: [PATCH 13/18] Add KEY_RFKILL to keys database.

Signed-off-by: Peter Hutterer <peter.hutterer@who-t.net>
---
 evtest.c |    5 ++++-
 1 files changed, 4 insertions(+), 1 deletions(-)

diff --git a/evtest.c b/evtest.c
index c0d1f15..632e5af 100644
--- a/evtest.c
+++ b/evtest.c
@@ -280,7 +280,10 @@ char *keys[KEY_MAX + 1] = {
 	[KEY_SCREENLOCK] = "Screen Lock",	[KEY_SWITCHVIDEOMODE] = "Switch Video Mode",
 	[KEY_UWB] = "UWB",			[KEY_VIDEO_NEXT] = "Video Next",
 	[KEY_VIDEO_PREV] = "Video Prev",	[KEY_WIMAX] = "WIMAX",
-	[KEY_WLAN] = "WLAN"
+	[KEY_WLAN] = "WLAN",
+#ifdef KEY_RFKILL
+	[KEY_RFKILL] = "RF kill"
+#endif
 
 };
 
-- 
1.7.3.1

