diff -paur e2fsprogs-1.41.9.orig/misc/filefrag.c e2fsprogs-1.41.9.work/misc/filefrag.c
--- e2fsprogs-1.41.9.orig/misc/filefrag.c	2009-08-22 19:44:35.000000000 -0700
+++ e2fsprogs-1.41.9.work/misc/filefrag.c	2011-05-25 10:52:36.318791000 -0700
@@ -39,8 +39,8 @@ extern int optind;
 #include <sys/vfs.h>
 #include <sys/ioctl.h>
 #include <linux/fd.h>
-#include <ext2fs/ext2_types.h>
 #include <ext2fs/fiemap.h>
+#include <ext2fs/ext2_types.h>
 
 int verbose = 0;
 int no_bs = 0;		/* Don't use the files blocksize, use 1K blocksize */
Only in e2fsprogs-1.41.9.work/misc: filefrag.c~
