diff --git a/src/bind_lambda.h b/src/bind_lambda.h
index e6f367c..3e8389f 100644
--- a/src/bind_lambda.h
+++ b/src/bind_lambda.h
@@ -31,7 +31,7 @@ class LambdaAdapter<Lambda, R (Lambda::*)(Args... args)> {
  public:
   typedef R(RunType)(Args...);
   LambdaAdapter(Lambda lambda) : lambda_(lambda) {}
-  R Run(Args... args) { return lambda_(CallbackForward(args)...); }
+  R Run(Args... args) { return lambda_(std::move(args)...); }
 
  private:
   Lambda lambda_;
@@ -43,7 +43,7 @@ class LambdaAdapter<Lambda, R (Lambda::*)(Args... args) const> {
  public:
   typedef R(RunType)(Args...);
   LambdaAdapter(Lambda lambda) : lambda_(lambda) {}
-  R Run(Args... args) { return lambda_(CallbackForward(args)...); }
+  R Run(Args... args) { return lambda_(std::move(args)...); }
 
  private:
   Lambda lambda_;
diff --git a/src/config.cc b/src/config.cc
index 52f5ea1..a243ebd 100644
--- a/src/config.cc
+++ b/src/config.cc
@@ -11,6 +11,7 @@
 #include <base/json/json_reader.h>
 #include <base/json/json_writer.h>
 #include <base/logging.h>
+#include <base/memory/scoped_ptr.h>
 #include <base/strings/string_number_conversions.h>
 #include <base/values.h>
 #include <weave/enum_to_string.h>
diff --git a/src/streams.h b/src/streams.h
index 990f47c..cefcf94 100644
--- a/src/streams.h
+++ b/src/streams.h
@@ -5,6 +5,8 @@
 #ifndef LIBWEAVE_SRC_STREAMS_H_
 #define LIBWEAVE_SRC_STREAMS_H_
 
+#include <vector>
+
 #include <base/memory/weak_ptr.h>
 #include <weave/stream.h>
 
-- 
2.1.2

