diff --git a/include/evdev-properties.h b/include/evdev-properties.h
index 6ab42b3..2b00177 100644
--- a/include/evdev-properties.h
+++ b/include/evdev-properties.h
@@ -46,6 +46,9 @@
 /* CARD8, value range 0-32, 0 to always scroll */
 #define EVDEV_PROP_WHEEL_BUTTON "Evdev Wheel Emulation Button"
 
+/* Swap left and right buttons */
+#define EVDEV_PROP_SWITCH_BUTTONS "Evdev Switch Left and Right Buttons"
+
 /* Drag lock */
 /* CARD8, either 1 value or pairs, value range 0-32, 0 to disable a value*/
 #define EVDEV_PROP_DRAGLOCK "Evdev Drag Lock Buttons"
diff --git a/src/evdev.c b/src/evdev.c
index c888d6a..4352693 100644
--- a/src/evdev.c
+++ b/src/evdev.c
@@ -125,6 +125,7 @@ static void EvdevInitButtonLabels(EvdevPtr pEvdev, int natoms, Atom *atoms);
 static void EvdevInitProperty(DeviceIntPtr dev);
 static int EvdevSetProperty(DeviceIntPtr dev, Atom atom,
                             XIPropertyValuePtr val, BOOL checkonly);
+static Atom prop_switch_buttons = 0;
 static Atom prop_invert = 0;
 static Atom prop_calibration = 0;
 static Atom prop_swap = 0;
@@ -2407,6 +2408,8 @@ _X_EXPORT XF86ModuleData evdevModuleData =
 unsigned int
 EvdevUtilButtonEventToButtonNumber(EvdevPtr pEvdev, int code)
 {
+    if (pEvdev->switch_buttons && (code == BTN_LEFT || code == BTN_RIGHT))
+        code = code == BTN_LEFT ? BTN_RIGHT : BTN_LEFT;
     switch (code)
     {
         /* Mouse buttons */
@@ -2750,6 +2753,16 @@ EvdevInitProperty(DeviceIntPtr dev)
 
         XISetDevicePropertyDeletable(dev, prop_swap, FALSE);
 
+        prop_switch_buttons = MakeAtom(EVDEV_PROP_SWITCH_BUTTONS,
+                strlen(EVDEV_PROP_SWITCH_BUTTONS), TRUE);
+
+        rc = XIChangeDeviceProperty(dev, prop_switch_buttons, XA_INTEGER, 8,
+                PropModeReplace, 1, &pEvdev->switch_buttons, FALSE);
+        if (rc != Success)
+            return;
+
+        XISetDevicePropertyDeletable(dev, prop_switch_buttons, FALSE);
+
 #ifdef HAVE_LABELS
         /* Axis labelling */
         if ((pEvdev->num_vals > 0) && (prop_axis_label = XIGetKnownProperty(AXIS_LABEL_PROP)))
@@ -2793,6 +2806,17 @@ EvdevSetProperty(DeviceIntPtr dev, Atom atom, XIPropertyValuePtr val,
             pEvdev->invert_x = data[0];
             pEvdev->invert_y = data[1];
         }
+    } else if (atom == prop_switch_buttons)
+    {
+        BOOL* data;
+        if (val->format != 8 || val->size != 1 || val->type != XA_INTEGER)
+            return BadMatch;
+
+        if (!checkonly)
+        {
+            data = (BOOL*)val->data;
+            pEvdev->switch_buttons = data[0];
+        }
     } else if (atom == prop_calibration)
     {
         if (val->format != 32 || val->type != XA_INTEGER)
diff --git a/src/evdev.h b/src/evdev.h
index 4102672..34153aa 100644
--- a/src/evdev.h
+++ b/src/evdev.h
@@ -143,6 +143,7 @@ typedef struct {
     BOOL swap_axes;
     BOOL invert_x;
     BOOL invert_y;
+    BOOL switch_buttons;
 
     int delta[REL_CNT];
     unsigned int abs_queued, rel_queued, prox_queued;
