diff -Naur openssh-5.2p1/sshd_config openssh-5.2p1-chromiumos/sshd_config
--- openssh-5.2p1/sshd_config	2010-02-22 20:33:12.000000000 +0000
+++ openssh-5.2p1-chromiumos/sshd_config	2010-02-22 20:37:22.000000000 +0000
@@ -20,11 +20,11 @@
 # activation of protocol 1
 Protocol 2
 
-# HostKey for protocol version 1
-#HostKey /etc/ssh/ssh_host_key
+# /etc is read-only.  Fetch keys from stateful partition
+# Not using v1, so no v1 key
 # HostKeys for protocol version 2
-#HostKey /etc/ssh/ssh_host_rsa_key
-#HostKey /etc/ssh/ssh_host_dsa_key
+HostKey /mnt/stateful_partition/etc/ssh/ssh_host_rsa_key
+HostKey /mnt/stateful_partition/etc/ssh/ssh_host_dsa_key
 
 # Lifetime and size of ephemeral version 1 server key
 #KeyRegenerationInterval 1h
