diff --git a/base/message_loop/message_loop.cc b/base/message_loop/message_loop.cc
index dd1a393..711430a 100644
--- a/base/message_loop/message_loop.cc
+++ b/base/message_loop/message_loop.cc
@@ -32,7 +32,7 @@
 #if defined(OS_ANDROID)
 #include "base/message_loop/message_pump_android.h"
 #endif
-#if defined(USE_GLIB)
+#if defined(USE_GLIB) || defined(OS_CHROMEOS)
 #include "base/message_loop/message_pump_glib.h"
 #endif
 
@@ -213,7 +213,7 @@ bool MessageLoop::InitMessagePumpForUIFactory(MessagePumpFactory* factory) {
 // static
 scoped_ptr<MessagePump> MessageLoop::CreateMessagePumpForType(Type type) {
 // TODO(rvargas): Get rid of the OS guards.
-#if defined(USE_GLIB) && !defined(OS_NACL)
+#if (defined(USE_GLIB) && !defined(OS_NACL)) || defined(OS_CHROMEOS)
   typedef MessagePumpGlib MessagePumpForUI;
 #elif defined(OS_LINUX) && !defined(OS_NACL)
   typedef MessagePumpLibevent MessagePumpForUI;
