--- src/pluto/connections.corig	2011-06-07 16:38:39.000000000 -0700
+++ src/pluto/connections.c	2011-06-07 16:39:00.000000000 -0700
@@ -897,6 +897,9 @@
 
 	dst->is_left = is_left;
 	dst->id = identification_create_from_string(src->id);
+	if (!is_left && src->id && streq(src->id, "%usepeercert")) {
+		dst->use_peer_cert = TRUE;
+	}
 	dst->ca = NULL;
 
 	/* decode CA distinguished name, if any */
--- src/pluto/connections.h.orig	2011-06-07 16:35:33.000000000 -0700
+++ src/pluto/connections.h	2011-06-07 16:35:52.000000000 -0700
@@ -159,6 +159,7 @@
 								/* rules if client behind host is a subnet */
 	bool allow_any;             /* IP address is subject to change */
 	certpolicy_t sendcert;      /* whether or not to send the certificate */
+        bool use_peer_cert;
 };
 
 struct spd_route {
--- src/pluto/ipsec_doi.c.orig	2011-06-07 16:40:25.000000000 -0700
+++ src/pluto/ipsec_doi.c	2011-06-07 16:42:21.000000000 -0700
@@ -2416,7 +2416,11 @@
 	{
 		int pathlen;
 
-		if (!peer->equals(peer, c->spd.that.id))
+		if (c->spd.that.use_peer_cert)
+		{
+			loglog(RC_LOG_SERIOUS, "skipping id check");
+		}
+		else if (!peer->equals(peer, c->spd.that.id))
 		{
 			loglog(RC_LOG_SERIOUS,
 					"we require peer to have ID '%Y', but peer declares '%Y'",
@@ -5862,4 +5866,3 @@
 		loglog(RC_LOG_SERIOUS, "DPD: unknown action");
 	}
 }
-
