--- strongswan-4.6.4/src/pluto/ipsec_doi.c.orig	2013-01-20 16:00:51.833391729 -0800
+++ strongswan-4.6.4/src/pluto/ipsec_doi.c	2013-01-20 16:02:20.873407218 -0800
@@ -2713,6 +2713,7 @@
 						 const char *which)
 {
 	ip_subnet net_temp;
+	extern bool ignore_peer_id_check;
 
 	if (!decode_net_id(id, id_pbs, &net_temp, which))
 	{
@@ -2722,7 +2723,11 @@
 	|| *protoid != id->isaiid_protoid || *port != id->isaiid_port)
 	{
 		loglog(RC_LOG_SERIOUS, "%s ID returned doesn't match my proposal", which);
-		return FALSE;
+		if (ignore_peer_id_check) {
+			loglog(RC_LOG_SERIOUS, "ignoring mismatch.");
+		} else {
+			return FALSE;
+		}
 	}
 	return TRUE;
 }
