--- src/pluto/plutomain.c.realorig	2010-09-06 02:46:15.000000000 -0700
+++ src/pluto/plutomain.c	2011-06-07 11:13:54.000000000 -0700
@@ -701,7 +701,9 @@
 	char buf[1024];
 
 		if (getpwnam_r(IPSEC_USER, &passwd, buf, sizeof(buf), &pwp) != 0 ||
-				pwp == NULL || setuid(pwp->pw_uid) != 0)
+				pwp == NULL ||
+				initgroups(pwp->pw_name, pwp->pw_gid) != 0 ||
+				setuid(pwp->pw_uid) != 0)
 		{
 			plog("unable to change daemon user");
 			abort();
