diff --git a/lib/lsan/lsan_common.cc b/lib/lsan/lsan_common.cc
index 746244c..f99cee0 100644
--- a/llvm/projects/compiler-rt/lib/lsan/lsan_common.cc
+++ b/llvm/projects/compiler-rt/lib/lsan/lsan_common.cc
@@ -79,7 +79,7 @@ static void InitializeFlags(bool standalone) {
     SetCommonFlagsDefaults(cf);
     cf->external_symbolizer_path = GetEnv("LSAN_SYMBOLIZER_PATH");
     cf->malloc_context_size = 30;
-    cf->detect_leaks = true;
+    cf->detect_leaks = false;
   }
   ParseCommonFlagsFromString(cf, options);
 }
diff --git a/lib/sanitizer_common/sanitizer_flags.cc b/lib/sanitizer_common/sanitizer_flags.cc
index 40b6ec0..1f0c9f2 100644
--- a/llvm/projects/compiler-rt/lib/sanitizer_common/sanitizer_flags.cc
+++ b/llvm/projects/compiler-rt/lib/sanitizer_common/sanitizer_flags.cc
@@ -46,7 +46,7 @@ void SetCommonFlagsDefaults(CommonFlags *f) {
   f->malloc_context_size = 1;
   f->log_path = "stderr";
   f->verbosity = 0;
-  f->detect_leaks = true;
+  f->detect_leaks = false;
   f->leak_check_at_exit = true;
   f->allocator_may_return_null = false;
   f->print_summary = true;
--- a/llvm/projects/compiler-rt/lib/asan/asan_rtl.cc
+++ b/llvm/projects/compiler-rt/lib/asan/asan_rtl.cc
@@ -233,7 +233,7 @@ static void ParseFlagsFromString(Flags *f, const char *str) {
 void InitializeFlags(Flags *f, const char *env) {
   CommonFlags *cf = common_flags();
   SetCommonFlagsDefaults(cf);
-  cf->detect_leaks = CAN_SANITIZE_LEAKS;
+  cf->detect_leaks = false;
   cf->external_symbolizer_path = GetEnv("ASAN_SYMBOLIZER_PATH");
   cf->malloc_context_size = kDefaultMallocContextSize;
   cf->intercept_tls_get_addr = true;

