https://code.google.com/p/gyp/issues/detail?id=438

Index: pylib/gyp/input.py
===================================================================
--- pylib/gyp/input.py	(revision 1590)
+++ pylib/gyp/input.py	(working copy)
@@ -932,7 +932,7 @@
         output = replacement[:]
       else:
         # Split it the same way sh would split arguments.
-        output = shlex.split(str(replacement))
+        output = shlex.split(str(replacement), posix=False)
     else:
       # Expanding in string context.
       encoded_replacement = ''
