diff -rNu xorg-server-1.11.99.902.orig/dix/getevents.c xorg-server-1.11.99.902/dix/getevents.c
--- xorg-server-1.11.99.902.orig/dix/getevents.c	2012-01-24 00:02:46.000000000 -0500
+++ xorg-server-1.11.99.902/dix/getevents.c	2012-03-08 12:27:45.544212478 -0500
@@ -891,11 +891,11 @@
                double *devx, double *devy,
                double *screenx, double *screeny)
 {
-    ScreenPtr scr = miPointerGetScreen(dev);
+    ScreenPtr scr;
     double tmpx, tmpy;
 
     if (!dev->valuator || dev->valuator->numAxes < 2)
-        return scr;
+        return miPointerGetScreen(dev);
 
     tmpx = *screenx;
     tmpy = *screeny;
@@ -904,6 +904,8 @@
      * clipping to the current screen. Coordinates returned are in desktop
      * coord system */
     scr = miPointerSetPosition(dev, mode, screenx, screeny);
+    if (!scr)
+        scr = miPointerGetScreen(dev);
 
     /* If we were constrained, rescale x/y from the screen coordinates so
      * the device valuators reflect the correct position. For screen
