diff --git a/hw/xfree86/dri2/dri2.c b/hw/xfree86/dri2/dri2.c
index 5cc9068..d03d082 100644
--- a/hw/xfree86/dri2/dri2.c
+++ b/hw/xfree86/dri2/dri2.c
@@ -216,7 +216,9 @@ DRI2SwapLimit(DrawablePtr pDraw, int swap_limit)
 
     if (pPriv->target_sbc == -1 && !pPriv->blockedOnMsc) {
 	if (pPriv->blockedClient) {
-	    AttendClient(pPriv->blockedClient);
+	    if ((pPriv->blockedClient->clientState != ClientStateGone) &&
+	        (pPriv->blockedClient->clientState != ClientStateInitial))
+		AttendClient(pPriv->blockedClient);
 	    pPriv->blockedClient = NULL;
 	}
     }
