From 17e52f63a3a1dc80af1a5b21a62c8a1b2d22f3f5 Mon Sep 17 00:00:00 2001
From: Sadrul Habib Chowdhury <sadrul@chromium.org>
Date: Fri, 15 Apr 2011 13:50:20 -0400
Subject: [PATCH] Add a label for ABS_MT_SLOT


Signed-off-by: Sadrul Habib Chowdhury <sadrul@chromium.org>
---
 Xi/xiproperty.c              |    1 +
 include/xserver-properties.h |    1 +
 2 files changed, 2 insertions(+), 0 deletions(-)

diff --git a/Xi/xiproperty.c b/Xi/xiproperty.c
index b9f53f7..4dd39c3 100644
--- a/Xi/xiproperty.c
+++ b/Xi/xiproperty.c
@@ -93,6 +93,7 @@ static struct dev_properties
     {0, AXIS_LABEL_PROP_ABS_TILT_Y},
     {0, AXIS_LABEL_PROP_ABS_TOOL_WIDTH},
     {0, AXIS_LABEL_PROP_ABS_VOLUME},
+    {0, AXIS_LABEL_PROP_ABS_MT_SLOT},
     {0, AXIS_LABEL_PROP_ABS_MT_TOUCH_MAJOR},
     {0, AXIS_LABEL_PROP_ABS_MT_TOUCH_MINOR},
     {0, AXIS_LABEL_PROP_ABS_MT_WIDTH_MAJOR},
diff --git a/include/xserver-properties.h b/include/xserver-properties.h
index c6259ae..fd7ce91 100644
--- a/include/xserver-properties.h
+++ b/include/xserver-properties.h
@@ -96,6 +96,7 @@
 #define AXIS_LABEL_PROP_ABS_TILT_Y      "Abs Tilt Y"
 #define AXIS_LABEL_PROP_ABS_TOOL_WIDTH  "Abs Tool Width"
 #define AXIS_LABEL_PROP_ABS_VOLUME      "Abs Volume"
+#define AXIS_LABEL_PROP_ABS_MT_SLOT        "Abs MT Slot ID"
 #define AXIS_LABEL_PROP_ABS_MT_TOUCH_MAJOR "Abs MT Touch Major"
 #define AXIS_LABEL_PROP_ABS_MT_TOUCH_MINOR "Abs MT Touch Minor"
 #define AXIS_LABEL_PROP_ABS_MT_WIDTH_MAJOR "Abs MT Width Major"
--
1.7.3.2

