diff -paur xorg-server-1.9.3.orig/damageext/damageext.c xorg-server-1.9.3.work/damageext/damageext.c
--- xorg-server-1.9.3.orig/damageext/damageext.c	2010-12-22 18:58:55.424518000 -0800
+++ xorg-server-1.9.3.work/damageext/damageext.c	2011-05-06 15:39:19.805940000 -0700
@@ -217,7 +217,6 @@ ProcDamageCreate (ClientPtr client)
     if (!AddResource (stuff->damage, DamageExtType, (pointer) pDamageExt))
 	return BadAlloc;
 
-    DamageSetReportAfterOp (pDamageExt->pDamage, TRUE);
     DamageRegister (pDamageExt->pDrawable, pDamageExt->pDamage);
 
     if (pDrawable->type == DRAWABLE_WINDOW)
Only in xorg-server-1.9.3.work/damageext: damageext.c~
