From 78054ab511078ed205f3a5a403009d44a61e9c0d Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Ville=20Syrj=C3=A4l=C3=A4?= <ville.syrjala@nokia.com>
Date: Wed, 12 Jan 2011 10:48:54 -0800
Subject: [PATCH 1/4] ChangeGC changes the GC, so ValidateGC should be called after it, not
 before.
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit
X-NVConfidentiality: public

Signed-off-by: Ville Syrjälä <ville.syrjala@nokia.com>
Reviewed-by: Keith Packard <keithp@keithp.com>
Reviewed-by: Adam Jackson <ajax@redhat.com>
---
 composite/compalloc.c |    3 +--
 1 files changed, 1 insertions(+), 2 deletions(-)

diff --git a/composite/compalloc.c b/composite/compalloc.c
index 246e4c5..1138b9a 100644
--- a/composite/compalloc.c
+++ b/composite/compalloc.c
@@ -500,9 +500,8 @@ compNewPixmap (WindowPtr pWin, int x, int y, int w, int h)
 	{
 	    ChangeGCVal val;
 	    val.val = IncludeInferiors;
-	    
+	    ChangeGC (NullClient, pGC, GCSubwindowMode, &val);
 	    ValidateGC(&pPixmap->drawable, pGC);
-	    ChangeGC (serverClient, pGC, GCSubwindowMode, &val);
 	    (*pGC->ops->CopyArea) (&pParent->drawable,
 				   &pPixmap->drawable,
 				   pGC,
-- 
1.7.1

