diff -aurp xorg-server-1.9.3/dix/window.c xorg-server-1.9.3-root-none//dix/window.c
--- xorg-server-1.9.3/dix/window.c	2011-02-21 00:10:46.004374317 -0800
+++ xorg-server-1.9.3-root-none//dix/window.c	2011-02-21 00:15:13.524249906 -0800
@@ -992,12 +992,8 @@ ChangeWindowAttributes(WindowPtr pWin, M
 	    {
 		if (pWin->backgroundState == BackgroundPixmap)
 		    (*pScreen->DestroyPixmap)(pWin->background.pixmap);
-		if (!pWin->parent)
-		    MakeRootTile(pWin);
-		else {
-		    pWin->backgroundState = XaceBackgroundNoneState(pWin);
-		    pWin->background.pixel = pScreen->whitePixel;
-		}
+		pWin->backgroundState = XaceBackgroundNoneState(pWin);
+		pWin->background.pixel = pScreen->whitePixel;
 	    }
 	    else if (pixID == ParentRelative)
 	    {
