diff -paur xorg-server-1.9.3.orig/mi/mieq.c xorg-server-1.9.3.work/mi/mieq.c
--- xorg-server-1.9.3.orig/mi/mieq.c	2010-12-22 18:59:00.677024000 -0800
+++ xorg-server-1.9.3.work/mi/mieq.c	2011-05-09 13:37:50.672865000 -0700
@@ -58,7 +58,7 @@ in this Software without prior written a
 # include <X11/extensions/dpmsconst.h>
 #endif
 
-#define QUEUE_SIZE  512
+#define QUEUE_SIZE  2048
 
 #define EnqueueScreen(dev) dev->spriteInfo->sprite->pEnqueueScreen
 #define DequeueScreen(dev) dev->spriteInfo->sprite->pDequeueScreen
