--- xorg-server-1.9.3.orig/glx/glapi.c	2010-12-22 18:58:56.004461000 -0800
+++ xorg-server-1.9.3.work/glx/glapi.c	2011-02-09 13:44:37.054739000 -0800
@@ -892,7 +892,7 @@ _glapi_get_proc_offset(const char *funcN
  * in the name of static functions, try generating a new API entrypoint on
  * the fly with assembly language.
  */
-_glapi_proc
+PUBLIC _glapi_proc
 _glapi_get_proc_address(const char *funcName)
 {
    struct _glapi_function * entry;
