From 0803a3b013729c337951e532f4bf89f10cc6eadb Mon Sep 17 00:00:00 2001
From: Simon Glass <sjg@chromium.org>
Date: Mon, 20 Jun 2011 16:21:44 -0700
Subject: [PATCH 4/4] Add missing tests to .gitignore


Signed-off-by: Simon Glass <sjg@chromium.org>
---
 tests/.gitignore |    2 ++
 1 files changed, 2 insertions(+), 0 deletions(-)

diff --git a/tests/.gitignore b/tests/.gitignore
index c4e1205..f4e58b2 100644
--- a/tests/.gitignore
+++ b/tests/.gitignore
@@ -45,3 +45,5 @@
 /sw_tree1
 /truncated_property
 /value-labels
+/dtb_reverse
+/dtbs_equal_unordered
-- 
1.7.3.1

