diff -urN xsendevt-0.6-org/xsendevt.cpp xsendevt-0.6-endptr/xsendevt.cpp
--- xsendevt-0.6-org/xsendevt.cpp	2006-04-19 17:33:55.000000000 -0700
+++ xsendevt-0.6-endptr/xsendevt.cpp	2013-11-14 20:42:48.039901574 -0800
@@ -141,7 +141,7 @@
                 type=Release;
                 strptr++;
             }
-            char *endptr=strchr(strptr,']');
+            const char *endptr=strchr(strptr,']');
             if (endptr == '\0')
             {
                printerr("No closing \"]\" after %s\n", strptr);
