diff -Naur dbus-glib-0.80.orig/dbus/Makefile.am dbus-glib-0.80.new/dbus/Makefile.am
--- dbus/Makefile.am	2009-01-30 21:05:36.000000000 +0000
+++ dbus/Makefile.am	2009-10-20 01:23:26.000000000 +0000
@@ -1,4 +1,4 @@
-SUBDIRS = . examples
+SUBDIRS = .

 INCLUDES = \
 	-I$(top_srcdir)						\
diff -Naur dbus-glib-0.80.orig/dbus/Makefile.in dbus-glib-0.80.new/dbus/Makefile.in
--- dbus/Makefile.in	2009-02-02 20:41:21.000000000 +0000
+++ dbus/Makefile.in	2009-10-20 01:23:37.000000000 +0000
@@ -315,7 +315,7 @@
 top_build_prefix = @top_build_prefix@
 top_builddir = @top_builddir@
 top_srcdir = @top_srcdir@
-SUBDIRS = . examples
+SUBDIRS = .
 INCLUDES = \
 	-I$(top_srcdir)						\
 	-I$(top_builddir)					\
