diff --git a/proto/chrome_user_metrics_extension.proto b/proto/chrome_user_metrics_extension.proto
index 7efa5b4..b4267f4 100644
--- a/proto/chrome_user_metrics_extension.proto
+++ b/proto/chrome_user_metrics_extension.proto
@@ -25,9 +25,13 @@ message ChromeUserMetricsExtension {
     // Google Chrome product family.
     CHROME = 0;
   }
-  // The product corresponding to this log. Note: The default value is Chrome,
-  // so Chrome products will not transmit this field.
-  optional Product product = 10 [default = CHROME];
+  // The product corresponding to this log. The field type is int32 instead of
+  // Product so that downstream users of the Chromium metrics component can
+  // introduce products without needing to make changes to the Chromium code
+  // (though they still need to add the new product to the server-side enum).
+  // Note: The default value is Chrome, so Chrome products will not transmit
+  // this field.
+  optional int32 product = 10 [default = 0];
 
   // The id of the client install that generated these events.
   //
