diff -paur xf86-video-intel-2.14.0.orig/src/intel.h xf86-video-intel-2.14.0.work/src/intel.h
--- xf86-video-intel-2.14.0.orig/src/intel.h	2011-01-10 15:02:16.527507000 -0800
+++ xf86-video-intel-2.14.0.work/src/intel.h	2011-01-12 16:35:15.310834000 -0800
@@ -472,6 +472,7 @@ typedef struct intel_screen_private {
 	OptionInfoPtr Options;
 
 	/* Driver phase/state information */
+	Bool starting;
 	Bool suspended;
 
 	enum last_3d last_3d;
@@ -501,6 +502,7 @@ extern void intel_mode_fini(intel_screen
 extern int intel_get_pipe_from_crtc_id(drm_intel_bufmgr *bufmgr, xf86CrtcPtr crtc);
 extern int intel_crtc_id(xf86CrtcPtr crtc);
 extern int intel_output_dpms_status(xf86OutputPtr output);
+extern void drmmode_copy_fb(ScrnInfoPtr scrn);
 
 extern Bool intel_do_pageflip(intel_screen_private *intel,
 			      dri_bo *new_front,
diff -paur xf86-video-intel-2.14.0.orig/src/intel_display.c xf86-video-intel-2.14.0.work/src/intel_display.c
--- xf86-video-intel-2.14.0.orig/src/intel_display.c	2011-01-10 15:02:16.534513000 -0800
+++ xf86-video-intel-2.14.0.work/src/intel_display.c	2011-01-13 10:35:19.498460000 -0800
@@ -35,6 +35,7 @@
 #include <unistd.h>
 #include <errno.h>
 #include <poll.h>
+#include <sys/ioctl.h>
 
 #include "xorgVersion.h"
 
@@ -928,8 +929,15 @@ intel_output_dpms(xf86OutputPtr output, 
 	drmModeConnectorPtr koutput = intel_output->mode_output;
 	struct intel_mode *mode = intel_output->mode;
 	int i;
+	intel_screen_private *intel = intel_get_screen_private(output->scrn);
 
-	for (i = 0; i < koutput->count_props; i++) {
+	/* xf86Crtc.c calls dpms off in set desired modes, so ignore
+	 * the request if we're starting up. */
+
+	if (intel->starting)
+		return;
+
+ 	for (i = 0; i < koutput->count_props; i++) {
 		drmModePropertyPtr props;
 
 		props = drmModeGetProperty(mode->fd, koutput->props[i]);
@@ -1614,6 +1622,8 @@ Bool intel_mode_pre_init(ScrnInfoPtr scr
 	}
 
 	intel->modes = mode;
+	scrn->canDoBGNoneRoot = TRUE;
+
 	return TRUE;
 }
 
@@ -1684,3 +1694,98 @@ intel_crtc_id(xf86CrtcPtr crtc)
 {
 	return crtc_id(crtc->driver_private);
 }
+
+
+static PixmapPtr
+drmmode_create_pixmap_for_fbcon(ScrnInfoPtr scrn)
+{
+	xf86CrtcConfigPtr xf86_config = XF86_CRTC_CONFIG_PTR(scrn);
+	struct intel_crtc* drmmode_crtc = xf86_config->crtc[0]->driver_private;
+	ScreenPtr pScreen = screenInfo.screens[scrn->scrnIndex];
+	struct intel_mode* drmmode = drmmode_crtc->mode;
+	intel_screen_private *intel = intel_get_screen_private(scrn);
+	drmModeFBPtr fbcon = NULL;
+	struct drm_gem_flink flink;
+	drm_intel_bo *bo;
+	PixmapPtr pixmap = NULL;
+	int i;
+
+	for (i = 0; i < drmmode->mode_res->count_crtcs; i++) {
+		drmmode_crtc = xf86_config->crtc[i]->driver_private;
+		if (drmmode_crtc->mode_crtc->buffer_id == 0)
+			continue;
+		fbcon = drmModeGetFB(drmmode->fd,
+				     drmmode_crtc->mode_crtc->buffer_id);
+		if (fbcon != NULL)
+			break;
+	}
+	if (i == drmmode->mode_res->count_crtcs)
+		return NULL;
+
+	flink.handle = fbcon->handle;
+	if (ioctl(drmmode->fd, DRM_IOCTL_GEM_FLINK, &flink) < 0) {
+		xf86DrvMsg(scrn->scrnIndex, X_ERROR,
+			   "Couldn't flink fbcon handle\n");
+		return NULL;
+	}
+
+	bo = drm_intel_bo_gem_create_from_name(intel->bufmgr,
+					       "fbcon", flink.name);
+	if (bo == NULL) {
+		xf86DrvMsg(scrn->scrnIndex, X_ERROR,
+			   "Couldn't allocate bo for fbcon handle\n");
+		return NULL;
+	}
+
+	pixmap = GetScratchPixmapHeader(pScreen,
+					fbcon->width, fbcon->height,
+					fbcon->depth, fbcon->bpp,
+					fbcon->pitch, NULL);
+	if (pixmap == NULL) {
+		xf86DrvMsg(scrn->scrnIndex, X_ERROR,
+			   "Couldn't allocate pixmap fbcon contents\n");
+		return NULL;
+	}
+
+	intel_set_pixmap_bo(pixmap, bo);
+	drm_intel_bo_unreference(bo);
+	drmModeFreeFB(fbcon);
+
+	return pixmap;
+}
+
+void drmmode_copy_fb(ScrnInfoPtr scrn)
+{
+	ScreenPtr pScreen = screenInfo.screens[scrn->scrnIndex];
+	intel_screen_private *intel = intel_get_screen_private(scrn);
+	PixmapPtr src, dst;
+	unsigned int pitch = scrn->displayWidth * intel->cpp;
+
+	src = drmmode_create_pixmap_for_fbcon(scrn);
+	if (src == NULL) {
+		xf86DrvMsg(scrn->scrnIndex, X_ERROR,
+			   "Couldn't create pixmap for fbcon\n");
+		return;
+	}
+
+	/* We dont have a screen Pixmap yet */
+	dst = GetScratchPixmapHeader(pScreen,
+				     scrn->virtualX, scrn->virtualY,
+				     scrn->depth, scrn->bitsPerPixel,
+				     pitch,
+				     NULL);
+	intel_set_pixmap_bo(dst, intel->front_buffer);
+
+	intel->uxa_driver->prepare_copy(src, dst, -1, -1, GXcopy, FB_ALLONES);
+
+	intel->uxa_driver->copy(dst, 0, 0, 0, 0,
+				scrn->virtualX, scrn->virtualY);
+
+	intel->uxa_driver->done_copy(dst);
+
+	intel_batch_submit (scrn, TRUE);
+
+	(*pScreen->DestroyPixmap)(src);
+	(*pScreen->DestroyPixmap)(dst);
+}
+
diff -paur xf86-video-intel-2.14.0.orig/src/intel_driver.c xf86-video-intel-2.14.0.work/src/intel_driver.c
--- xf86-video-intel-2.14.0.orig/src/intel_driver.c	2011-01-10 15:02:16.539518000 -0800
+++ xf86-video-intel-2.14.0.work/src/intel_driver.c	2011-01-13 10:35:01.569939000 -0800
@@ -1018,6 +1018,8 @@ I830ScreenInit(int scrnIndex, ScreenPtr 
 	if (INTEL_INFO(intel)->gen >= 40)
 		gen4_render_state_init(scrn);
 
+	intel->starting = TRUE;
+
 	miClearVisualTypes();
 	if (!miSetVisualTypes(scrn->depth,
 			      miGetDefaultVisualMask(scrn->depth),
@@ -1153,6 +1155,7 @@ I830ScreenInit(int scrnIndex, ScreenPtr 
 
 	intel_mode_init(intel);
 
+	intel->starting = FALSE;
 	intel->suspended = FALSE;
 
 #if HAVE_UDEV
@@ -1222,6 +1225,11 @@ static Bool I830EnterVT(int scrnIndex, i
 
 	intel_set_gem_max_sizes(scrn);
 
+	if (!CreateScratchPixmapsForScreen(scrn->scrnIndex))
+		return FALSE;
+
+	drmmode_copy_fb(scrn);
+
 	if (!xf86SetDesiredModes(scrn))
 		return FALSE;
 
