diff -paur xf86-video-intel-2.15.0.orig/src/intel_display.c xf86-video-intel-2.15.0.work/src/intel_display.c
--- xf86-video-intel-2.15.0.orig/src/intel_display.c	2011-04-14 02:16:40.000000000 -0700
+++ xf86-video-intel-2.15.0.work/src/intel_display.c	2011-06-16 20:44:35.336894000 -0700
@@ -1607,6 +1607,7 @@ Bool intel_mode_pre_init(ScrnInfoPtr scr
 	gp.value = &has_flipping;
 	(void)drmCommandWriteRead(intel->drmSubFD, DRM_I915_GETPARAM, &gp,
 				  sizeof(gp));
+	has_flipping = 0;
 	if (has_flipping) {
 		xf86DrvMsg(scrn->scrnIndex, X_INFO,
 			   "Kernel page flipping support detected, enabling\n");
Only in xf86-video-intel-2.15.0.work/src: intel_display.c~
