Index: unix/ibus/mozc_engine.cc
===================================================================
--- unix/ibus/mozc_engine.cc	(revision 52)
+++ unix/ibus/mozc_engine.cc	(working copy)
@@ -1030,6 +1030,7 @@
   }
 
   config::Config mozc_config;
+  session_->GetConfig(&mozc_config);
   const google::protobuf::Descriptor *descriptor = mozc_config.GetDescriptor();
   const google::protobuf::Reflection *reflection = mozc_config.GetReflection();
   const google::protobuf::FieldDescriptor *field_to_update =
