--- unix/ibus/main.cc	2010-11-08 12:07:28.000000000 -0800
+++ unix/ibus/main.cc	2010-11-11 06:04:37.000000000 -0800
@@ -17,7 +17,10 @@
 namespace {
 
 IBusBus *g_bus = NULL;
+#ifdef OS_CHROMEOS
+// We use the ibus configuration daemon only on Chromium OS.
 IBusConfig *g_config = NULL;
+#endif
 
 #ifndef OS_CHROMEOS
 void EnableVerboseLog() {
@@ -71,12 +74,15 @@
                    "disconnected",
                    G_CALLBACK(mozc::ibus::MozcEngine::Disconnected),
                    NULL);
+
+#ifdef OS_CHROMEOS
   g_config = ibus_bus_get_config(g_bus);
   g_object_ref_sink(g_config);
   g_signal_connect(g_config,
                    "value-changed",
                    G_CALLBACK(mozc::ibus::MozcEngine::ConfigValueChanged),
                    NULL);
+#endif
 
   IBusComponent *component = GetIBusComponent();
   IBusFactory *factory = ibus_factory_new(ibus_bus_get_connection(g_bus));
@@ -113,8 +119,12 @@
   IgnoreSigChild();
 #endif
   ibus_main();
+
+#ifdef OS_CHROMEOS
   if (g_config) {
     g_object_unref(g_config);
   }
+#endif
+
   return 0;
 }
