diff -ur mozc-1.3.975.102.orig/unix/ibus/mozc_engine.cc mozc-1.3.975.102/unix/ibus/mozc_engine.cc
--- mozc-1.3.975.102.orig/unix/ibus/mozc_engine.cc	2012-01-26 18:15:23.000000000 +0900
+++ mozc-1.3.975.102/unix/ibus/mozc_engine.cc	2012-04-02 18:20:45.222060147 +0900
@@ -640,8 +640,10 @@
   }
 
   i = 0;
-  while ((prop = ibus_prop_list_get(prop_composition_mode_->sub_props, i++))) {
-    if (!g_strcmp0(property_name, prop->key)) {
+  while ((prop = ibus_prop_list_get(
+             const_cast<IBusPropList*>(
+                 ibus_property_get_sub_props(prop_composition_mode_)), i++))) {
+    if (!g_strcmp0(property_name, ibus_property_get_key(prop))) {
       const MozcEngineProperty *entry =
           reinterpret_cast<const MozcEngineProperty*>(
               g_object_get_data(G_OBJECT(prop), kGObjectDataKey));
@@ -1096,8 +1098,10 @@
 
   size_t i = 0;
   IBusProperty *prop = NULL;
-  while ((prop = ibus_prop_list_get(prop_composition_mode_->sub_props, i++))) {
-    if (!g_strcmp0(entry->key, prop->key)) {
+  while ((prop = ibus_prop_list_get(
+             const_cast<IBusPropList*>(
+                 ibus_property_get_sub_props(prop_composition_mode_)), i++))) {
+    if (!g_strcmp0(entry->key, ibus_property_get_key(prop))) {
       // Update the language panel.
       ibus_property_set_icon(prop_composition_mode_,
                              GetIconPath(entry->icon).c_str());
