diff -uNr mozc-1.1.690.102.orig/gui/gui.gyp mozc-1.1.690.102/gui/gui.gyp
--- mozc-1.1.690.102.orig/gui/gui.gyp	2011-04-15 19:41:28.000000000 +0900
+++ mozc-1.1.690.102/gui/gui.gyp	2011-04-16 13:06:24.115728586 +0900
@@ -354,7 +354,6 @@
       ],
       'dependencies': [
         '../base/base.gyp:base',
-        'zinnia.gyp:zinnia',
         'gen_character_pad_files',
         'gen_character_pad_cp932_data',
         'gen_character_pad_data',
