diff -ur elfutils-0.166.orig/configure.ac elfutils-0.166/configure.ac
--- elfutils-0.166.orig/configure.ac	2016-03-31 02:48:08.000000000 -0600
+++ elfutils-0.166/configure.ac	2017-10-05 10:28:52.917364437 -0600
@@ -336,31 +336,6 @@
 AM_CONDITIONAL(HAVE_NULL_DEREFERENCE_WARNING,
 	       [test "x$ac_cv_null_dereference" != "xno"])
 
-dnl Check if we have argp available from our libc
-AC_LINK_IFELSE(
-	[AC_LANG_PROGRAM(
-		[#include <argp.h>],
-		[int argc=1; char *argv[]={"test"}; argp_parse(0,argc,argv,0,0,0); return 0;]
-		)],
-	[libc_has_argp="true"],
-	[libc_has_argp="false"]
-)
-
-dnl If our libc doesn't provide argp, then test for libargp
-if test "$libc_has_argp" = "false" ; then
-	AC_MSG_WARN("libc does not have argp")
-	AC_CHECK_LIB([argp], [argp_parse], [have_argp="true"], [have_argp="false"])
-
-	if test "$have_argp" = "false"; then
-		AC_MSG_ERROR("no libargp found")
-	else
-		argp_LDADD="-largp"
-	fi
-else
-	argp_LDADD=""
-fi
-AC_SUBST([argp_LDADD])
-
 dnl The directories with content.
 
 dnl Documentation.
diff -ur elfutils-0.166.orig/lib/system.h elfutils-0.166/lib/system.h
--- elfutils-0.166.orig/lib/system.h	2015-08-21 06:22:37.000000000 -0600
+++ elfutils-0.166/lib/system.h	2017-10-05 10:28:52.917364437 -0600
@@ -29,7 +29,6 @@
 #ifndef LIB_SYSTEM_H
 #define LIB_SYSTEM_H	1
 
-#include <argp.h>
 #include <stddef.h>
 #include <stdint.h>
 #include <endian.h>
diff -ur elfutils-0.166.orig/libelf/elf_error.c elfutils-0.166/libelf/elf_error.c
--- elfutils-0.166.orig/libelf/elf_error.c	2016-01-12 05:49:19.000000000 -0700
+++ elfutils-0.166/libelf/elf_error.c	2017-10-05 10:28:52.917364437 -0600
@@ -32,13 +32,15 @@
 #endif
 
 #include <assert.h>
-#include <libintl.h>
 #include <stdbool.h>
 #include <stdint.h>
 #include <stdlib.h>
 
 #include "libelfP.h"
 
+/* no internalization */
+#define gettext(x)      (x)
+#define dgettext(x,y)   (y)
 
 /* The error number.  */
 static __thread int global_error;
diff -ur elfutils-0.166.orig/libelf/elf_getarsym.c elfutils-0.166/libelf/elf_getarsym.c
--- elfutils-0.166.orig/libelf/elf_getarsym.c	2015-11-27 06:36:29.000000000 -0700
+++ elfutils-0.166/libelf/elf_getarsym.c	2017-10-05 10:29:27.557389223 -0600
@@ -45,6 +45,14 @@
 #include <dl-hash.h>
 #include "libelfP.h"
 
+static inline void *rawmemchr(const void *s, int c)
+{
+  const unsigned char *ptr = s;
+  while (1) {
+    if (*ptr == c) return (void *) ptr;
+    ptr++;
+  }
+}
 
 static int
 read_number_entries (uint64_t *nump, Elf *elf, size_t *offp, bool index64_p)
Only in elfutils-0.166/libelf: .elf_getarsym.c.swp
