gold: turn on --experimental-use-relr by default.

--- binutils-2.27/gold/options.h
+++ binutils-2.27/gold/options.h
@@ -770,7 +770,7 @@ class General_options
 		 N_("Exclude libraries from automatic export"),
 		 N_(("lib,lib ...")));
 
-  DEFINE_bool(experimental_use_relr, options::TWO_DASHES, '\0', false,
+  DEFINE_bool(experimental_use_relr, options::TWO_DASHES, '\0', true,
 	      N_("Generate RELR dynamic relocations"),
 	      N_("Do not generate RELR dynamic relocations"));
 
