--- string/bits/string3.h	2007-10-03 11:33:51.000000000 -0700
+++ string/bits/string3.h	2007-10-03 11:33:51.000000000 -0700
@@ -53,7 +53,7 @@
 }
 
 __extern_always_inline void *
-__NTH (memmove (void *__restrict __dest, __const void *__restrict __src,
+__NTH (memmove (void *__dest, __const void *__src,
 		size_t __len))
 {
   return __builtin___memmove_chk (__dest, __src, __len, __bos0 (__dest));
@@ -87,7 +87,7 @@
 
 #ifdef __USE_BSD
 __extern_always_inline void
-__NTH (bcopy (__const void *__restrict __src, void *__restrict __dest,
+__NTH (bcopy (__const void *__src, void *__dest,
 	      size_t __len))
 {
   (void) __builtin___memmove_chk (__dest, __src, __len, __bos0 (__dest));
