From 8129f03cb82129f7782753f7c05aa7821c80fe12 Mon Sep 17 00:00:00 2001
From: Peng Huang <shawn.p.huang@gmail.com>
Date: Fri, 22 Jul 2011 16:16:03 -0400
Subject: [PATCH] Always enable the new focused BusInputContext

---
 bus/ibusimpl.c |    3 +--
 1 files changed, 1 insertions(+), 2 deletions(-)

diff --git a/bus/ibusimpl.c b/bus/ibusimpl.c
index b6b2441..853465c 100644
--- a/bus/ibusimpl.c
+++ b/bus/ibusimpl.c
@@ -1203,8 +1203,7 @@ bus_ibus_impl_set_focused_context (BusIBusImpl     *ibus,
         /* attach engine to the focused context */
         if (engine != NULL) {
             bus_input_context_set_engine (context, engine);
-            if (bus_engine_proxy_is_enabled (engine))
-                bus_input_context_enable (context);
+            bus_input_context_enable (context);
             g_object_unref (engine);
         }
 
-- 
1.7.1

