diff -uNr ibus-1.3.99.20110422.orig/client/gtk2/ibusimcontext.c ibus-1.3.99.20110422/client/gtk2/ibusimcontext.c
--- ibus-1.3.99.20110422.orig/client/gtk2/ibusimcontext.c	2011-04-20 12:34:25.000000000 +0900
+++ ibus-1.3.99.20110422/client/gtk2/ibusimcontext.c	2011-04-22 15:03:58.717291695 +0900
@@ -270,7 +270,7 @@
 static void
 _request_surrounding_text (IBusIMContext *context, gboolean force)
 {
-    if (context->enable &&
+    if (context && context->enable &&
         (context->caps & IBUS_CAP_SURROUNDING_TEXT) != 0 &&
         (force ||
          ibus_input_context_needs_surrounding_text (context->ibuscontext))) {
