diff --git a/bus/panelproxy.c b/bus/panelproxy.c
index 3005909..7a0df6f 100644
--- a/bus/panelproxy.c
+++ b/bus/panelproxy.c
@@ -480,7 +480,7 @@ _context_set_cursor_location_cb (BusInputContext *context,
 
     g_return_if_fail (panel->focused_context == context);
 
-    bus_panel_proxy_set_cursor_location (panel, x, y, w, h);
+    // bus_panel_proxy_set_cursor_location (panel, x, y, w, h);
 }
 
 static void
