--- laptop-mode-tools-1.59/usr/sbin/laptop_mode
+++ laptop-mode-tools-1.59/usr/sbin/laptop_mode
@@ -593,7 +593,7 @@
     ON_AC=0
     for POWER_SUPPLY in /sys/class/power_supply/* ; do
 	    if [ -f $POWER_SUPPLY/type ] ; then
-		    if [ "$(cat $POWER_SUPPLY/type)" = "Mains" ] ;then
+		    if [ "$(cat $POWER_SUPPLY/type)" != "Battery" ] ;then
 			    log "VERBOSE" "Determining power state from $POWER_SUPPLY/online."
 			    FOUND_SYS_CLASS_POWER_SUPPLY_AC=1
 			    if [ "$(cat $POWER_SUPPLY/online)" = 1 ] ; then
