--- laptop-mode-tools-1.59/etc/laptop-mode/conf.d/intel-sata-powermgmt.conf
+++ laptop-mode-tools-1.59/etc/laptop-mode/conf.d/intel-sata-powermgmt.conf
@@ -24,5 +24,5 @@ CONTROL_INTEL_SATA_POWER="auto"

 # Handle power management of the Intel SATA deivce under specific circumstances
 BATT_ACTIVATE_SATA_POWER=1
-LM_AC_ACTIVATE_SATA_POWER=0
-NOLM_AC_ACTIVATE_SATA_POWER=0
+LM_AC_ACTIVATE_SATA_POWER=1
+NOLM_AC_ACTIVATE_SATA_POWER=1
