--- ./nss-3.14.3/mozilla/security/nss/lib/pk11wrap/secmod.h.org	2013-09-24 11:09:49.000000000 -0700
+++ ./nss-3.14.3/mozilla/security/nss/lib/pk11wrap/secmod.h	2013-09-24 11:10:04.000000000 -0700
@@ -2,7 +2,7 @@
  * License, v. 2.0. If a copy of the MPL was not distributed with this
  * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
 #ifndef _SECMOD_H_
-#define _SEDMOD_H_
+#define _SECMOD_H_
 #include "seccomon.h"
 #include "secmodt.h"
 #include "prinrval.h"
