==== chewing/chewing.gyp#58 - chewing/chewing.gyp ====
--- chewing/chewing.gyp#58	(revision 58)
+++ chewing/chewing.gyp	(working copy)
@@ -58,6 +58,9 @@
       ],
       'dependencies': [
         '../base/base.gyp:base',
+        '../session/session_base.gyp:config_handler',
+        '../session/session_base.gyp:genproto_session',
+        '../session/session_base.gyp:session_protocol',
         'chewing_session_server',
       ],
     },
