==== build_mozc.py#119 - build_mozc.py ====
--- build_mozc.py#119	2011-06-27 17:02:04.000000000 +0900
+++ build_mozc.py	2011-07-06 13:06:38.000000000 +0900
@@ -111,6 +111,15 @@ def GetGeneratorName():
   return generator
 
 
+def GetPkgConfigCommand():
+  """Get the pkg-config command name.
+
+  Returns:
+    pkg-config config command name.
+  """
+  return os.environ.get('PKG_CONFIG', 'pkg-config')
+
+
 def GenerateVersionFile(version_template_path, version_path):
   """Reads the version template file and stores it into version_path.
 
@@ -201,13 +210,13 @@ def GetGypFileNames(options):
     # Add ibus.gyp if ibus is installed.
     # Ubuntu 8.04 (Hardy) does not contain ibus package.
     try:
-      RunOrDie(['pkg-config', '--exists', 'ibus-1.0'])
+      RunOrDie([GetPkgConfigCommand(), '--exists', 'ibus-1.0'])
     except RunOrDieError:
       print 'ibus-1.0 was not found.'
       gyp_file_names.remove('%s/unix/ibus/ibus.gyp' % SRC_DIR)
     # Add gui.gyp if Qt libraries are installed.
     try:
-      RunOrDie(['pkg-config', '--exists', 'QtCore', 'QtGui'])
+      RunOrDie([GetPkgConfigCommand(), '--exists', 'QtCore', 'QtGui'])
       qt_found = True
     except RunOrDieError:
       print 'QtCore or QtGui was not found.'
@@ -218,7 +227,7 @@ def GetGypFileNames(options):
       gyp_file_names.remove('%s/gui/gui.gyp' % SRC_DIR)
     # Add scim.gyp if SCIM is installed.
     try:
-      RunOrDie(['pkg-config', '--exists', 'scim'])
+      RunOrDie([GetPkgConfigCommand(), '--exists', 'scim'])
     except RunOrDieError:
       print 'scim was not found.'
       gyp_file_names.remove('%s/unix/scim/scim.gyp' % SRC_DIR)
@@ -479,6 +488,15 @@ def GypMain(deps_file_name):
   else:
     command_line.extend(['-D', 'dictionary=desktop'])
 
+  # Specifying pkg-config command.  Some environment (such like
+  # cross-platform ChromeOS build) requires us to call a different
+  # command for pkg-config.  Here we catch the environment variable
+  # and use the specified command instead of actual pkg-config
+  # command.
+  if IsLinux():
+    command_line.extend([
+        '-D', 'pkg_config_command=%s' % GetPkgConfigCommand()])
+
   RunOrDie(command_line)
 
   # <internal>
==== chewing/chewing.gyp#10 - chewing/chewing.gyp ====
--- chewing/chewing.gyp#10	2011-06-23 15:01:32.000000000 +0900
+++ chewing/chewing.gyp	2011-07-06 13:06:38.000000000 +0900
@@ -63,7 +63,7 @@
             }],
            ],
           'cflags': [
-            '<!@(pkg-config --cflags <@(pkg_config_libs))',
+            '<!@(<(pkg_config_command) --cflags <@(pkg_config_libs))',
           ],
         },
         {
@@ -89,13 +89,13 @@
             }],
            ],
           'cflags': [
-            '<!@(pkg-config --cflags <@(pkg_config_libs))',
+            '<!@(<(pkg_config_command) --cflags <@(pkg_config_libs))',
           ],
           'libraries': [
-            '<!@(pkg-config --libs-only-l <@(pkg_config_libs))',
+            '<!@(<(pkg_config_command) --libs-only-l <@(pkg_config_libs))',
           ],
           'ldflags': [
-            '<!@(pkg-config --libs-only-L <@(pkg_config_libs))',
+            '<!@(<(pkg_config_command) --libs-only-L <@(pkg_config_libs))',
           ],
         },
       ],
==== gui/qt_libraries.gypi#5 - gui/qt_libraries.gypi ====
--- gui/qt_libraries.gypi#5	2011-06-23 16:27:36.000000000 +0900
+++ gui/qt_libraries.gypi	2011-07-06 13:06:38.000000000 +0900
@@ -15,7 +15,7 @@
       }, {
         'conditions': [
           ['OS=="linux"', {
-            'qt_cflags': ['<!@(pkg-config --cflags QtGui QtCore)'],
+            'qt_cflags': ['<!@(<(pkg_config_command) --cflags QtGui QtCore)'],
             'qt_include_dirs': [],
           }, {
             'qt_cflags': [],
@@ -72,7 +72,7 @@
             # pthread library is removed because it must not be specific to Qt,
             # and libpng12, which is missing on the list, is added because we
             # use PNG images.
-            '<!@(pkg-config --libs-only-L --libs-only-l'
+            '<!@(<(pkg_config_command) --libs-only-L --libs-only-l'
             ' xrender xrandr xcursor xfixes xinerama fontconfig freetype2'
             ' xi xt xext x11'
             ' sm ice'
@@ -81,7 +81,7 @@
           ],
         }, {
           'libraries': [
-            '<!@(pkg-config --libs QtGui QtCore)',
+            '<!@(<(pkg_config_command) --libs QtGui QtCore)',
           ],
         }],
       ],
==== gui/qt_moc.gypi#4 - gui/qt_moc.gypi ====
--- gui/qt_moc.gypi#4	2011-06-23 12:12:03.000000000 +0900
+++ gui/qt_moc.gypi	2011-07-06 13:06:38.000000000 +0900
@@ -9,7 +9,7 @@
       }, {
         'conditions': [
           ['OS=="linux"', {
-            'moc_path': '<!(pkg-config --variable=moc_location QtGui)',
+            'moc_path': '<!(<(pkg_config_command) --variable=moc_location QtGui)',
           }, {
             'moc_path': '<(qt_dir_env)/bin/moc<(EXECUTABLE_SUFFIX)',
           }],
==== gui/qt_rcc.gypi#3 - gui/qt_rcc.gypi ====
--- gui/qt_rcc.gypi#3	2011-06-22 16:30:42.000000000 +0900
+++ gui/qt_rcc.gypi	2011-07-06 13:06:38.000000000 +0900
@@ -10,7 +10,8 @@
         'conditions': [
           ['OS=="linux"', {
             # seems that --variable=rcc_location is not supported
-            'rcc_path': '<!(pkg-config --variable=exec_prefix QtGui)/bin/rcc',
+            'rcc_path':
+              '<!(<(pkg_config_command) --variable=exec_prefix QtGui)/bin/rcc',
           }, {
             'rcc_path': '<(qt_dir_env)/bin/rcc<(EXECUTABLE_SUFFIX)',
           }],
==== gui/qt_uic.gypi#4 - gui/qt_uic.gypi ====
--- gui/qt_uic.gypi#4	2011-06-22 16:30:42.000000000 +0900
+++ gui/qt_uic.gypi	2011-07-06 13:06:38.000000000 +0900
@@ -9,7 +9,8 @@
       }, {
         'conditions': [
           ['OS=="linux"', {
-            'uic_path': '<!(pkg-config --variable=uic_location QtGui)',
+            'uic_path':
+              '<!(<(pkg_config_command) --variable=uic_location QtGui)',
           }, {
             'uic_path': '<(qt_dir_env)/bin/uic<(EXECUTABLE_SUFFIX)',
           }],
==== gui/zinnia.gyp#1 - gui/zinnia.gyp ====
--- gui/zinnia.gyp#1	2011-02-28 12:01:58.000000000 +0900
+++ gui/zinnia.gyp	2011-07-06 13:06:38.000000000 +0900
@@ -23,7 +23,7 @@
             ['use_libzinnia==1', {
               'link_settings': {
                 'libraries': [
-                  ' <!@(pkg-config --libs zinnia)',
+                  ' <!@(<(pkg_config_command) --libs zinnia)',
                 ],
               },
             }, {  # OS=="linux" and use_libzinnia==0
==== gyp/common.gypi#73 - gyp/common.gypi ====
--- gyp/common.gypi#73	2011-06-27 21:05:26.000000000 +0900
+++ gyp/common.gypi	2011-07-06 13:06:38.000000000 +0900
@@ -117,6 +117,11 @@
     # project file called "coverage".
     'coverage%': 0,
 
+    # The pkg-config command to get the cflags/ldflags for Linux
+    # builds.  We make it customizable to allow building in a special
+    # environment such like cross-platform build.
+    'pkg_config_command%': 'pkg-config',
+
     'mozc_data_dir': '<(SHARED_INTERMEDIATE_DIR)',
   },
   'target_defaults': {
==== unix/ibus/ibus.gyp#24 - unix/ibus/ibus.gyp ====
--- unix/ibus/ibus.gyp#24	2011-06-07 18:37:03.000000000 +0900
+++ unix/ibus/ibus.gyp	2011-07-06 13:06:38.000000000 +0900
@@ -90,7 +90,7 @@
         }],
        ],
       'cflags': [
-        '<!@(pkg-config --cflags <@(pkg_config_libs))',
+        '<!@(<(pkg_config_command) --cflags <@(pkg_config_libs))',
       ],
     },
     {
@@ -159,13 +159,13 @@
         }],
        ],
       'cflags': [
-        '<!@(pkg-config --cflags <@(pkg_config_libs))',
+        '<!@(<(pkg_config_command) --cflags <@(pkg_config_libs))',
       ],
       'libraries': [
-        '<!@(pkg-config --libs-only-l <@(pkg_config_libs))',
+        '<!@(<(pkg_config_command) --libs-only-l <@(pkg_config_libs))',
       ],
       'ldflags': [
-        '<!@(pkg-config --libs-only-L <@(pkg_config_libs))',
+        '<!@(<(pkg_config_command) --libs-only-L <@(pkg_config_libs))',
       ],
     },
     {
@@ -203,13 +203,13 @@
         }],
        ],
       'cflags': [
-        '<!@(pkg-config --cflags <@(pkg_config_libs))',
+        '<!@(<(pkg_config_command) --cflags <@(pkg_config_libs))',
       ],
       'libraries': [
-        '<!@(pkg-config --libs-only-l <@(pkg_config_libs))',
+        '<!@(<(pkg_config_command) --libs-only-l <@(pkg_config_libs))',
       ],
       'ldflags': [
-        '<!@(pkg-config --libs-only-L <@(pkg_config_libs))',
+        '<!@(<(pkg_config_command) --libs-only-L <@(pkg_config_libs))',
       ],
       'variables': {
         'test_size': 'small',
==== unix/scim/scim.gyp#6 - unix/scim/scim.gyp ====
--- unix/scim/scim.gyp#6	2011-04-20 20:54:01.000000000 +0900
+++ unix/scim/scim.gyp	2011-07-06 13:06:38.000000000 +0900
@@ -16,7 +16,7 @@
       '../../session/session.gyp:session',
     ],
     'scim_defines': [
-      'SCIM_ICONDIR="<!@(pkg-config --variable=icondir scim)"',
+      'SCIM_ICONDIR="<!@(<(pkg_config_command) --variable=icondir scim)"',
     ]
   },
   'targets': [
@@ -36,7 +36,7 @@
         '<@(scim_dependencies)',
       ],
       'cflags': [
-        '<!@(pkg-config --cflags <@(pkg_config_libs))',
+        '<!@(<(pkg_config_command) --cflags <@(pkg_config_libs))',
       ],
       'defines': [
         '<@(scim_defines)',
@@ -52,7 +52,7 @@
         '<@(scim_dependencies)',
       ],
       'cflags': [
-        '<!@(pkg-config --cflags <@(pkg_config_libs))',
+        '<!@(<(pkg_config_command) --cflags <@(pkg_config_libs))',
       ],
       'defines': [
         '<@(scim_defines)',
@@ -72,13 +72,13 @@
         'scim_mozc',
       ],
       'cflags': [
-        '<!@(pkg-config --cflags <@(pkg_config_libs))',
+        '<!@(<(pkg_config_command) --cflags <@(pkg_config_libs))',
       ],
       'libraries': [
-        '<!@(pkg-config --libs-only-l <@(pkg_config_libs))',
+        '<!@(<(pkg_config_command) --libs-only-l <@(pkg_config_libs))',
       ],
       'ldflags': [
-        '<!@(pkg-config --libs-only-L <@(pkg_config_libs))',
+        '<!@(<(pkg_config_command) --libs-only-L <@(pkg_config_libs))',
         # Add a library path to the directory of libscim_mozc.so.
         '-Wl,-rpath,<(PRODUCT_DIR)/lib.target',
       ],
