From c42bccca5557f09d311f2cef4bfd2e58a6de92df Mon Sep 17 00:00:00 2001
From: Rob Clark <robclark@freedesktop.org>
Date: Sun, 13 Oct 2013 09:49:09 -0400
Subject: [PATCH 8/8] squash! libdrm: atomic mode set

Signed-off-by: Sean Paul <seanpaul@chromium.org>
---
 include/drm/drm_mode.h | 12 +++++++++---
 1 file changed, 9 insertions(+), 3 deletions(-)

diff --git a/include/drm/drm_mode.h b/include/drm/drm_mode.h
index fc8dd3b..cec906c 100644
--- a/include/drm/drm_mode.h
+++ b/include/drm/drm_mode.h
@@ -499,9 +499,15 @@ struct drm_mode_destroy_dumb {
 	__u32 handle;
 };
 
-#define DRM_MODE_ATOMIC_TEST_ONLY (1<<0)
-#define DRM_MODE_ATOMIC_EVENT (1<<1)
-#define DRM_MODE_ATOMIC_NONBLOCK (1<<2)
+/* page-flip flags are valid, plus: */
+#define DRM_MODE_ATOMIC_TEST_ONLY 0x0100
+#define DRM_MODE_ATOMIC_NONBLOCK  0x0200
+
+#define DRM_MODE_ATOMIC_FLAGS (\
+		DRM_MODE_PAGE_FLIP_EVENT |\
+		DRM_MODE_PAGE_FLIP_ASYNC |\
+		DRM_MODE_ATOMIC_TEST_ONLY |\
+		DRM_MODE_ATOMIC_NONBLOCK)
 
 /* FIXME come up with some sane error reporting mechanism? */
 struct drm_mode_atomic {
-- 
1.9.0.rc1.175.g0b1dcb5

