diff -ur ibus-pinyin-1.3.99.20110119.orig/src/PYMain.cc ibus-pinyin-1.3.99.20110119/src/PYMain.cc
--- ibus-pinyin-1.3.99.20110119.orig/src/PYMain.cc	2011-01-15 09:32:10.000000000 +0900
+++ ibus-pinyin-1.3.99.20110119/src/PYMain.cc	2011-06-09 13:29:30.755999213 +0900
@@ -82,7 +82,9 @@
 
     Database::init ();
     PinyinConfig::init (bus);
+#if 0
     BopomofoConfig::init (bus);
+#endif
 
     g_signal_connect ((IBusBus *)bus, "disconnected", G_CALLBACK (ibus_disconnected_cb), NULL);
 
@@ -119,6 +121,7 @@
 
     if (ibus) {
         ibus_factory_add_engine (factory, "pinyin", IBUS_TYPE_PINYIN_ENGINE);
+        ibus_factory_add_engine (factory, "pinyin-dv", IBUS_TYPE_PINYIN_ENGINE);
         ibus_factory_add_engine (factory, "bopomofo", IBUS_TYPE_PINYIN_ENGINE);
         ibus_bus_request_name (bus, "org.freedesktop.IBus.Pinyin", 0);
     }
diff -ur ibus-pinyin-1.3.99.20110119.orig/src/pinyin.xml.in.in ibus-pinyin-1.3.99.20110119/src/pinyin.xml.in.in
--- ibus-pinyin-1.3.99.20110119.orig/src/pinyin.xml.in.in	2011-01-15 09:32:10.000000000 +0900
+++ ibus-pinyin-1.3.99.20110119/src/pinyin.xml.in.in	2011-06-09 13:27:54.477750852 +0900
@@ -3,7 +3,7 @@
 <component>
 	<name>org.freedesktop.IBus.Pinyin</name>
 	<description>Pinyin Component</description>
-	<exec>${libexecdir}/ibus-engine-pinyin --ibus</exec>
+	<exec>/bin/env XDG_CACHE_HOME=/home/chronos/user/.cache ${libexecdir}/ibus-engine-pinyin --ibus</exec>
 	<version>@VERSION@</version>
 	<author>Peng Huang &lt;shawn.p.huang@gmail.com&gt;</author>
 	<license>GPL</license>
@@ -13,7 +13,7 @@
 	<engines>
 		<engine>
 			<name>pinyin</name>
-			<language>zh</language>
+			<language>zh-CN</language>
 			<license>GPL</license>
 			<author>Peng Huang &lt;shawn.p.huang@gmail.com&gt;
 BYVoid &lt;byvoid1@gmail.com&gt;
@@ -25,17 +25,17 @@
 			<rank>99</rank>
 		</engine>
 		<engine>
-			<name>bopomofo</name>
-			<language>zh</language>
+			<name>pinyin-dv</name>
+			<language>zh-CN</language>
 			<license>GPL</license>
-			<author>BYVoid &lt;byvoid1@gmail.com&gt;
-Peng Huang &lt;shawn.p.huang@gmail.com&gt;
+			<author>Peng Huang &lt;shawn.p.huang@gmail.com&gt;
+BYVoid &lt;byvoid1@gmail.com&gt;
 Peng Wu &lt;alexepico@gmail.com&gt;</author>
-			<icon>${pkgdatadir}/icons/ibus-bopomofo.svg</icon>
-			<layout>us</layout>
-			<longname>Bopomofo</longname>
-			<description>Bopomofo input method</description>
-			<rank>98</rank>
+			<icon>${pkgdatadir}/icons/ibus-pinyin.svg</icon>
+			<layout>us(dvorak)</layout>
+			<longname>Pinyin (for US Dvorak keyboard)</longname>
+			<description>Pinyin input method</description>
+			<rank>99</rank>
 		</engine>
 	</engines>
 
Only in ibus-pinyin-1.3.99.20110119/src: pinyin.xml.in.in.orig
