From 5e602178f00657db8a447db27b196df54834dfae Mon Sep 17 00:00:00 2001
From: Ben Chan <benchan@chromium.org>
Date: Tue, 10 Mar 2015 17:15:11 -0700
Subject: [PATCH] Force building network service from source on Chrome OS.

---
 build/module_args/mojo.gni | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/build/module_args/mojo.gni b/build/module_args/mojo.gni
index a20ef9e..7554663 100644
--- a/build/module_args/mojo.gni
+++ b/build/module_args/mojo.gni
@@ -13,4 +13,4 @@ build_mojo_shell_from_source = true
 
 # To build the network service from source, set this variable to true. To use
 # the prebuilt network service, omit this variable or set it to false.
-build_network_service_from_source = false
+build_network_service_from_source = true
-- 
2.2.0.rc0.207.ga3a616c

