diff -ru binutils-2.20.1.orig/gold/options.h binutils-2.20.1/gold/options.h
--- binutils-2.20.1.orig/gold/options.h	2010-01-14 19:48:42.000000000 +0900
+++ binutils-2.20.1/gold/options.h	2011-02-28 15:38:27.539876965 +0900
@@ -705,7 +705,8 @@
 		N_("Min fraction of empty buckets in dynamic hash"),
 		N_("FRACTION"));
 
-  DEFINE_enum(hash_style, options::TWO_DASHES, '\0', "sysv",
+  /* Chromium-local modification, make --hash-style=gnu the default. */
+  DEFINE_enum(hash_style, options::TWO_DASHES, '\0', "gnu",
 	      N_("Dynamic hash style"), N_("[sysv,gnu,both]"),
 	      {"sysv", "gnu", "both"});
 
diff -ru binutils-2.20.1.orig/ld/ldmain.c binutils-2.20.1/ld/ldmain.c
--- binutils-2.20.1.orig/ld/ldmain.c	2009-09-14 20:43:29.000000000 +0900
+++ binutils-2.20.1/ld/ldmain.c	2011-02-28 15:38:14.479131019 +0900
@@ -264,7 +264,8 @@
   link_info.keep_memory = TRUE;
   link_info.combreloc = TRUE;
   link_info.strip_discarded = TRUE;
-  link_info.emit_hash = TRUE;
+  /* Chromium-local modification, make --hash-style=gnu the default. */
+  link_info.emit_gnu_hash = TRUE;
   link_info.callbacks = &link_callbacks;
   link_info.input_bfds_tail = &link_info.input_bfds;
   /* SVR4 linkers seem to set DT_INIT and DT_FINI based on magic _init
