From 0be8cdbaac652d89d50a2b7a389dabbbf953ff75 Mon Sep 17 00:00:00 2001
From: Peng Huang <shawn.p.huang@gmail.com>
Date: Sun, 19 Sep 2010 16:13:49 +0800
Subject: [PATCH] Fix bug in ibus_m17n_engine_reset.

---
 src/engine.c |   32 +++++++++++++++++---------------
 1 files changed, 17 insertions(+), 15 deletions(-)

diff --git a/src/engine.c b/src/engine.c
index dcfbe48..53e9672 100644
--- a/src/engine.c
+++ b/src/engine.c
@@ -73,6 +73,8 @@ static void ibus_m17n_engine_callback       (MInputContext          *context,
 
 static IBusEngineClass *parent_class = NULL;
 static GHashTable      *im_table = NULL;
+/* The hash table associates MInputContext with IBusM17nEngine. */
+static GHashTable      *m17n_inputcontexts = NULL;
 
 GType
 ibus_m17n_engine_get_type (void)
@@ -129,6 +131,8 @@ ibus_m17n_engine_class_init (IBusM17NEngineClass *klass)
     engine_class->cursor_down = ibus_m17n_engine_cursor_down;
 
     engine_class->property_activate = ibus_m17n_engine_property_activate;
+
+    m17n_inputcontexts = g_hash_table_new (g_direct_hash, g_direct_equal);
 }
 
 static void
@@ -204,7 +208,10 @@ ibus_m17n_engine_constructor (GType                   type,
             mplist_put (im->driver.callback_list, Minput_candidates_done, ibus_m17n_engine_callback);
             mplist_put (im->driver.callback_list, Minput_set_spot, ibus_m17n_engine_callback);
             mplist_put (im->driver.callback_list, Minput_toggle, ibus_m17n_engine_callback);
-            mplist_put (im->driver.callback_list, Minput_reset, ibus_m17n_engine_callback);
+            /*
+              Does not set reset callback, uses the default callback in m17n.
+              mplist_put (im->driver.callback_list, Minput_reset, ibus_m17n_engine_callback);
+            */
             mplist_put (im->driver.callback_list, Minput_get_surrounding_text, ibus_m17n_engine_callback);
             mplist_put (im->driver.callback_list, Minput_delete_surrounding_text, ibus_m17n_engine_callback);
 
@@ -221,7 +228,7 @@ ibus_m17n_engine_constructor (GType                   type,
     }
 
     m17n->context = minput_create_ic (im, NULL);
-    mplist_add (m17n->context->plist, msymbol ("IBusEngine"), m17n);
+    g_hash_table_insert (m17n_inputcontexts, m17n->context, m17n);
 
     return (GObject *) m17n;
 }
@@ -246,6 +253,7 @@ ibus_m17n_engine_destroy (IBusM17NEngine *m17n)
     }
 
     if (m17n->context) {
+        g_hash_table_remove (m17n_inputcontexts, m17n->context);
         minput_destroy_ic (m17n->context);
         m17n->context = NULL;
     }
@@ -386,7 +394,7 @@ ibus_m17n_engine_focus_in (IBusEngine *engine)
     IBusM17NEngine *m17n = (IBusM17NEngine *) engine;
 
     ibus_engine_register_properties (engine, m17n->prop_list);
-    ibus_m17n_engine_process_key (m17n, msymbol ("input-focus-in"));
+    ibus_m17n_engine_process_key (m17n, Minput_focus_in);
 
     parent_class->focus_in (engine);
 }
@@ -396,7 +404,7 @@ ibus_m17n_engine_focus_out (IBusEngine *engine)
 {
     IBusM17NEngine *m17n = (IBusM17NEngine *) engine;
 
-    ibus_m17n_engine_process_key (m17n, msymbol ("input-focus-out"));
+    ibus_m17n_engine_process_key (m17n, Minput_focus_out);
 
     parent_class->focus_out (engine);
 }
@@ -407,7 +415,8 @@ ibus_m17n_engine_reset (IBusEngine *engine)
     IBusM17NEngine *m17n = (IBusM17NEngine *) engine;
 
     parent_class->reset (engine);
-    ibus_m17n_engine_focus_in (engine);
+
+    minput_reset_ic (m17n->context);
 }
 
 static void
@@ -477,7 +486,7 @@ ibus_m17n_engine_property_activate (IBusEngine  *engine,
 static void
 ibus_m17n_engine_update_lookup_table (IBusM17NEngine *m17n)
 {
-   ibus_lookup_table_clear (m17n->table);
+    ibus_lookup_table_clear (m17n->table);
 
     if (m17n->context->candidate_list && m17n->context->candidate_show) {
         IBusText *text;
@@ -550,16 +559,9 @@ ibus_m17n_engine_callback (MInputContext *context,
                            MSymbol        command)
 {
     IBusM17NEngine *m17n = NULL;
-    MPlist *p = NULL;
 
-    p = mplist_find_by_key (context->plist,  msymbol ("IBusEngine"));
-    if (p) {
-        m17n = (IBusM17NEngine *) mplist_value (p);
-    }
-
-    if (m17n == NULL) {
-        return;
-    }
+    m17n = g_hash_table_lookup (m17n_inputcontexts, context);
+    g_return_if_fail (m17n != NULL);
 
     if (command == Minput_preedit_start) {
         ibus_engine_hide_preedit_text ((IBusEngine *)m17n);
-- 
1.7.1

