--- src/engine.c.orig	2011-08-05 13:58:51.765377533 -0400
+++ src/engine.c	2011-08-05 14:00:03.625377567 -0400
@@ -368,10 +368,10 @@
 
             g_signal_connect (config, "value-changed",
                               G_CALLBACK(ibus_config_value_changed), m17n);
-            m17n->config_section = g_strdup_printf ("engine/M17N/%s/%s",
-                                                    lang, name);
         }
     }
+
+    m17n->config_section = g_strdup_printf ("engine/M17N/%s/%s", lang, name);
     g_strfreev (strv);
 
     if (im == NULL) {
