--- a/etc/portage/bashrc	2010-02-20 01:41:57.000000000 +0000
+++ b/etc/portage/bashrc	2010-02-20 00:33:16.000000000 +0000
@@ -8,10 +8,10 @@
 		cmdline+=" -e ^${EM}[[:space:]]";        
 	done
 	output="$( cd ${D} && scanelf -RmyBF%a . | grep -v ${cmdline} )"
 	[[ $output != "" ]] && { echo; echo "* Wrong EM_TYPE. Expected ${E_MACHINE}"; echo -e "${output}"; echo; exit 1; }
 }
 
 # We dont functionize this to avoid sandboxing.
 if [[ $EBUILD_PHASE == "postinst" ]]; then
-	[[ $EMERGE_FROM != binary ]] && cross-fix-root ${CHOST}
+	cross-fix-root ${SYSROOT}
 fi
