Rename <linux/switch.h> to <linux/switchdev.h>

diff -rup a/src/cli.c b/src/cli.c
--- a/src/cli.c
+++ b/src/cli.c
@@ -31,7 +31,7 @@
 #include <netlink/netlink.h>
 #include <netlink/genl/genl.h>
 #include <netlink/genl/ctrl.h>
-#include <linux/switch.h>
+#include <linux/switchdev.h>
 #include "swlib.h"
 
 enum {
diff -rup a/src/swlib.c b/src/swlib.c
--- a/src/swlib.c
+++ b/src/swlib.c
@@ -22,7 +22,7 @@
 #include <getopt.h>
 #include <sys/types.h>
 #include <sys/socket.h>
-#include <linux/switch.h>
+#include <linux/switchdev.h>
 #include "swlib.h"
 #include <netlink/netlink.h>
 #include <netlink/genl/genl.h>
diff -rup a/src/uci.c b/src/uci.c
--- a/src/uci.c
+++ b/src/uci.c
@@ -30,7 +30,7 @@
 #include <netlink/netlink.h>
 #include <netlink/genl/genl.h>
 #include <netlink/genl/ctrl.h>
-#include <linux/switch.h>
+#include <linux/switchdev.h>
 #include "swlib.h"
 
 #ifndef ARRAY_SIZE
