From d56a40e05a1f1d2ae2ed79a638bccdee4c3126ab Mon Sep 17 00:00:00 2001
From: Zach Reizner <zachr@google.com>
Date: Wed, 30 Aug 2017 13:48:04 -0700
Subject: [PATCH] recognize armv7a as an armv7 linux target

---
 src/lib.rs | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/src/lib.rs b/src/lib.rs
index 510df12..3aa4cac 100644
--- a/src/vendor/gcc/src/lib.rs
+++ b/src/vendor/gcc/src/lib.rs
@@ -1009,7 +1009,7 @@ impl Build {
                 }
 
                 // armv7 targets get to use armv7 instructions
-                if target.starts_with("armv7-") && target.contains("-linux-") {
+                if (target.starts_with("armv7-") || target.starts_with("armv7a-")) && target.contains("-linux-") {
                     cmd.args.push("-march=armv7-a".into());
                 }
 
-- 
2.14.1.581.gf28d330327-goog
