diff --git a/memconf/config.c b/memconf/config.c
index 34d8f7c..ba738c4 100644
--- a/memconf/config.c
+++ b/memconf/config.c
@@ -113,7 +113,7 @@ ibus_config_memconf_set_value (IBusConfigService *config,
     g_assert (value);
     g_assert (error == NULL || *error == NULL);
 
-    gchar *key = g_strdup_printf ("%s:%s", section, name);
+    gchar *key = g_strdup_printf ("%s%s", section, name);
 
 
     if (g_hash_table_lookup (
@@ -141,26 +141,22 @@ ibus_config_memconf_get_value (IBusConfigService *config,
     g_assert (name);
     g_assert (error == NULL || *error == NULL);
 
-    gchar *key = g_strdup_printf ("%s:%s", section, name);
-
-    g_hash_table_remove (IBUS_CONFIG_MEMCONF (config)->unread, key);
+    gchar *key = g_strdup_printf ("%s%s", section, name);
 
     GVariant *value = (GVariant *)g_hash_table_lookup (IBUS_CONFIG_MEMCONF (config)->values, key);
 
     if (value == NULL) {
         g_hash_table_insert (IBUS_CONFIG_MEMCONF (config)->unwritten,
                              g_strdup (key), NULL);
-    }
-
-    g_free (key);
-
-    if (value != NULL) {
+        if (error != NULL) {
+            *error = g_error_new (G_DBUS_ERROR, G_DBUS_ERROR_FAILED,
+                                  "Config value [%s:%s] does not exist.", section, name);
+        }
+    } else {
+        g_hash_table_remove (IBUS_CONFIG_MEMCONF (config)->unread, key);
         g_variant_ref (value);
     }
-    else if (error != NULL) {
-        *error = g_error_new (G_DBUS_ERROR, G_DBUS_ERROR_FAILED,
-                              "Config value [%s:%s] does not exist.", section, name);
-    }
+    g_free (key);
     return value;
 }
 
@@ -175,11 +171,11 @@ ibus_config_memconf_unset_value (IBusConfigService *config,
     g_assert (name);
     g_assert (error == NULL || *error == NULL);
 
-    gchar *key = g_strdup_printf ("%s:%s", section, name);
+    gchar *key = g_strdup_printf ("%s%s", section, name);
     gboolean retval = g_hash_table_remove (IBUS_CONFIG_MEMCONF (config)->values, key);
-    g_free (key);
 
     if (retval) {
+        g_hash_table_remove (IBUS_CONFIG_MEMCONF (config)->unread, key);
     }
     else {
         if (error && *error) {
@@ -187,6 +183,7 @@ ibus_config_memconf_unset_value (IBusConfigService *config,
                                   "Config value [%s:%s] does not exist.", section, name);
         }
     }
+    g_free (key);
     return retval;
 }
 
