diff --git a/src/ibusconfig.c b/src/ibusconfig.c
index 62c434e..e8dd21b 100644
--- a/src/ibusconfig.c
+++ b/src/ibusconfig.c
@@ -323,9 +323,12 @@ ibus_config_get_unused (IBusConfig  *config,
         return FALSE;
     }
 
-    *unread = g_variant_get_child_value (result, 0);
-    *unwritten = g_variant_get_child_value (result, 1);
-
+    GVariant *value = g_variant_get_child_value (result, 0);
+    if (value != NULL) {
+        *unread = g_variant_get_child_value (value, 0);
+        *unwritten = g_variant_get_child_value (value, 1);
+        g_variant_unref (value);
+    }
     g_variant_unref (result);
 
     return TRUE;
diff --git a/src/ibusconfigservice.c b/src/ibusconfigservice.c
index 412cc8b..82c22c7 100644
--- a/src/ibusconfigservice.c
+++ b/src/ibusconfigservice.c
@@ -256,7 +256,8 @@ ibus_config_service_service_method_call (IBusService           *service,
         }
         return;
     }
-    else if (g_strcmp0 (method_name, "GetUnused") == 0) {
+
+    if (g_strcmp0 (method_name, "GetUnused") == 0) {
         GVariant *unread = NULL;
         GVariant *unwritten = NULL;
         gboolean retval;
@@ -266,7 +267,7 @@ ibus_config_service_service_method_call (IBusService           *service,
                         config, &unread, &unwritten, &error);
         if (retval) {
             g_dbus_method_invocation_return_value (invocation,
-                            g_variant_new ("(@as@as)", unread, unwritten));
+                            g_variant_new ("((@as@as))", unread, unwritten));
         }
         else {
             g_dbus_method_invocation_return_gerror (invocation, error);
