remove target flags from host compiler
--- fc-case/Makefile.am.bak	2013-04-05 11:16:59.062934420 -0700
+++ fc-case/Makefile.am	2013-04-05 11:17:30.983145425 -0700
@@ -24,7 +24,7 @@
 
 CC = @CC_FOR_BUILD@
 EXEEXT = @EXEEXT_FOR_BUILD@
-LINK = $(CCLD) $(AM_CFLAGS) $(CFLAGS) $(AM_LDFLAGS) $(LDFLAGS) -o $@
+LINK = $(CCLD) $(AM_CFLAGS) $(AM_LDFLAGS) -o $@
 
 INCLUDES=-I${top_srcdir}/src -I${top_srcdir} $(WARN_CFLAGS)
 
--- fc-case/Makefile.in.bak	2013-04-05 11:17:06.162981353 -0700
+++ fc-case/Makefile.in	2013-04-05 11:18:17.643453883 -0700
@@ -81,7 +81,7 @@
 am__depfiles_maybe = depfiles
 am__mv = mv -f
 COMPILE = $(CC) $(DEFS) $(DEFAULT_INCLUDES) $(INCLUDES) $(AM_CPPFLAGS) \
-	$(CPPFLAGS) $(AM_CFLAGS) $(CFLAGS)
+	$(AM_CFLAGS)
 CCLD = $(CC)
 SOURCES = fc-case.c
 DIST_SOURCES = fc-case.c
@@ -237,7 +237,7 @@
 top_build_prefix = @top_build_prefix@
 top_builddir = @top_builddir@
 top_srcdir = @top_srcdir@
-LINK = $(CCLD) $(AM_CFLAGS) $(CFLAGS) $(AM_LDFLAGS) $(LDFLAGS) -o $@
+LINK = $(CCLD) $(AM_CFLAGS) $(AM_LDFLAGS) -o $@
 INCLUDES = -I${top_srcdir}/src -I${top_srcdir} $(WARN_CFLAGS)
 TMPL = fccase.tmpl.h
 STMPL = ${top_srcdir}/fc-case/${TMPL}
--- fc-lang/Makefile.am.bak	2013-04-05 11:24:33.775931018 -0700
+++ fc-lang/Makefile.am	2013-04-05 11:26:27.996681039 -0700
@@ -24,7 +24,7 @@
 
 CC = @CC_FOR_BUILD@
 EXEEXT = @EXEEXT_FOR_BUILD@
-LINK = $(CCLD) $(AM_CFLAGS) $(CFLAGS) $(AM_LDFLAGS) $(LDFLAGS) -o $@
+LINK = $(CCLD) $(AM_CFLAGS) $(AM_LDFLAGS) -o $@
 
 INCLUDES=-I${top_srcdir}/src -I${top_srcdir} $(WARN_CFLAGS)
 
--- fc-lang/Makefile.in.bak	2013-04-05 11:24:26.325882104 -0700
+++ fc-lang/Makefile.in	2013-04-05 11:26:00.096497813 -0700
@@ -79,7 +79,7 @@
 am__depfiles_maybe = depfiles
 am__mv = mv -f
 COMPILE = $(CC) $(DEFS) $(DEFAULT_INCLUDES) $(INCLUDES) $(AM_CPPFLAGS) \
-	$(CPPFLAGS) $(AM_CFLAGS) $(CFLAGS)
+	$(AM_CFLAGS)
 CCLD = $(CC)
 SOURCES = fc-lang.c
 DIST_SOURCES = fc-lang.c
@@ -234,7 +234,7 @@
 top_build_prefix = @top_build_prefix@
 top_builddir = @top_builddir@
 top_srcdir = @top_srcdir@
-LINK = $(CCLD) $(AM_CFLAGS) $(CFLAGS) $(AM_LDFLAGS) $(LDFLAGS) -o $@
+LINK = $(CCLD) $(AM_CFLAGS) $(AM_LDFLAGS) -o $@
 INCLUDES = -I${top_srcdir}/src -I${top_srcdir} $(WARN_CFLAGS)
 TMPL = fclang.tmpl.h
 STMPL = ${top_srcdir}/fc-lang/fclang.tmpl.h
--- fc-glyphname/Makefile.am.bak	2013-04-05 11:29:09.507741970 -0700
+++ fc-glyphname/Makefile.am	2013-04-05 11:31:39.888730168 -0700
@@ -24,7 +24,7 @@
 
 CC = @CC_FOR_BUILD@
 EXEEXT = @EXEEXT_FOR_BUILD@
-LINK = $(CCLD) $(AM_CFLAGS) $(CFLAGS) $(AM_LDFLAGS) $(LDFLAGS) -o $@
+LINK = $(CCLD) $(AM_CFLAGS) $(AM_LDFLAGS) -o $@
 
 INCLUDES=-I${top_srcdir}/src -I${top_srcdir} $(WARN_CFLAGS)
 
--- fc-glyphname/Makefile.in.bak	2013-04-05 11:29:16.867790324 -0700
+++ fc-glyphname/Makefile.in	2013-04-05 11:30:19.478201719 -0700
@@ -79,7 +79,7 @@
 am__depfiles_maybe = depfiles
 am__mv = mv -f
 COMPILE = $(CC) $(DEFS) $(DEFAULT_INCLUDES) $(INCLUDES) $(AM_CPPFLAGS) \
-	$(CPPFLAGS) $(AM_CFLAGS) $(CFLAGS)
+	$(AM_CFLAGS)
 CCLD = $(CC)
 SOURCES = fc-glyphname.c
 DIST_SOURCES = fc-glyphname.c
@@ -234,7 +234,7 @@
 top_build_prefix = @top_build_prefix@
 top_builddir = @top_builddir@
 top_srcdir = @top_srcdir@
-LINK = $(CCLD) $(AM_CFLAGS) $(CFLAGS) $(AM_LDFLAGS) $(LDFLAGS) -o $@
+LINK = $(CCLD) $(AM_CFLAGS) $(AM_LDFLAGS) -o $@
 INCLUDES = -I${top_srcdir}/src -I${top_srcdir} $(WARN_CFLAGS)
 TMPL = fcglyphname.tmpl.h
 STMPL = ${top_srcdir}/fc-glyphname/${TMPL}
--- fc-arch/Makefile.am.bak     2013-04-05 11:33:07.059303153 -0700
+++ fc-arch/Makefile.am 2013-04-05 11:33:27.299436217 -0700
@@ -24,7 +24,7 @@

 CC = @CC_FOR_BUILD@
 EXEEXT = @EXEEXT_FOR_BUILD@
-LINK = $(CCLD) $(AM_CFLAGS) $(CFLAGS) $(AM_LDFLAGS) $(LDFLAGS) -o $@
+LINK = $(CCLD) $(AM_CFLAGS) $(AM_LDFLAGS) -o $@

 INCLUDES=-I${top_srcdir}/src -I${top_srcdir} $(WARN_CFLAGS)
                                                                   
--- fc-arch/Makefile.in.bak	2013-04-05 11:33:13.209343587 -0700
+++ fc-arch/Makefile.in	2013-04-05 11:33:59.759649618 -0700
@@ -79,7 +79,7 @@
 am__depfiles_maybe = depfiles
 am__mv = mv -f
 COMPILE = $(CC) $(DEFS) $(DEFAULT_INCLUDES) $(INCLUDES) $(AM_CPPFLAGS) \
-	$(CPPFLAGS) $(AM_CFLAGS) $(CFLAGS)
+	$(AM_CFLAGS)
 CCLD = $(CC)
 SOURCES = fc-arch.c
 DIST_SOURCES = fc-arch.c
@@ -234,7 +234,7 @@
 top_build_prefix = @top_build_prefix@
 top_builddir = @top_builddir@
 top_srcdir = @top_srcdir@
-LINK = $(CCLD) $(AM_CFLAGS) $(CFLAGS) $(AM_LDFLAGS) $(LDFLAGS) -o $@
+LINK = $(CCLD) $(AM_CFLAGS) $(AM_LDFLAGS) -o $@
 INCLUDES = -I${top_srcdir}/src -I${top_srcdir} $(WARN_CFLAGS)
 TMPL = fcarch.tmpl.h
 STMPL = ${top_srcdir}/fc-arch/fcarch.tmpl.h
