diff -uNr xcb-proto-1.6.old/xcbgen/expr.py xcb-proto-1.6/xcbgen/expr.py
--- xcb-proto-1.6.old/xcbgen/expr.py	2012-06-05 16:06:46.971402977 -0700
+++ xcb-proto-1.6/xcbgen/expr.py	2012-06-05 16:07:16.191953201 -0700
@@ -26,7 +26,7 @@
     Represents a mathematical expression for a list length or exprfield.
 
     Public fields:
-    op is the operation (text +,*,/,<<) or None.
+    op is the operation (text +,*,/,<<,~) or None.
     lhs and rhs are the sub-Expressions if op is set.
     lenfield_name is the name of the length field, or None for request lists.
     lenfield is the Field object for the length field, or None.
@@ -75,13 +75,20 @@
             if self.lenfield_name == None:
                 self.lenfield_name = self.rhs.lenfield_name
 
+        elif elt.tag == 'unop':
+            # Op field.  Need to recurse.
+            self.op = elt.get('op')
+            self.rhs = Expression(list(elt)[0], parent)
+
+            self.lenfield_name = self.rhs.lenfield_name
+            
         elif elt.tag == 'value':
             # Constant expression
-            self.nmemb = int(elt.text)
+            self.nmemb = int(elt.text, 0)
 
         else:
             # Notreached
-            raise Exception('XXX')
+            raise Exception("undefined tag '%s'" % elt.tag)
 
 
     def fixed_size(self):
